{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "3ee91ff5",
   "metadata": {},
   "outputs": [],
   "source": [
    "%load_ext autoreload\n",
    "%autoreload 2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "15865688",
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ff5d5753",
   "metadata": {},
   "source": [
    "## Read data"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "235e3bf2",
   "metadata": {},
   "source": [
    "Load eCLIP-seq sequences"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "abe42208",
   "metadata": {},
   "outputs": [],
   "source": [
    "eclip = pd.read_csv(\"HNRNPC/HNRNPC_eclip.txt\", sep='\\t', header=None)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "32e0bc25",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr7:138151219-138151300</td>\n",
       "      <td>TTTTTTTGGAGACAGCATCTCACTGTGTTGCCCATACTGGAGTGCA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr7:139051735-139051798</td>\n",
       "      <td>ATTTTTTGGAGACAGAGTCTCTGTCTCCCAGGCGTGAAATGCAGTG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr7:918222-918272</td>\n",
       "      <td>TTTTTTGCGACGGAGTCTCGCTCTGTCCTCTGTCCCAGGCTGGAGT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr7:138154088-138154142</td>\n",
       "      <td>TTTTTGAGACAGTGTCTCGTGCAGTGGCCATCTTGGCTCACTGCAA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr7:138167839-138167908</td>\n",
       "      <td>TTTTTGTAGCGGGGGGGGTCTCACTTTGTTGCCCAGGCTGGTGTCA...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                          0                                                  1\n",
       "0  chr7:138151219-138151300  TTTTTTTGGAGACAGCATCTCACTGTGTTGCCCATACTGGAGTGCA...\n",
       "1  chr7:139051735-139051798  ATTTTTTGGAGACAGAGTCTCTGTCTCCCAGGCGTGAAATGCAGTG...\n",
       "2        chr7:918222-918272  TTTTTTGCGACGGAGTCTCGCTCTGTCCTCTGTCCCAGGCTGGAGT...\n",
       "3  chr7:138154088-138154142  TTTTTGAGACAGTGTCTCGTGCAGTGGCCATCTTGGCTCACTGCAA...\n",
       "4  chr7:138167839-138167908  TTTTTGTAGCGGGGGGGGTCTCACTTTGTTGCCCAGGCTGGTGTCA..."
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "eclip.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8e72135a",
   "metadata": {},
   "source": [
    "Load RBNS binding affinities"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "7bc981f7",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>[hnRNPC]</th>\n",
       "      <th>0</th>\n",
       "      <th>5</th>\n",
       "      <th>20</th>\n",
       "      <th>80</th>\n",
       "      <th>320</th>\n",
       "      <th>1300</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>TTTT</td>\n",
       "      <td>1.050108</td>\n",
       "      <td>1.213564</td>\n",
       "      <td>1.234794</td>\n",
       "      <td>2.755430</td>\n",
       "      <td>11.477239</td>\n",
       "      <td>5.958550</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ATTT</td>\n",
       "      <td>1.052096</td>\n",
       "      <td>1.124278</td>\n",
       "      <td>0.990108</td>\n",
       "      <td>1.255270</td>\n",
       "      <td>2.505993</td>\n",
       "      <td>1.978384</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>CTTT</td>\n",
       "      <td>1.094396</td>\n",
       "      <td>1.111101</td>\n",
       "      <td>1.087047</td>\n",
       "      <td>1.272162</td>\n",
       "      <td>2.264619</td>\n",
       "      <td>1.867528</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>TTTA</td>\n",
       "      <td>1.023649</td>\n",
       "      <td>1.074103</td>\n",
       "      <td>0.975594</td>\n",
       "      <td>1.183402</td>\n",
       "      <td>2.129294</td>\n",
       "      <td>1.799562</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>TTTC</td>\n",
       "      <td>1.097361</td>\n",
       "      <td>1.150883</td>\n",
       "      <td>1.052075</td>\n",
       "      <td>1.217492</td>\n",
       "      <td>1.999443</td>\n",
       "      <td>1.715131</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>251</th>\n",
       "      <td>TCGG</td>\n",
       "      <td>0.830022</td>\n",
       "      <td>0.846412</td>\n",
       "      <td>0.842917</td>\n",
       "      <td>0.818757</td>\n",
       "      <td>0.693297</td>\n",
       "      <td>0.672110</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>252</th>\n",
       "      <td>TCGT</td>\n",
       "      <td>0.849285</td>\n",
       "      <td>0.844630</td>\n",
       "      <td>0.870984</td>\n",
       "      <td>0.837076</td>\n",
       "      <td>0.685304</td>\n",
       "      <td>0.681841</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>253</th>\n",
       "      <td>CGTC</td>\n",
       "      <td>0.845628</td>\n",
       "      <td>0.842135</td>\n",
       "      <td>0.878028</td>\n",
       "      <td>0.833263</td>\n",
       "      <td>0.678240</td>\n",
       "      <td>0.691214</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>254</th>\n",
       "      <td>ATCG</td>\n",
       "      <td>0.849546</td>\n",
       "      <td>0.843249</td>\n",
       "      <td>0.866847</td>\n",
       "      <td>0.828345</td>\n",
       "      <td>0.666320</td>\n",
       "      <td>0.674459</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>255</th>\n",
       "      <td>GTCG</td>\n",
       "      <td>0.782238</td>\n",
       "      <td>0.781878</td>\n",
       "      <td>0.814882</td>\n",
       "      <td>0.774402</td>\n",
       "      <td>0.633157</td>\n",
       "      <td>0.619229</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>256 rows Ã— 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    [hnRNPC]         0         5        20        80        320      1300\n",
       "0       TTTT  1.050108  1.213564  1.234794  2.755430  11.477239  5.958550\n",
       "1       ATTT  1.052096  1.124278  0.990108  1.255270   2.505993  1.978384\n",
       "2       CTTT  1.094396  1.111101  1.087047  1.272162   2.264619  1.867528\n",
       "3       TTTA  1.023649  1.074103  0.975594  1.183402   2.129294  1.799562\n",
       "4       TTTC  1.097361  1.150883  1.052075  1.217492   1.999443  1.715131\n",
       "..       ...       ...       ...       ...       ...        ...       ...\n",
       "251     TCGG  0.830022  0.846412  0.842917  0.818757   0.693297  0.672110\n",
       "252     TCGT  0.849285  0.844630  0.870984  0.837076   0.685304  0.681841\n",
       "253     CGTC  0.845628  0.842135  0.878028  0.833263   0.678240  0.691214\n",
       "254     ATCG  0.849546  0.843249  0.866847  0.828345   0.666320  0.674459\n",
       "255     GTCG  0.782238  0.781878  0.814882  0.774402   0.633157  0.619229\n",
       "\n",
       "[256 rows x 7 columns]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "four_mers = pd.read_csv(\"HNRNPC/HNRNPC_4.tsv\", sep='\\t')\n",
    "five_mers = pd.read_csv(\"HNRNPC/HNRNPC_5.tsv\", sep='\\t',)\n",
    "six_mers = pd.read_csv(\"HNRNPC/HNRNPC_6.tsv\", sep='\\t',)\n",
    "four_mers"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "af6ffd4e",
   "metadata": {},
   "source": [
    "Load knockdown differential rna-seq data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "4740db51",
   "metadata": {},
   "outputs": [],
   "source": [
    "diff_exp_HNRNPC = pd.read_csv(\"HNRNPC/HNRNPC.tsv\", sep='\\t',)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "ab40bc90",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>test_id</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene</th>\n",
       "      <th>locus</th>\n",
       "      <th>sample_1</th>\n",
       "      <th>sample_2</th>\n",
       "      <th>status</th>\n",
       "      <th>value_1</th>\n",
       "      <th>value_2</th>\n",
       "      <th>log2(fold_change)</th>\n",
       "      <th>test_stat</th>\n",
       "      <th>p_value</th>\n",
       "      <th>q_value</th>\n",
       "      <th>significant</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>XLOC_000001</td>\n",
       "      <td>XLOC_000001</td>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>chr1:11868-31109</td>\n",
       "      <td>HNRNPC-BGHLV20-HepG2</td>\n",
       "      <td>Control</td>\n",
       "      <td>NOTEST</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.021689</td>\n",
       "      <td>inf</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>no</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>XLOC_000002</td>\n",
       "      <td>XLOC_000002</td>\n",
       "      <td>MIR1302-11</td>\n",
       "      <td>chr1:11868-31109</td>\n",
       "      <td>HNRNPC-BGHLV20-HepG2</td>\n",
       "      <td>Control</td>\n",
       "      <td>NOTEST</td>\n",
       "      <td>0.017697</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>-inf</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>no</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>XLOC_000003</td>\n",
       "      <td>XLOC_000003</td>\n",
       "      <td>OR4G4P</td>\n",
       "      <td>chr1:52472-54936</td>\n",
       "      <td>HNRNPC-BGHLV20-HepG2</td>\n",
       "      <td>Control</td>\n",
       "      <td>NOTEST</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>no</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>XLOC_000004</td>\n",
       "      <td>XLOC_000004</td>\n",
       "      <td>OR4G11P</td>\n",
       "      <td>chr1:62947-63887</td>\n",
       "      <td>HNRNPC-BGHLV20-HepG2</td>\n",
       "      <td>Control</td>\n",
       "      <td>NOTEST</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>no</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>XLOC_000005</td>\n",
       "      <td>XLOC_000005</td>\n",
       "      <td>OR4F5</td>\n",
       "      <td>chr1:69090-70008</td>\n",
       "      <td>HNRNPC-BGHLV20-HepG2</td>\n",
       "      <td>Control</td>\n",
       "      <td>NOTEST</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>no</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       test_id      gene_id        gene             locus  \\\n",
       "0  XLOC_000001  XLOC_000001     DDX11L1  chr1:11868-31109   \n",
       "1  XLOC_000002  XLOC_000002  MIR1302-11  chr1:11868-31109   \n",
       "2  XLOC_000003  XLOC_000003      OR4G4P  chr1:52472-54936   \n",
       "3  XLOC_000004  XLOC_000004     OR4G11P  chr1:62947-63887   \n",
       "4  XLOC_000005  XLOC_000005       OR4F5  chr1:69090-70008   \n",
       "\n",
       "               sample_1 sample_2  status   value_1   value_2  \\\n",
       "0  HNRNPC-BGHLV20-HepG2  Control  NOTEST  0.000000  0.021689   \n",
       "1  HNRNPC-BGHLV20-HepG2  Control  NOTEST  0.017697  0.000000   \n",
       "2  HNRNPC-BGHLV20-HepG2  Control  NOTEST  0.000000  0.000000   \n",
       "3  HNRNPC-BGHLV20-HepG2  Control  NOTEST  0.000000  0.000000   \n",
       "4  HNRNPC-BGHLV20-HepG2  Control  NOTEST  0.000000  0.000000   \n",
       "\n",
       "   log2(fold_change)  test_stat  p_value  q_value significant  \n",
       "0                inf        0.0      1.0      1.0          no  \n",
       "1               -inf        0.0      1.0      1.0          no  \n",
       "2                0.0        0.0      1.0      1.0          no  \n",
       "3                0.0        0.0      1.0      1.0          no  \n",
       "4                0.0        0.0      1.0      1.0          no  "
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "diff_exp_HNRNPC.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "d0e331a6",
   "metadata": {},
   "outputs": [],
   "source": [
    "# get other diff expr for subsetting to shared genes later\n",
    "diff_exp_SRSF9 = pd.read_csv(\"SRSF9/SRSF9.tsv\", sep='\\t',)\n",
    "diff_exp_PCBP1 = pd.read_csv(\"PCBP1/PCBP1.tsv\", sep='\\t',)\n",
    "diff_exp_TIA1 = pd.read_csv(\"TIA1/TIA1.tsv\", sep='\\t',)\n",
    "diff_exp_TRA2A = pd.read_csv(\"TRA2A/TRA2A.tsv\", sep='\\t',)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1612ca75",
   "metadata": {},
   "source": [
    "## Generate positive pairs"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b0787ccc",
   "metadata": {},
   "source": [
    "EClip positive binding sites"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "592b1f36",
   "metadata": {},
   "outputs": [],
   "source": [
    "in_vivo_pos_seqs = eclip[1].values"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "18057862",
   "metadata": {},
   "outputs": [],
   "source": [
    "# filter for length and unique values\n",
    "in_vivo_pos_seqs = [s for s in in_vivo_pos_seqs if len(s) >= 10]\n",
    "in_vivo_pos_seqs = list(set(in_vivo_pos_seqs))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "adf61072",
   "metadata": {},
   "source": [
    "RBNS significant binding affinity"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "fd120254",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_signficant_sequences(df):\n",
    "    num_cols = df.columns[1:]\n",
    "    filtered = df[df[num_cols].gt(2).any(axis=1)]\n",
    "    pairs = [\n",
    "        (row[df.columns[0]], tuple(row[num_cols]))\n",
    "        for _, row in filtered.iterrows()\n",
    "    ]\n",
    "    return pairs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "ed65dc0f",
   "metadata": {},
   "outputs": [],
   "source": [
    "pos_four_mers = get_signficant_sequences(four_mers)\n",
    "pos_five_mers = get_signficant_sequences(five_mers)\n",
    "pos_six_mers = get_signficant_sequences(six_mers)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "f995d1a6",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[('TTTTTT', (1.071962, 1.692866, 3.672444, 17.832708, 94.802148, 35.212764)),\n",
       " ('ATTTTT', (1.076105, 1.348466, 1.481458, 3.968517, 19.883022, 10.642442)),\n",
       " ('CTTTTT', (1.067404, 1.24024, 1.319701, 3.161729, 14.967689, 8.264041)),\n",
       " ('TTTTTA', (1.023226, 1.254425, 1.27889, 2.985524, 14.87552, 8.87071)),\n",
       " ('TTTTTC', (1.121678, 1.305399, 1.288073, 2.738826, 12.616572, 7.35691)),\n",
       " ('TATTTT', (1.023793, 1.20788, 1.17275, 2.489906, 10.644826, 6.290213)),\n",
       " ('TTTTTG', (0.996192, 1.196576, 1.128267, 2.251156, 9.510792, 5.751316)),\n",
       " ('GTTTTT', (1.00087, 1.27983, 1.1027, 2.268313, 9.301687, 5.271012)),\n",
       " ('TTTTAT', (1.046737, 1.178148, 1.108257, 1.996217, 7.586432, 4.897009)),\n",
       " ('TCTTTT', (1.089925, 1.15815, 1.188027, 1.980937, 6.885662, 4.33642)),\n",
       " ('AATTTT', (1.123924, 1.302106, 1.129826, 1.970238, 6.694241, 4.646509)),\n",
       " ('CATTTT', (1.046779, 1.159885, 1.076534, 1.901975, 6.610391, 4.293748)),\n",
       " ('TTTTCT', (1.13576, 1.25046, 1.147984, 1.876544, 6.591292, 4.298317)),\n",
       " ('TTATTT', (1.053787, 1.141478, 1.093976, 1.87168, 6.521348, 4.10437)),\n",
       " ('ACTTTT', (1.088376, 1.1713, 1.120796, 1.811183, 5.857198, 3.996817)),\n",
       " ('TTTATT', (1.083366, 1.142882, 1.114286, 1.765261, 5.720871, 3.707175)),\n",
       " ('TTTTAC', (1.062612, 1.161388, 1.061676, 1.648229, 5.43187, 3.927529)),\n",
       " ('CCTTTT', (1.202271, 1.242453, 1.170172, 1.744092, 5.255853, 3.74814)),\n",
       " ('TTCTTT', (1.183498, 1.204123, 1.233877, 1.732844, 5.024888, 3.356681)),\n",
       " ('TGTTTT', (1.00434, 1.174187, 1.013673, 1.608081, 4.913791, 3.195835)),\n",
       " ('TTTTAA', (1.045768, 1.155133, 1.050492, 1.64703, 4.883637, 3.714949)),\n",
       " ('TTTTCC', (1.30157, 1.415685, 1.236204, 1.762325, 4.879859, 3.619581)),\n",
       " ('TTTCTT', (1.1614, 1.227511, 1.18246, 1.670956, 4.853057, 3.302071)),\n",
       " ('TTTTGT', (0.973967, 1.091387, 1.019211, 1.5346, 4.604834, 3.085827)),\n",
       " ('TTTTCA', (1.107123, 1.20631, 1.092615, 1.564608, 4.354419, 3.213496)),\n",
       " ('TTTTGC', (1.024522, 1.101646, 1.033033, 1.475901, 4.138848, 3.061661)),\n",
       " ('GCTTTT', (0.994315, 1.065108, 1.022309, 1.491927, 4.094022, 2.859432)),\n",
       " ('AGTTTT', (1.089128, 1.271, 1.091084, 1.526687, 3.722683, 2.777795)),\n",
       " ('TTTTAG', (0.966068, 1.093385, 0.940434, 1.329479, 3.519792, 2.662876)),\n",
       " ('TTGTTT', (1.025964, 1.102577, 1.050802, 1.371497, 3.321745, 2.29336)),\n",
       " ('TTTTGA', (1.083472, 1.191468, 1.051302, 1.402626, 3.312505, 2.622345)),\n",
       " ('TTTGTT', (1.01907, 1.09595, 1.067826, 1.377193, 3.241837, 2.254773)),\n",
       " ('GATTTT', (1.049193, 1.194336, 0.943174, 1.320831, 3.229903, 2.360217)),\n",
       " ('ATATTT', (1.036791, 1.133854, 0.996907, 1.402496, 3.2146, 2.496918)),\n",
       " ('CTATTT', (1.063153, 1.109794, 1.014948, 1.347273, 3.113188, 2.394362)),\n",
       " ('TAATTT', (1.063995, 1.163698, 0.98004, 1.356062, 3.088496, 2.3978)),\n",
       " ('ATTTTA', (1.037355, 1.161858, 0.941295, 1.326071, 3.077406, 2.710868)),\n",
       " ('TACTTT', (1.072751, 1.104316, 1.05824, 1.360276, 2.98615, 2.328794)),\n",
       " ('CTTTTA', (1.080541, 1.117157, 1.023982, 1.332811, 2.859152, 2.534685)),\n",
       " ('CGTTTT', (0.925589, 1.03418, 0.884141, 1.202144, 2.717356, 1.980505)),\n",
       " ('ACATTT', (1.058285, 1.082465, 1.042439, 1.31747, 2.700537, 2.145311)),\n",
       " ('CCCTTT', (1.322174, 1.338083, 1.221096, 1.460462, 2.689696, 2.324784)),\n",
       " ('TCATTT', (0.995825, 1.028557, 0.997097, 1.264295, 2.642042, 1.991773)),\n",
       " ('CCATTT', (1.188851, 1.212594, 1.11519, 1.365551, 2.619526, 2.172033)),\n",
       " ('TTTACC', (1.224051, 1.260226, 1.149296, 1.382187, 2.61431, 2.323086)),\n",
       " ('CAATTT', (1.070547, 1.135088, 0.985793, 1.256936, 2.55934, 2.090553)),\n",
       " ('TTTACT', (1.075011, 1.092976, 1.044093, 1.263046, 2.526121, 2.049058)),\n",
       " ('CACTTT', (1.117398, 1.150381, 1.1061, 1.340924, 2.522245, 2.136232)),\n",
       " ('AAATTT', (1.092131, 1.210988, 0.96313, 1.277373, 2.507801, 2.156247)),\n",
       " ('ATTTTC', (1.088829, 1.197339, 0.962744, 1.251661, 2.506086, 2.238154)),\n",
       " ('CTCTTT', (1.106018, 1.127844, 1.142336, 1.342419, 2.492982, 2.067625)),\n",
       " ('TTTTGG', (0.937089, 1.054827, 0.907264, 1.17374, 2.492037, 1.922124)),\n",
       " ('CTTTTC', (1.164016, 1.219319, 1.10391, 1.334712, 2.488472, 2.278367)),\n",
       " ('ACCTTT', (1.240697, 1.227335, 1.233794, 1.426128, 2.465058, 2.121239)),\n",
       " ('TTTAAT', (1.06272, 1.108855, 1.023571, 1.283042, 2.456559, 2.045962)),\n",
       " ('ACACCC', (2.387527, 2.282171, 2.019541, 2.13895, 2.432033, 2.689552)),\n",
       " ('TTTCCC', (1.355923, 1.400906, 1.244972, 1.439235, 2.425335, 2.194756)),\n",
       " ('GGTTTT', (0.961226, 1.19214, 0.904943, 1.165857, 2.377518, 1.774106)),\n",
       " ('TTTCCT', (1.293798, 1.344679, 1.227248, 1.429927, 2.32793, 2.036784)),\n",
       " ('AACTTT', (1.091947, 1.112787, 1.064095, 1.286034, 2.299967, 1.925819)),\n",
       " ('TTTATA', (1.027563, 1.114956, 0.984499, 1.250078, 2.285906, 1.973772)),\n",
       " ('ATCTTT', (1.066621, 1.068272, 1.06643, 1.275994, 2.284345, 1.882184)),\n",
       " ('TTTCAT', (1.110277, 1.143161, 1.088105, 1.28131, 2.222439, 1.836058)),\n",
       " ('TTTCTA', (1.145047, 1.203835, 1.069236, 1.292814, 2.212533, 1.907289)),\n",
       " ('ATTTTG', (0.994758, 1.111871, 0.914621, 1.155213, 2.205814, 1.979945)),\n",
       " ('TTACTT', (1.082758, 1.076837, 1.07702, 1.249475, 2.197227, 1.821073)),\n",
       " ('GTATTT', (0.97927, 1.102852, 0.946439, 1.19871, 2.192105, 1.778512)),\n",
       " ('TTTTCG', (0.907062, 1.00087, 0.859899, 1.084858, 2.192062, 1.699221)),\n",
       " ('GCATTT', (1.016073, 1.057616, 0.996381, 1.183288, 2.180142, 1.765)),\n",
       " ('TCCTTT', (1.214518, 1.184983, 1.165382, 1.330348, 2.176655, 1.839746)),\n",
       " ('CCACCC', (1.974508, 2.016051, 1.908299, 1.982787, 2.168829, 2.385569)),\n",
       " ('TGCTTT', (1.003509, 1.028136, 1.045835, 1.202524, 2.147666, 1.75216)),\n",
       " ('TTAATT', (1.101491, 1.136987, 1.035496, 1.251181, 2.128661, 1.801665)),\n",
       " ('TTTCTC', (1.132957, 1.173515, 1.090708, 1.273628, 2.096104, 1.821551)),\n",
       " ('TTTACA', (1.069593, 1.091288, 1.023837, 1.203204, 2.089612, 1.865682)),\n",
       " ('CACCCC', (1.925714, 1.982213, 1.8017, 1.888431, 2.073407, 2.27656)),\n",
       " ('ATTATT', (1.064509, 1.112253, 1.004782, 1.250609, 2.067834, 1.824916)),\n",
       " ('TTTGCC', (1.152487, 1.17751, 1.156211, 1.302585, 2.040692, 1.88373)),\n",
       " ('TTTAAC', (1.062856, 1.086587, 1.006652, 1.208014, 2.03526, 1.81487)),\n",
       " ('CCCCCC', (1.67347, 1.822388, 1.64548, 1.76817, 2.034291, 2.072124)),\n",
       " ('TTTCAC', (1.180192, 1.218606, 1.146301, 1.291149, 2.026388, 1.890576)),\n",
       " ('TTTCCA', (1.300347, 1.345255, 1.233874, 1.386534, 2.015055, 1.897454)),\n",
       " ('CTTATT', (1.063676, 1.067599, 1.020397, 1.216595, 2.009885, 1.761087)),\n",
       " ('GCACCC', (1.921784, 1.971797, 1.802746, 1.898316, 2.008097, 2.261786)),\n",
       " ('TTTGCT', (1.016384, 1.025315, 1.039804, 1.165808, 2.005694, 1.67906)),\n",
       " ('CACCCT', (1.840973, 1.839547, 1.722294, 1.807765, 1.985276, 2.164302)),\n",
       " ('CACCCA', (1.950627, 1.883832, 1.813668, 1.849386, 1.978095, 2.227374)),\n",
       " ('GACACC', (2.041942, 1.907944, 1.828951, 1.827588, 1.905933, 2.024611)),\n",
       " ('ACACCT', (1.802165, 1.731956, 1.680782, 1.712761, 1.859659, 2.048757)),\n",
       " ('TCACCC', (1.795574, 1.753808, 1.67115, 1.711491, 1.847075, 2.008855)),\n",
       " ('CACCCG', (2.001298, 1.966189, 1.829228, 1.819881, 1.727738, 1.844854)),\n",
       " ('CCCGAG', (2.088873, 2.017272, 1.847135, 1.807011, 1.644216, 1.740085))]"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pos_six_mers"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "6894c643",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "<__array_function__ internals>:180: VisibleDeprecationWarning: Creating an ndarray from ragged nested sequences (which is a list-or-tuple of lists-or-tuples-or ndarrays with different lengths or shapes) is deprecated. If you meant to do this, you must specify 'dtype=object' when creating the ndarray.\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "(117, 2)"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "in_vitro_pos_seqs = np.concatenate((pos_four_mers,pos_five_mers,pos_six_mers))\n",
    "in_vitro_pos_seqs.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "3bc23f33",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "117"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "unique_rbns = {}\n",
    "for s, arr in in_vitro_pos_seqs:\n",
    "    if s not in unique_rbns:\n",
    "        unique_rbns[s] = arr\n",
    "\n",
    "in_vitro_pos_seqs = list(unique_rbns.items())\n",
    "len(in_vitro_pos_seqs)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6ad2eb50",
   "metadata": {},
   "source": [
    "Filter diff expr"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "7a6b7538",
   "metadata": {},
   "outputs": [],
   "source": [
    "# remove when gene expression is always 0 or if logfold change is inf/-inf\n",
    "diff_exp_filtered_HNRNPC = diff_exp_HNRNPC[\n",
    "    ~(\n",
    "        ((diff_exp_HNRNPC[\"value_1\"] == 0) & (diff_exp_HNRNPC[\"value_2\"] == 0)) |\n",
    "        (diff_exp_HNRNPC[\"log2(fold_change)\"].isin([np.inf, -np.inf]))\n",
    "    )\n",
    "]\n",
    "\n",
    "diff_exp_filtered_SRSF9 = diff_exp_SRSF9[\n",
    "    ~(\n",
    "        ((diff_exp_SRSF9[\"value_1\"] == 0) & (diff_exp_SRSF9[\"value_2\"] == 0)) |\n",
    "        (diff_exp_SRSF9[\"log2(fold_change)\"].isin([np.inf, -np.inf]))\n",
    "    )\n",
    "]\n",
    "\n",
    "diff_exp_filtered_PCBP1 = diff_exp_PCBP1[\n",
    "    ~(\n",
    "        ((diff_exp_PCBP1[\"value_1\"] == 0) & (diff_exp_PCBP1[\"value_2\"] == 0)) |\n",
    "        (diff_exp_PCBP1[\"log2(fold_change)\"].isin([np.inf, -np.inf]))\n",
    "    )\n",
    "]\n",
    "\n",
    "diff_exp_filtered_TIA1 = diff_exp_TIA1[\n",
    "    ~(\n",
    "        ((diff_exp_TIA1[\"value_1\"] == 0) & (diff_exp_TIA1[\"value_2\"] == 0)) |\n",
    "        (diff_exp_TIA1[\"log2(fold_change)\"].isin([np.inf, -np.inf]))\n",
    "    )\n",
    "]\n",
    "\n",
    "diff_exp_filtered_TRA2A = diff_exp_TRA2A[\n",
    "    ~(\n",
    "        ((diff_exp_TRA2A[\"value_1\"] == 0) & (diff_exp_TRA2A[\"value_2\"] == 0)) |\n",
    "        (diff_exp_TRA2A[\"log2(fold_change)\"].isin([np.inf, -np.inf]))\n",
    "    )\n",
    "]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "377252f2",
   "metadata": {},
   "outputs": [],
   "source": [
    "dfs = [diff_exp_filtered_HNRNPC, diff_exp_filtered_SRSF9, diff_exp_filtered_PCBP1, diff_exp_filtered_TIA1, diff_exp_filtered_TRA2A]\n",
    "col = \"gene_id\"\n",
    "shared_keys = set.intersection(*(set(df[col]) for df in dfs))\n",
    "\n",
    "# 2. Subset the dataframe you want (example: df1)\n",
    "diff_exp_filtered_HNRNPC = diff_exp_filtered_HNRNPC[diff_exp_filtered_HNRNPC[col].isin(shared_keys)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "fdeaa9e5",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>test_id</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene</th>\n",
       "      <th>locus</th>\n",
       "      <th>sample_1</th>\n",
       "      <th>sample_2</th>\n",
       "      <th>status</th>\n",
       "      <th>value_1</th>\n",
       "      <th>value_2</th>\n",
       "      <th>log2(fold_change)</th>\n",
       "      <th>test_stat</th>\n",
       "      <th>p_value</th>\n",
       "      <th>q_value</th>\n",
       "      <th>significant</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>XLOC_000008</td>\n",
       "      <td>XLOC_000008</td>\n",
       "      <td>OR4F29</td>\n",
       "      <td>chr1:317719-461954</td>\n",
       "      <td>HNRNPC-BGHLV20-HepG2</td>\n",
       "      <td>Control</td>\n",
       "      <td>OK</td>\n",
       "      <td>0.42339</td>\n",
       "      <td>1.53871</td>\n",
       "      <td>1.861660</td>\n",
       "      <td>0.889161</td>\n",
       "      <td>0.00805</td>\n",
       "      <td>0.035235</td>\n",
       "      <td>yes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>XLOC_000012</td>\n",
       "      <td>XLOC_000012</td>\n",
       "      <td>MTND1P23</td>\n",
       "      <td>chr1:536815-660283</td>\n",
       "      <td>HNRNPC-BGHLV20-HepG2</td>\n",
       "      <td>Control</td>\n",
       "      <td>OK</td>\n",
       "      <td>410.13200</td>\n",
       "      <td>336.58000</td>\n",
       "      <td>-0.285138</td>\n",
       "      <td>-0.149730</td>\n",
       "      <td>0.81500</td>\n",
       "      <td>0.897404</td>\n",
       "      <td>no</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>XLOC_000013</td>\n",
       "      <td>XLOC_000013</td>\n",
       "      <td>MTND2P28</td>\n",
       "      <td>chr1:536815-660283</td>\n",
       "      <td>HNRNPC-BGHLV20-HepG2</td>\n",
       "      <td>Control</td>\n",
       "      <td>OK</td>\n",
       "      <td>931.02300</td>\n",
       "      <td>834.29300</td>\n",
       "      <td>-0.158262</td>\n",
       "      <td>-0.292230</td>\n",
       "      <td>0.66870</td>\n",
       "      <td>0.802839</td>\n",
       "      <td>no</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>XLOC_000014</td>\n",
       "      <td>XLOC_000014</td>\n",
       "      <td>hsa-mir-6723</td>\n",
       "      <td>chr1:536815-660283</td>\n",
       "      <td>HNRNPC-BGHLV20-HepG2</td>\n",
       "      <td>Control</td>\n",
       "      <td>OK</td>\n",
       "      <td>1147.94000</td>\n",
       "      <td>1761.41000</td>\n",
       "      <td>0.617688</td>\n",
       "      <td>1.819070</td>\n",
       "      <td>0.00955</td>\n",
       "      <td>0.040072</td>\n",
       "      <td>yes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>XLOC_000015</td>\n",
       "      <td>XLOC_000015</td>\n",
       "      <td>RP5-857K21.7</td>\n",
       "      <td>chr1:536815-660283</td>\n",
       "      <td>HNRNPC-BGHLV20-HepG2</td>\n",
       "      <td>Control</td>\n",
       "      <td>OK</td>\n",
       "      <td>1200.71000</td>\n",
       "      <td>1284.82000</td>\n",
       "      <td>0.097686</td>\n",
       "      <td>0.164563</td>\n",
       "      <td>0.81710</td>\n",
       "      <td>0.898417</td>\n",
       "      <td>no</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        test_id      gene_id          gene               locus  \\\n",
       "7   XLOC_000008  XLOC_000008        OR4F29  chr1:317719-461954   \n",
       "11  XLOC_000012  XLOC_000012      MTND1P23  chr1:536815-660283   \n",
       "12  XLOC_000013  XLOC_000013      MTND2P28  chr1:536815-660283   \n",
       "13  XLOC_000014  XLOC_000014  hsa-mir-6723  chr1:536815-660283   \n",
       "14  XLOC_000015  XLOC_000015  RP5-857K21.7  chr1:536815-660283   \n",
       "\n",
       "                sample_1 sample_2 status     value_1     value_2  \\\n",
       "7   HNRNPC-BGHLV20-HepG2  Control     OK     0.42339     1.53871   \n",
       "11  HNRNPC-BGHLV20-HepG2  Control     OK   410.13200   336.58000   \n",
       "12  HNRNPC-BGHLV20-HepG2  Control     OK   931.02300   834.29300   \n",
       "13  HNRNPC-BGHLV20-HepG2  Control     OK  1147.94000  1761.41000   \n",
       "14  HNRNPC-BGHLV20-HepG2  Control     OK  1200.71000  1284.82000   \n",
       "\n",
       "    log2(fold_change)  test_stat  p_value   q_value significant  \n",
       "7            1.861660   0.889161  0.00805  0.035235         yes  \n",
       "11          -0.285138  -0.149730  0.81500  0.897404          no  \n",
       "12          -0.158262  -0.292230  0.66870  0.802839          no  \n",
       "13           0.617688   1.819070  0.00955  0.040072         yes  \n",
       "14           0.097686   0.164563  0.81710  0.898417          no  "
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "diff_exp_filtered_HNRNPC.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "1046db6c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(17985, 14)"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "diff_exp_filtered_HNRNPC.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "d8fe4752",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([ 1.86166 , -0.285138, -0.158262, ..., -1.65698 ,  0.128098,\n",
       "       -1.54899 ])"
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "logfold_changes = diff_exp_filtered_HNRNPC['log2(fold_change)'].values\n",
    "logfold_changes"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "14a14525",
   "metadata": {},
   "source": [
    "## Generate negative pairs"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "eaa775d1",
   "metadata": {},
   "source": [
    "Shuffle eCLIP sequences"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "68b0a794",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "647011"
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import random\n",
    "def shuffle_by_pairs(seq):\n",
    "    pairs = [seq[i:i+2] for i in range(0, len(seq), 2)]\n",
    "    random.shuffle(pairs)\n",
    "    return \"\".join(pairs)\n",
    "in_vivo_neg_seqs = [shuffle_by_pairs(seq) for seq in eclip[1].values]\n",
    "len(in_vivo_neg_seqs)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "1ec9c58c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "627721"
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# filter for length and unique values\n",
    "in_vivo_neg_seqs = [s for s in in_vivo_neg_seqs if len(s) >= 10]\n",
    "in_vivo_neg_seqs = list(set(in_vivo_neg_seqs))\n",
    "len(in_vivo_neg_seqs)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "02a5e383",
   "metadata": {},
   "source": [
    "Retrieve RBNS sequences that are not enriched"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "669ff6f6",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_insignficant_sequences(df):\n",
    "    num_cols = df.columns[1:]\n",
    "    filtered = df[df[num_cols].lt(2).any(axis=1)]\n",
    "    pairs = [\n",
    "        (row[df.columns[0]], tuple(row[num_cols]))\n",
    "        for _, row in filtered.iterrows()\n",
    "    ]\n",
    "    return pairs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "4a0c3f24",
   "metadata": {},
   "outputs": [],
   "source": [
    "neg_four_mers = get_insignficant_sequences(four_mers)\n",
    "neg_five_mers = get_insignficant_sequences(five_mers)\n",
    "neg_six_mers = get_insignficant_sequences(six_mers)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "7920a3d9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[('TTTT', (1.050108, 1.213564, 1.234794, 2.75543, 11.477239, 5.95855)),\n",
       " ('ATTT', (1.052096, 1.124278, 0.990108, 1.25527, 2.505993, 1.978384)),\n",
       " ('CTTT', (1.094396, 1.111101, 1.087047, 1.272162, 2.264619, 1.867528)),\n",
       " ('TTTA', (1.023649, 1.074103, 0.975594, 1.183402, 2.129294, 1.799562)),\n",
       " ('TTTC', (1.097361, 1.150883, 1.052075, 1.217492, 1.999443, 1.715131)),\n",
       " ('TATT', (1.039198, 1.06593, 1.000795, 1.150161, 1.713124, 1.496541)),\n",
       " ('TTTG', (1.004893, 1.050304, 0.987782, 1.093798, 1.596789, 1.390267)),\n",
       " ('GTTT', (0.973983, 1.046386, 0.956951, 1.079698, 1.564893, 1.351425)),\n",
       " ('CCCC', (1.419824, 1.45442, 1.343596, 1.41312, 1.549043, 1.601944)),\n",
       " ('TTAT', (1.011911, 1.034474, 0.975124, 1.09913, 1.524726, 1.373946)),\n",
       " ('ACCC', (1.48522, 1.480864, 1.389412, 1.432278, 1.504673, 1.613837)),\n",
       " ('TTCT', (1.115392, 1.128724, 1.087663, 1.167728, 1.499969, 1.362751)),\n",
       " ('CACC', (1.425643, 1.40017, 1.396304, 1.407528, 1.478731, 1.612886)),\n",
       " ('TCTT', (1.067138, 1.066816, 1.070433, 1.137513, 1.47739, 1.338711)),\n",
       " ('CCTT', (1.223428, 1.218957, 1.207205, 1.257616, 1.471449, 1.452822)),\n",
       " ('TTCC', (1.258922, 1.276476, 1.210496, 1.273769, 1.43923, 1.438196)),\n",
       " ('CCCT', (1.297303, 1.306835, 1.248285, 1.296923, 1.385793, 1.438717)),\n",
       " ('CATT', (1.059273, 1.045052, 1.052043, 1.103115, 1.354572, 1.276617)),\n",
       " ('GCCC', (1.267166, 1.286685, 1.284297, 1.290898, 1.344433, 1.4347)),\n",
       " ('AATT', (1.095289, 1.133786, 0.998282, 1.097207, 1.327526, 1.270973)),\n",
       " ('ACTT', (1.072442, 1.063085, 1.055059, 1.100418, 1.28865, 1.221723)),\n",
       " ('CCCA', (1.290003, 1.279884, 1.243526, 1.255731, 1.287662, 1.359885)),\n",
       " ('TTAC', (1.045513, 1.041314, 1.019919, 1.07411, 1.275841, 1.251561)),\n",
       " ('TACC', (1.210234, 1.197337, 1.179201, 1.20187, 1.257949, 1.327106)),\n",
       " ('CCCG', (1.357365, 1.363053, 1.287683, 1.297022, 1.243685, 1.305168)),\n",
       " ('CCAC', (1.190136, 1.18121, 1.206173, 1.199956, 1.223125, 1.308511)),\n",
       " ('ACCT', (1.209458, 1.196891, 1.193955, 1.207681, 1.217546, 1.279174)),\n",
       " ('TCCC', (1.204441, 1.185197, 1.11962, 1.16065, 1.209274, 1.234751)),\n",
       " ('TTCA', (1.055274, 1.054028, 1.045473, 1.079449, 1.203676, 1.187871)),\n",
       " ('CCAT', (1.143194, 1.120258, 1.136824, 1.142068, 1.178375, 1.22793)),\n",
       " ('TGCC', (1.105896, 1.109061, 1.149612, 1.142799, 1.170256, 1.238925)),\n",
       " ('TTAA', (1.014688, 1.024918, 0.95622, 1.031618, 1.166786, 1.16204)),\n",
       " ('TGTT', (0.965665, 0.986337, 0.976491, 1.020627, 1.166325, 1.088035)),\n",
       " ('TTGC', (1.01683, 1.014454, 1.045513, 1.050028, 1.160123, 1.150723)),\n",
       " ('ACAC', (1.208796, 1.166885, 1.183006, 1.161961, 1.144901, 1.222576)),\n",
       " ('AACC', (1.182685, 1.177179, 1.135118, 1.153375, 1.141967, 1.23317)),\n",
       " ('TCCT', (1.155229, 1.145474, 1.123786, 1.142958, 1.140403, 1.149314)),\n",
       " ('AGCC', (1.120876, 1.114207, 1.162778, 1.135222, 1.137418, 1.217739)),\n",
       " ('CACT', (1.113169, 1.090953, 1.129628, 1.120302, 1.136065, 1.1807)),\n",
       " ('GCTT', (0.979654, 0.970139, 1.014636, 1.021067, 1.135504, 1.100794)),\n",
       " ('CCTC', (1.108521, 1.114466, 1.122073, 1.129994, 1.120649, 1.172878)),\n",
       " ('TCCA', (1.144013, 1.132788, 1.136621, 1.135642, 1.117229, 1.175879)),\n",
       " ('CTCC', (1.148942, 1.12968, 1.104469, 1.119507, 1.11377, 1.140752)),\n",
       " ('CCTA', (1.166302, 1.163223, 1.130617, 1.145696, 1.112672, 1.169452)),\n",
       " ('TTGT', (0.974288, 0.984751, 0.978307, 1.005291, 1.11007, 1.037032)),\n",
       " ('ATTC', (1.157592, 1.162652, 1.105276, 1.129247, 1.108435, 1.144567)),\n",
       " ('ATCC', (1.207557, 1.194663, 1.168822, 1.168604, 1.104295, 1.171949)),\n",
       " ('CTCT', (1.074146, 1.071938, 1.096599, 1.099103, 1.097939, 1.122067)),\n",
       " ('CCGC', (1.064809, 1.087773, 1.11388, 1.103482, 1.091869, 1.143869)),\n",
       " ('ACCA', (1.127295, 1.102039, 1.124034, 1.109014, 1.09056, 1.172363)),\n",
       " ('CTAT', (1.046023, 1.036685, 1.031032, 1.056776, 1.088781, 1.099002)),\n",
       " ('GCCT', (1.046494, 1.070228, 1.106249, 1.092944, 1.087513, 1.1282)),\n",
       " ('TACT', (1.035715, 1.019237, 1.023107, 1.040522, 1.084153, 1.073811)),\n",
       " ('CGCC', (1.063165, 1.087612, 1.12552, 1.104155, 1.08312, 1.131778)),\n",
       " ('CTTC', (1.082049, 1.070985, 1.095449, 1.094022, 1.068158, 1.099821)),\n",
       " ('TTGA', (1.111597, 1.1122, 1.096822, 1.091737, 1.063546, 1.065769)),\n",
       " ('CCAG', (1.109368, 1.105508, 1.132628, 1.103283, 1.059743, 1.1278)),\n",
       " ('CCAA', (1.139536, 1.125944, 1.110603, 1.10289, 1.058203, 1.125772)),\n",
       " ('CCTG', (1.154119, 1.162398, 1.146092, 1.139134, 1.057779, 1.107168)),\n",
       " ('GCCA', (1.025333, 1.0196, 1.079771, 1.051842, 1.053953, 1.113197)),\n",
       " ('CTTA', (1.047841, 1.035736, 1.021286, 1.048683, 1.050647, 1.093398)),\n",
       " ('TAAT', (1.020177, 1.021903, 0.966731, 1.018445, 1.044392, 1.058038)),\n",
       " ('ATAT', (1.01467, 1.021111, 0.962297, 1.017869, 1.037211, 1.057666)),\n",
       " ('CAGC', (0.992877, 0.982756, 1.051294, 1.018022, 1.037053, 1.107017)),\n",
       " ('AGTT', (1.001615, 1.028361, 1.005502, 1.022721, 1.03588, 1.02796)),\n",
       " ('CACA', (1.068383, 1.039196, 1.07111, 1.051283, 1.032812, 1.0955)),\n",
       " ('CTTG', (1.134874, 1.132656, 1.157003, 1.122504, 1.030871, 1.08153)),\n",
       " ('ACAT', (1.035626, 1.007352, 1.020853, 1.02547, 1.030717, 1.056465)),\n",
       " ('CTAC', (1.069298, 1.052677, 1.072502, 1.060657, 1.028359, 1.078324)),\n",
       " ('GCAC', (1.049266, 1.038298, 1.085539, 1.056979, 1.025212, 1.091483)),\n",
       " ('TCAC', (1.034928, 1.017928, 1.051188, 1.036791, 1.014597, 1.065203)),\n",
       " ('CATC', (1.079669, 1.050094, 1.098459, 1.07099, 1.010368, 1.071309)),\n",
       " ('TCAT', (0.979843, 0.960044, 0.987098, 0.990595, 1.008102, 0.998219)),\n",
       " ('ACTC', (1.104655, 1.082017, 1.100156, 1.081797, 1.007197, 1.045725)),\n",
       " ('TATA', (1.008179, 1.019603, 0.96064, 1.018709, 1.005997, 1.047081)),\n",
       " ('TACA', (1.027512, 1.004611, 1.004238, 1.012611, 1.005249, 1.049728)),\n",
       " ('ATAC', (1.090318, 1.070538, 1.047526, 1.052456, 1.005021, 1.057875)),\n",
       " ('CCGA', (1.192177, 1.188492, 1.184839, 1.147941, 1.003003, 1.045055)),\n",
       " ('TGCT', (0.961991, 0.948393, 1.001237, 0.987444, 0.999196, 1.012194)),\n",
       " ('CAAT', (1.026397, 1.000356, 1.009331, 1.009914, 0.998847, 1.026424)),\n",
       " ('TCTC', (1.038349, 1.028823, 1.043395, 1.037923, 0.995729, 1.011432)),\n",
       " ('ATTA', (1.01825, 1.022742, 0.962543, 1.008196, 0.994588, 1.033436)),\n",
       " ('CTGC', (0.989938, 0.994956, 1.032927, 1.015658, 0.987842, 1.031733)),\n",
       " ('TCCG', (1.078999, 1.097424, 1.088433, 1.075787, 0.986141, 1.005762)),\n",
       " ('TAAC', (1.025791, 1.009699, 0.999649, 1.009898, 0.984679, 1.033093)),\n",
       " ('TTAG', (0.951621, 0.967427, 0.938225, 0.959081, 0.981914, 0.972419)),\n",
       " ('CTCA', (1.025935, 1.012608, 1.048235, 1.024879, 0.979828, 1.029937)),\n",
       " ('CAAC', (1.042279, 1.016183, 1.052903, 1.026252, 0.978039, 1.031157)),\n",
       " ('GCAT', (0.974611, 0.956565, 0.994824, 0.973668, 0.977567, 0.993519)),\n",
       " ('TCTA', (1.028603, 1.026204, 1.013511, 1.021839, 0.97054, 0.994116)),\n",
       " ('CATA', (1.016812, 0.99549, 0.985774, 0.993595, 0.967526, 1.015561)),\n",
       " ('TGCA', (0.979029, 0.969237, 0.998097, 0.981104, 0.964104, 1.006433)),\n",
       " ('AAAT', (1.012026, 1.055945, 0.885901, 0.954504, 0.964048, 1.0091)),\n",
       " ('GAGA', (1.189721, 1.173752, 1.205329, 1.141142, 0.962055, 0.99445)),\n",
       " ('CGCT', (0.936172, 0.940451, 0.996647, 0.972095, 0.955043, 0.983616)),\n",
       " ('TAGC', (0.968338, 0.955723, 1.002869, 0.973544, 0.949281, 1.001516)),\n",
       " ('ATGC', (1.001118, 0.98225, 1.019043, 0.996135, 0.948524, 1.014017)),\n",
       " ('GATT', (1.021351, 1.027089, 0.989115, 0.988988, 0.948466, 0.915075)),\n",
       " ('AACT', (1.012219, 1.004958, 0.978123, 0.988447, 0.947388, 0.965324)),\n",
       " ('TATC', (1.009953, 1.001967, 0.995658, 1.005619, 0.944625, 0.978343)),\n",
       " ('AGCA', (0.988322, 0.959359, 1.009964, 0.974151, 0.942942, 1.001039)),\n",
       " ('ATCT', (1.013181, 1.005519, 1.004818, 1.006746, 0.941451, 0.960804)),\n",
       " ('CGAG', (1.122606, 1.109578, 1.160536, 1.089322, 0.934257, 0.961991)),\n",
       " ('ACGC', (0.9645, 0.953817, 0.999574, 0.965999, 0.933208, 0.979767)),\n",
       " ('GAAT', (1.107281, 1.110901, 1.064613, 1.061805, 0.932724, 0.959449)),\n",
       " ('TAAA', (0.994246, 1.018058, 0.894872, 0.954532, 0.931015, 0.989319)),\n",
       " ('AATA', (1.027976, 1.034914, 0.94675, 0.986928, 0.930772, 0.995975)),\n",
       " ('AAAC', (1.030843, 1.044137, 0.940185, 0.972222, 0.929656, 0.998898)),\n",
       " ('GCGC', (0.879827, 0.891527, 0.974626, 0.929399, 0.929005, 0.972594)),\n",
       " ('CTAA', (1.031123, 1.025214, 0.98729, 0.994702, 0.928537, 0.979195)),\n",
       " ('GACA', (1.073567, 1.038709, 1.073477, 1.025916, 0.926852, 0.940955)),\n",
       " ('GGCC', (0.92651, 0.947857, 0.983426, 0.959889, 0.924447, 0.946828)),\n",
       " ('AGCT', (0.964786, 0.943186, 1.005371, 0.964593, 0.923824, 0.955799)),\n",
       " ('GACC', (1.032679, 1.027745, 1.057104, 1.022253, 0.922909, 0.955618)),\n",
       " ('CAGT', (0.959367, 0.951971, 0.995887, 0.967115, 0.919368, 0.955543)),\n",
       " ('TGAC', (1.059146, 1.03667, 1.076065, 1.030622, 0.917404, 0.94127)),\n",
       " ('CGCA', (0.939455, 0.932203, 0.977503, 0.945506, 0.916803, 0.960072)),\n",
       " ('ACTA', (1.005568, 0.985499, 0.983923, 0.981526, 0.916331, 0.95563)),\n",
       " ('GCTC', (0.958081, 0.954574, 1.00957, 0.972612, 0.91446, 0.943508)),\n",
       " ('AACA', (1.012846, 0.992017, 0.977567, 0.97418, 0.912258, 0.974464)),\n",
       " ('CCGT', (0.996878, 1.008147, 1.014106, 1.004257, 0.911436, 0.940281)),\n",
       " ('AAGC', (0.983695, 0.963021, 0.994414, 0.962796, 0.910511, 0.970855)),\n",
       " ('TGAG', (1.040285, 1.036538, 1.070812, 1.026752, 0.906922, 0.940146)),\n",
       " ('GTCC', (0.982279, 0.982682, 1.017505, 0.994736, 0.905886, 0.937037)),\n",
       " ('AGAA', (1.107299, 1.115278, 1.058872, 1.052557, 0.905256, 0.965466)),\n",
       " ('CAAA', (1.002968, 1.006737, 0.931463, 0.948457, 0.904108, 0.956021)),\n",
       " ('GCTA', (0.934516, 0.925263, 0.963565, 0.93712, 0.898954, 0.935439)),\n",
       " ('ATAA', (1.011087, 1.01005, 0.941562, 0.975473, 0.898754, 0.962043)),\n",
       " ('AAAA', (0.984943, 1.057085, 0.813962, 0.893712, 0.898639, 0.958259)),\n",
       " ('GAGC', (0.990403, 0.972898, 1.045256, 0.984548, 0.898332, 0.928608)),\n",
       " ('CCGG', (1.016151, 1.047111, 1.023705, 1.001449, 0.897393, 0.900657)),\n",
       " ('GTTC', (0.988996, 0.988526, 1.011861, 0.996383, 0.897029, 0.924378)),\n",
       " ('GTAT', (0.940661, 0.954345, 0.942212, 0.96079, 0.894814, 0.910711)),\n",
       " ('ACCG', (1.00186, 1.00694, 1.027913, 0.99936, 0.894409, 0.937697)),\n",
       " ('AGAG', (1.060566, 1.042821, 1.086206, 1.032529, 0.89437, 0.933877)),\n",
       " ('GGTT', (0.918632, 0.957237, 0.935975, 0.932902, 0.893798, 0.882657)),\n",
       " ('TTGG', (0.934826, 0.957479, 0.94418, 0.934235, 0.892486, 0.874403)),\n",
       " ('ACAA', (1.000924, 0.972322, 0.971183, 0.960256, 0.889475, 0.934811)),\n",
       " ('TGGC', (0.928486, 0.933453, 0.971492, 0.943975, 0.889157, 0.913915)),\n",
       " ('GGCA', (0.933725, 0.934506, 0.958764, 0.927794, 0.885706, 0.891014)),\n",
       " ('GAGT', (1.032969, 1.035107, 1.073527, 1.027081, 0.885292, 0.913323)),\n",
       " ('TCAA', (0.976194, 0.96218, 0.957411, 0.94619, 0.884585, 0.907178)),\n",
       " ('CGTT', (0.903484, 0.905783, 0.917726, 0.912069, 0.884472, 0.870356)),\n",
       " ('GCAA', (0.934169, 0.922806, 0.942978, 0.919749, 0.884029, 0.918273)),\n",
       " ('AATC', (1.040402, 1.037417, 0.985231, 0.99174, 0.883451, 0.929193)),\n",
       " ('CAGA', (1.042016, 1.030781, 1.047248, 1.008475, 0.88247, 0.921758)),\n",
       " ('CATG', (0.961533, 0.942247, 0.992536, 0.957524, 0.882282, 0.934371)),\n",
       " ('CTGA', (1.056814, 1.057607, 1.068676, 1.033971, 0.882141, 0.918159)),\n",
       " ('TATG', (0.953836, 0.950967, 0.94374, 0.94886, 0.879281, 0.908002)),\n",
       " ('TAGT', (0.942202, 0.950795, 0.949456, 0.94882, 0.877141, 0.889897)),\n",
       " ('CTAG', (1.00587, 1.000354, 1.023722, 0.990806, 0.877037, 0.920339)),\n",
       " ('TGAA', (1.027779, 1.03721, 1.011756, 1.002879, 0.875909, 0.921807)),\n",
       " ('GTGC', (0.889761, 0.891233, 0.956148, 0.921144, 0.874759, 0.922458)),\n",
       " ('GTTA', (0.937064, 0.952267, 0.933319, 0.944103, 0.873677, 0.901702)),\n",
       " ('CTGT', (0.953407, 0.953685, 0.968154, 0.955037, 0.870334, 0.880231)),\n",
       " ('CACG', (0.943842, 0.936722, 0.981334, 0.942427, 0.86984, 0.917154)),\n",
       " ('GGGG', (0.880416, 0.91135, 0.91534, 0.897819, 0.869444, 0.884753)),\n",
       " ('GGGA', (0.972424, 0.987712, 0.988174, 0.959354, 0.86824, 0.880547)),\n",
       " ('GCAG', (0.896421, 0.898455, 0.955377, 0.91394, 0.866934, 0.897666)),\n",
       " ('TGAT', (1.007141, 0.995084, 1.020845, 0.99019, 0.863727, 0.875367)),\n",
       " ('GAAA', (1.018535, 1.053006, 0.929396, 0.949545, 0.860486, 0.914878)),\n",
       " ('TCTG', (0.966941, 0.978207, 0.987294, 0.969795, 0.859892, 0.872628)),\n",
       " ('TTCG', (0.917363, 0.930885, 0.925342, 0.916228, 0.859194, 0.844192)),\n",
       " ('GTAC', (0.953479, 0.944615, 0.965826, 0.948586, 0.85835, 0.8976)),\n",
       " ('TGTA', (0.94925, 0.960079, 0.942979, 0.954397, 0.858216, 0.886339)),\n",
       " ('AGTA', (0.978887, 0.977341, 0.972256, 0.967131, 0.855891, 0.910043)),\n",
       " ('CTGG', (0.94751, 0.962459, 0.97979, 0.95215, 0.854334, 0.870595)),\n",
       " ('GATA', (1.02937, 1.021333, 1.018439, 0.99362, 0.852852, 0.88557)),\n",
       " ('ATGT', (0.93145, 0.928103, 0.937778, 0.938033, 0.851986, 0.878922)),\n",
       " ('TAGA', (1.01652, 1.016643, 1.002225, 0.98685, 0.84983, 0.882108)),\n",
       " ('AGAC', (1.024611, 0.996732, 1.027001, 0.977225, 0.849608, 0.865322)),\n",
       " ('CAAG', (0.973456, 0.952272, 0.993755, 0.947563, 0.847562, 0.890351)),\n",
       " ('ACAG', (0.955174, 0.939093, 0.970983, 0.93471, 0.845312, 0.883092)),\n",
       " ('GCCG', (0.88259, 0.90564, 0.965698, 0.927481, 0.842242, 0.850272)),\n",
       " ('GACT', (0.96969, 0.948313, 0.977445, 0.944041, 0.841155, 0.814819)),\n",
       " ('ATCA', (0.962131, 0.940552, 0.956792, 0.938927, 0.838061, 0.874873)),\n",
       " ('TAAG', (0.950827, 0.944538, 0.934521, 0.928496, 0.837329, 0.87442)),\n",
       " ('AGGC', (0.895752, 0.890376, 0.92422, 0.889122, 0.837041, 0.858391)),\n",
       " ('ATTG', (0.978336, 0.966229, 0.965804, 0.950129, 0.836832, 0.858508)),\n",
       " ('ATGA', (1.020403, 0.996461, 1.022293, 0.988827, 0.835811, 0.87842)),\n",
       " ('GGCT', (0.881231, 0.879786, 0.903216, 0.874534, 0.832709, 0.809339)),\n",
       " ('TGCG', (0.868044, 0.867863, 0.918055, 0.88394, 0.831829, 0.856435)),\n",
       " ('GTGT', (0.876736, 0.898941, 0.924135, 0.903868, 0.831756, 0.857951)),\n",
       " ('TGGG', (0.897055, 0.917415, 0.915085, 0.8954, 0.831588, 0.835259)),\n",
       " ('GTCT', (0.919587, 0.913816, 0.950098, 0.923144, 0.830509, 0.83049)),\n",
       " ('GAAC', (0.976489, 0.973011, 0.983178, 0.96098, 0.829648, 0.864946)),\n",
       " ('TACG', (0.908661, 0.901492, 0.914113, 0.893209, 0.829027, 0.840843)),\n",
       " ('AGAT', (1.024915, 1.012358, 1.020034, 0.988056, 0.82841, 0.852254)),\n",
       " ('TCAG', (0.926046, 0.921128, 0.948941, 0.917223, 0.82784, 0.852067)),\n",
       " ('GAGG', (0.931873, 0.941049, 0.967683, 0.924794, 0.82726, 0.841136)),\n",
       " ('CGAC', (0.976534, 0.964376, 1.008891, 0.953014, 0.8267, 0.834691)),\n",
       " ('AAGT', (0.942741, 0.950052, 0.931117, 0.925617, 0.82648, 0.864572)),\n",
       " ('GTAA', (0.936143, 0.94646, 0.913002, 0.922328, 0.826388, 0.86968)),\n",
       " ('ACTG', (0.943581, 0.933107, 0.958142, 0.930688, 0.826301, 0.840505)),\n",
       " ('TGTG', (0.895663, 0.901448, 0.933655, 0.906443, 0.82587, 0.849546)),\n",
       " ('GGGT', (0.89024, 0.923575, 0.914958, 0.898218, 0.825847, 0.837047)),\n",
       " ('CAGG', (0.894975, 0.898791, 0.926408, 0.894285, 0.825752, 0.855132)),\n",
       " ('GCTG', (0.876638, 0.878766, 0.938967, 0.894996, 0.823943, 0.835822)),\n",
       " ('AATG', (0.966047, 0.957514, 0.940864, 0.932223, 0.822909, 0.868559)),\n",
       " ('CTCG', (0.974714, 0.984824, 1.02258, 0.9763, 0.82203, 0.843974)),\n",
       " ('AAGA', (1.036342, 1.018162, 1.010581, 0.979442, 0.821459, 0.864259)),\n",
       " ('CGAA', (0.993351, 0.99305, 0.982331, 0.954032, 0.820021, 0.857659)),\n",
       " ('ATAG', (0.955812, 0.951138, 0.954883, 0.935137, 0.81992, 0.85979)),\n",
       " ('CGGG', (0.934765, 0.95929, 0.949559, 0.922098, 0.819686, 0.812488)),\n",
       " ('GGAA', (0.958915, 0.971359, 0.936426, 0.919787, 0.817301, 0.831306)),\n",
       " ('TGGT', (0.888175, 0.90691, 0.931717, 0.906158, 0.816943, 0.833334)),\n",
       " ('CGGC', (0.861225, 0.879626, 0.913433, 0.87127, 0.815174, 0.80362)),\n",
       " ('GAAG', (0.968715, 0.964808, 0.991836, 0.944281, 0.812877, 0.843112)),\n",
       " ('TGTC', (0.934184, 0.92165, 0.960281, 0.928225, 0.812056, 0.834717)),\n",
       " ('AGTG', (0.90312, 0.909336, 0.95048, 0.909522, 0.809704, 0.852359)),\n",
       " ('AGGG', (0.87155, 0.888459, 0.888353, 0.867234, 0.808662, 0.827776)),\n",
       " ('CGAT', (0.962159, 0.949585, 0.989474, 0.942955, 0.806985, 0.822217)),\n",
       " ('CGCG', (0.856248, 0.872076, 0.929204, 0.881681, 0.806496, 0.836921)),\n",
       " ('GGTA', (0.895293, 0.91864, 0.910047, 0.896692, 0.805977, 0.835333)),\n",
       " ('TCGC', (0.875566, 0.886217, 0.930616, 0.896901, 0.804493, 0.820223)),\n",
       " ('GGGC', (0.844511, 0.853628, 0.871961, 0.845194, 0.803781, 0.778689)),\n",
       " ('GCGA', (0.895454, 0.893171, 0.953933, 0.898992, 0.800479, 0.820317)),\n",
       " ('AGCG', (0.858214, 0.850698, 0.912411, 0.861526, 0.799939, 0.835404)),\n",
       " ('GGAG', (0.909171, 0.92477, 0.94926, 0.906299, 0.797591, 0.80409)),\n",
       " ('AGTC', (0.953812, 0.93493, 0.971322, 0.931454, 0.797149, 0.819132)),\n",
       " ('AAAG', (0.950953, 0.956631, 0.895675, 0.898535, 0.792975, 0.841761)),\n",
       " ('GTGA', (0.928235, 0.930255, 0.975923, 0.926216, 0.792814, 0.824437)),\n",
       " ('GATG', (0.945062, 0.930501, 0.987624, 0.939211, 0.789376, 0.813587)),\n",
       " ('ACGA', (0.969887, 0.945599, 0.979188, 0.927458, 0.787162, 0.809851)),\n",
       " ('AGGT', (0.884018, 0.898229, 0.918267, 0.889823, 0.787072, 0.814165)),\n",
       " ('GGTG', (0.837552, 0.862057, 0.907047, 0.869256, 0.782553, 0.806807)),\n",
       " ('ATGG', (0.887627, 0.88476, 0.905478, 0.877065, 0.780076, 0.801549)),\n",
       " ('TAGG', (0.880431, 0.89335, 0.884686, 0.869574, 0.779552, 0.788457)),\n",
       " ('GCGT', (0.821426, 0.821948, 0.874339, 0.835369, 0.776887, 0.797741)),\n",
       " ('GTCA', (0.88758, 0.871012, 0.908677, 0.871253, 0.775644, 0.783138)),\n",
       " ('CGTA', (0.892773, 0.891044, 0.900192, 0.888345, 0.775564, 0.809623)),\n",
       " ('TCGA', (0.955351, 0.953888, 0.97501, 0.932088, 0.774104, 0.784404)),\n",
       " ('CGGT', (0.852555, 0.872742, 0.891982, 0.862018, 0.773064, 0.776421)),\n",
       " ('TGGA', (0.908613, 0.914038, 0.919718, 0.888734, 0.773035, 0.764296)),\n",
       " ('GTTG', (0.882991, 0.889494, 0.917694, 0.883311, 0.769349, 0.777367)),\n",
       " ('GGAC', (0.856849, 0.848502, 0.864726, 0.844319, 0.769167, 0.732115)),\n",
       " ('GTAG', (0.88697, 0.900635, 0.90408, 0.884957, 0.768779, 0.79393)),\n",
       " ('CGTG', (0.870193, 0.8764, 0.927925, 0.880588, 0.768401, 0.806791)),\n",
       " ('GTGG', (0.829766, 0.849445, 0.886653, 0.848969, 0.767731, 0.782529)),\n",
       " ('GGAT', (0.917706, 0.922679, 0.920589, 0.891905, 0.764151, 0.747988)),\n",
       " ('AGGA', (0.914914, 0.908836, 0.913552, 0.882754, 0.761694, 0.764249)),\n",
       " ('ACGT', (0.858806, 0.850958, 0.876, 0.853158, 0.758055, 0.778538)),\n",
       " ('GCGG', (0.795141, 0.814458, 0.858279, 0.818072, 0.757455, 0.753177)),\n",
       " ('GGCG', (0.799445, 0.809731, 0.868317, 0.817205, 0.754536, 0.752886)),\n",
       " ('AAGG', (0.872069, 0.873518, 0.871329, 0.846499, 0.749246, 0.772297)),\n",
       " ('AACG', (0.888444, 0.88379, 0.886085, 0.861099, 0.748755, 0.781717)),\n",
       " ('ACGG', (0.843131, 0.850978, 0.862152, 0.830013, 0.737784, 0.735588)),\n",
       " ('GACG', (0.884408, 0.872712, 0.914321, 0.85928, 0.736082, 0.722351)),\n",
       " ('CGGA', (0.874923, 0.8837, 0.875329, 0.847998, 0.730706, 0.711075)),\n",
       " ('GGTC', (0.848245, 0.852181, 0.892421, 0.849731, 0.72725, 0.727803)),\n",
       " ('GATC', (0.908302, 0.899281, 0.941015, 0.891437, 0.719511, 0.723544)),\n",
       " ('TCGG', (0.830022, 0.846412, 0.842917, 0.818757, 0.693297, 0.67211)),\n",
       " ('TCGT', (0.849285, 0.84463, 0.870984, 0.837076, 0.685304, 0.681841)),\n",
       " ('CGTC', (0.845628, 0.842135, 0.878028, 0.833263, 0.67824, 0.691214)),\n",
       " ('ATCG', (0.849546, 0.843249, 0.866847, 0.828345, 0.66632, 0.674459)),\n",
       " ('GTCG', (0.782238, 0.781878, 0.814882, 0.774402, 0.633157, 0.619229))]"
      ]
     },
     "execution_count": 34,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "neg_four_mers"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "47bc768a",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "<__array_function__ internals>:180: VisibleDeprecationWarning: Creating an ndarray from ragged nested sequences (which is a list-or-tuple of lists-or-tuples-or ndarrays with different lengths or shapes) is deprecated. If you meant to do this, you must specify 'dtype=object' when creating the ndarray.\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "(5375, 2)"
      ]
     },
     "execution_count": 35,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "in_vitro_neg_seqs = np.concatenate((neg_four_mers,neg_five_mers,neg_six_mers))\n",
    "in_vitro_neg_seqs.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "88397cff",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "5375"
      ]
     },
     "execution_count": 36,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "unique_rbns = {}\n",
    "for s, arr in in_vitro_neg_seqs:\n",
    "    if s not in unique_rbns:\n",
    "        unique_rbns[s] = arr\n",
    "\n",
    "in_vitro_neg_seqs = list(unique_rbns.items())\n",
    "len(in_vitro_neg_seqs)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "94988d48",
   "metadata": {},
   "source": [
    "Construct placeholder expression change vector"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "950281ea",
   "metadata": {},
   "outputs": [],
   "source": [
    "placeholder_logfold_changes = np.zeros(shape=logfold_changes.shape)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b14cf59a",
   "metadata": {},
   "source": [
    "## Split into train/val/test"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "id": "e9d8ef74",
   "metadata": {},
   "outputs": [],
   "source": [
    "# number of total samples per class\n",
    "combinations_size = 100000"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a99c5d06",
   "metadata": {},
   "source": [
    "Positive samples"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "id": "c5fdbf5e",
   "metadata": {},
   "outputs": [],
   "source": [
    "from itertools import product\n",
    "\n",
    "random.shuffle(in_vivo_pos_seqs)\n",
    "random.shuffle(in_vitro_pos_seqs)\n",
    "\n",
    "n_eclip = len(in_vivo_pos_seqs)\n",
    "n_rbns  = len(in_vitro_pos_seqs)\n",
    "\n",
    "# Split indices\n",
    "train_eclip = in_vivo_pos_seqs[:int(0.7*n_eclip)]\n",
    "val_eclip   = in_vivo_pos_seqs[int(0.7*n_eclip):int(0.85*n_eclip)]\n",
    "test_eclip  = in_vivo_pos_seqs[int(0.85*n_eclip):]\n",
    "\n",
    "train_rbns = in_vitro_pos_seqs[:int(0.7*n_rbns)]\n",
    "val_rbns   = in_vitro_pos_seqs[int(0.7*n_rbns):int(0.85*n_rbns)]\n",
    "test_rbns  = in_vitro_pos_seqs[int(0.85*n_rbns):]\n",
    "\n",
    "pos_train_samples = [\n",
    "    (random.choice(train_eclip), random.choice(train_rbns))\n",
    "    for _ in range(combinations_size)\n",
    "]\n",
    "pos_val_samples = [\n",
    "    (random.choice(val_eclip), random.choice(val_rbns))\n",
    "    for _ in range(combinations_size)\n",
    "]\n",
    "pos_test_samples = [\n",
    "    (random.choice(test_eclip), random.choice(test_rbns))\n",
    "    for _ in range(combinations_size)\n",
    "]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "91baf0a1",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "eCLIP sequences are disjoint across splits\n"
     ]
    }
   ],
   "source": [
    "train_eclips = set(eclip for eclip, _ in pos_train_samples)\n",
    "val_eclips   = set(eclip for eclip, _ in pos_val_samples)\n",
    "test_eclips  = set(eclip for eclip, _ in pos_test_samples)\n",
    "\n",
    "assert train_eclips.isdisjoint(val_eclips), \"Overlap between train and val eCLIP sequences!\"\n",
    "assert train_eclips.isdisjoint(test_eclips), \"Overlap between train and test eCLIP sequences!\"\n",
    "assert val_eclips.isdisjoint(test_eclips), \"Overlap between val and test eCLIP sequences!\"\n",
    "print(\"eCLIP sequences are disjoint across splits\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "id": "0ffaaf81",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "RBNS sequences are disjoint across splits\n"
     ]
    }
   ],
   "source": [
    "train_rbns = set(rbns for _, (rbns, _) in pos_train_samples)\n",
    "val_rbns   = set(rbns for _, (rbns, _) in pos_val_samples)\n",
    "test_rbns  = set(rbns for _, (rbns, _) in pos_test_samples)\n",
    "\n",
    "assert train_rbns.isdisjoint(val_rbns), \"Overlap between train and val RBNS sequences!\"\n",
    "assert train_rbns.isdisjoint(test_rbns), \"Overlap between train and test RBNS sequences!\"\n",
    "assert val_rbns.isdisjoint(test_rbns), \"Overlap between val and test RBNS sequences!\"\n",
    "print(\"RBNS sequences are disjoint across splits\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "id": "330d4bc3",
   "metadata": {},
   "outputs": [],
   "source": [
    "pos_train_label = np.ones(len(pos_train_samples))\n",
    "pos_val_label = np.ones(len(pos_val_samples))\n",
    "pos_test_label = np.ones(len(pos_test_samples))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c3da2c3a",
   "metadata": {},
   "source": [
    "Negative samples"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "id": "d7c965b7",
   "metadata": {},
   "outputs": [],
   "source": [
    "random.shuffle(in_vivo_neg_seqs)\n",
    "random.shuffle(in_vitro_neg_seqs)\n",
    "\n",
    "n_eclip = len(in_vivo_neg_seqs)\n",
    "n_rbns  = len(in_vitro_neg_seqs)\n",
    "\n",
    "# Split indices\n",
    "train_eclip = in_vivo_neg_seqs[:int(0.7*n_eclip)]\n",
    "val_eclip   = in_vivo_neg_seqs[int(0.7*n_eclip):int(0.85*n_eclip)]\n",
    "test_eclip  = in_vivo_neg_seqs[int(0.85*n_eclip):]\n",
    "\n",
    "train_rbns = in_vitro_neg_seqs[:int(0.7*n_rbns)]\n",
    "val_rbns   = in_vitro_neg_seqs[int(0.7*n_rbns):int(0.85*n_rbns)]\n",
    "test_rbns  = in_vitro_neg_seqs[int(0.85*n_rbns):]\n",
    "\n",
    "neg_train_samples = [\n",
    "    (random.choice(train_eclip), random.choice(train_rbns))\n",
    "    for _ in range(combinations_size)\n",
    "]\n",
    "neg_val_samples = [\n",
    "    (random.choice(val_eclip), random.choice(val_rbns))\n",
    "    for _ in range(combinations_size)\n",
    "]\n",
    "neg_test_samples = [\n",
    "    (random.choice(test_eclip), random.choice(test_rbns))\n",
    "    for _ in range(combinations_size)\n",
    "]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "id": "19d1f6e3",
   "metadata": {},
   "outputs": [],
   "source": [
    "neg_train_label = np.zeros(len(neg_train_samples))\n",
    "neg_val_label = np.zeros(len(neg_val_samples))\n",
    "neg_test_label = np.zeros(len(neg_test_samples))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9cc54a9e",
   "metadata": {},
   "source": [
    "Combine for full splits"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "id": "726d4838",
   "metadata": {},
   "outputs": [],
   "source": [
    "train_data = np.concatenate((pos_train_samples, neg_train_samples))\n",
    "val_data = np.concatenate((pos_val_samples, neg_val_samples))\n",
    "test_data = np.concatenate((pos_test_samples, neg_test_samples))\n",
    "\n",
    "train_labels = np.concatenate((pos_train_label, neg_train_label))\n",
    "val_labels = np.concatenate((pos_val_label, neg_val_label))\n",
    "test_labels = np.concatenate((pos_test_label, neg_test_label))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "id": "8ce65d9c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# get positive/negative label values\n",
    "positive_value = (\"HNRNPC\", logfold_changes)\n",
    "negative_value = (\"Not HNRNPC\", placeholder_logfold_changes)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "877f0473",
   "metadata": {},
   "source": [
    "## Save as dictionaries"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "16dcdac6",
   "metadata": {},
   "source": [
    "Save as pickle files since there are nested data structures"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "id": "eff9d82c",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pickle"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "id": "cfda55f9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'\\nSaved data format for each split:\\n- Input\\n    - List containing:\\n        - eCLIP sequence: text\\n        - Tuple of RBNS sequence and binding affinity: (text, vector)\\n- Output\\n    - List containing:\\n        - RBP label: text\\n        - Logfold gene expression change: vector\\n        \\nLabels save separately to save storage:\\n- List containing:\\n    - RBP label: text\\n    - Logfold gene expression change: vector\\n'"
      ]
     },
     "execution_count": 48,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "'''\n",
    "Saved data format for each split:\n",
    "- Input\n",
    "    - List containing:\n",
    "        - eCLIP sequence: text\n",
    "        - Tuple of RBNS sequence and binding affinity: (text, vector)\n",
    "- Output\n",
    "    - List containing:\n",
    "        - RBP label: text\n",
    "        - Logfold gene expression change: vector\n",
    "        \n",
    "Labels save separately to save storage:\n",
    "- List containing:\n",
    "    - RBP label: text\n",
    "    - Logfold gene expression change: vector\n",
    "'''"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "id": "6c84bd69",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Save train\n",
    "with open(\"HNRNPC/train_split.pkl\", \"wb\") as f:\n",
    "    pickle.dump((train_data, train_labels), f)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "id": "afb7be9c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Save val\n",
    "with open(\"HNRNPC/val_split.pkl\", \"wb\") as f:\n",
    "    pickle.dump((val_data, val_labels), f)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "id": "6266df79",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Save test\n",
    "with open(\"HNRNPC/test_split.pkl\", \"wb\") as f:\n",
    "    pickle.dump((test_data, test_labels), f)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "id": "95900830",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Save positive label\n",
    "with open(\"HNRNPC/HNRNPC_positive_label.pkl\", \"wb\") as f:\n",
    "    pickle.dump(positive_value, f)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "id": "fe972c58",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Save negative label\n",
    "with open(\"HNRNPC/HNRNPC_negative_label.pkl\", \"wb\") as f:\n",
    "    pickle.dump(negative_value, f)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "521fb44e",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "perturb",
   "language": "python",
   "name": "perturb"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
