{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "3ee91ff5",
   "metadata": {},
   "outputs": [],
   "source": [
    "%load_ext autoreload\n",
    "%autoreload 2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "15865688",
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ff5d5753",
   "metadata": {},
   "source": [
    "## Read data"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "235e3bf2",
   "metadata": {},
   "source": [
    "Load eCLIP-seq sequences"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "abe42208",
   "metadata": {},
   "outputs": [],
   "source": [
    "eclip = pd.read_csv(\"PCBP1/PCBP1_eclip.txt\", sep='\\t', header=None)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "32e0bc25",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr7:45928349-45928392</td>\n",
       "      <td>CTCCGCCGAGAAGCTCGCGCTCTGCCCGCCGGTGTCCGCCTCG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr7:45928417-45928492</td>\n",
       "      <td>GCTGCGGCTGTTGCCCGATGTGCGCCCTGCCTCTGGGCGCCGCGTG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr7:45928229-45928349</td>\n",
       "      <td>TGTCTGCTGCTCGCGCCTGGAGATGTCAGAGGTCCCCGTTGCTCGC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr7:45928392-45928417</td>\n",
       "      <td>TGCTCGGAGGTCACCCGGTCCGCCG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr7:45928531-45928577</td>\n",
       "      <td>TGCACGCCCTCACCCGCGGCCAAGGCGCCTGCGTGCAGGAGTCTGA</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                        0                                                  1\n",
       "0  chr7:45928349-45928392        CTCCGCCGAGAAGCTCGCGCTCTGCCCGCCGGTGTCCGCCTCG\n",
       "1  chr7:45928417-45928492  GCTGCGGCTGTTGCCCGATGTGCGCCCTGCCTCTGGGCGCCGCGTG...\n",
       "2  chr7:45928229-45928349  TGTCTGCTGCTCGCGCCTGGAGATGTCAGAGGTCCCCGTTGCTCGC...\n",
       "3  chr7:45928392-45928417                          TGCTCGGAGGTCACCCGGTCCGCCG\n",
       "4  chr7:45928531-45928577     TGCACGCCCTCACCCGCGGCCAAGGCGCCTGCGTGCAGGAGTCTGA"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "eclip.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8e72135a",
   "metadata": {},
   "source": [
    "Load RBNS binding affinities"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "7bc981f7",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>[PCBP1]</th>\n",
       "      <th>5 nM</th>\n",
       "      <th>20 nM</th>\n",
       "      <th>80 nM</th>\n",
       "      <th>320 nM</th>\n",
       "      <th>1300 nM</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CCCC</td>\n",
       "      <td>1.1704</td>\n",
       "      <td>1.2545</td>\n",
       "      <td>2.2991</td>\n",
       "      <td>3.6549</td>\n",
       "      <td>3.5584</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GCCC</td>\n",
       "      <td>1.5532</td>\n",
       "      <td>1.5971</td>\n",
       "      <td>2.4116</td>\n",
       "      <td>3.4126</td>\n",
       "      <td>3.5377</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>CCCA</td>\n",
       "      <td>1.0617</td>\n",
       "      <td>1.1591</td>\n",
       "      <td>1.8562</td>\n",
       "      <td>2.8203</td>\n",
       "      <td>2.9079</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>CCCG</td>\n",
       "      <td>1.5667</td>\n",
       "      <td>1.5468</td>\n",
       "      <td>2.0563</td>\n",
       "      <td>2.6961</td>\n",
       "      <td>2.7707</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ACCC</td>\n",
       "      <td>1.0986</td>\n",
       "      <td>1.1723</td>\n",
       "      <td>1.7423</td>\n",
       "      <td>2.5136</td>\n",
       "      <td>2.6790</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>251</th>\n",
       "      <td>GAAA</td>\n",
       "      <td>0.9294</td>\n",
       "      <td>0.9057</td>\n",
       "      <td>0.8140</td>\n",
       "      <td>0.6666</td>\n",
       "      <td>0.5976</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>252</th>\n",
       "      <td>TAGA</td>\n",
       "      <td>0.8989</td>\n",
       "      <td>0.8808</td>\n",
       "      <td>0.8041</td>\n",
       "      <td>0.6606</td>\n",
       "      <td>0.5888</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>253</th>\n",
       "      <td>TGAA</td>\n",
       "      <td>0.9222</td>\n",
       "      <td>0.9076</td>\n",
       "      <td>0.8075</td>\n",
       "      <td>0.6521</td>\n",
       "      <td>0.5876</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>254</th>\n",
       "      <td>GAAT</td>\n",
       "      <td>0.9098</td>\n",
       "      <td>0.8883</td>\n",
       "      <td>0.7917</td>\n",
       "      <td>0.6393</td>\n",
       "      <td>0.5715</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>255</th>\n",
       "      <td>AGAA</td>\n",
       "      <td>0.9096</td>\n",
       "      <td>0.8724</td>\n",
       "      <td>0.7830</td>\n",
       "      <td>0.6258</td>\n",
       "      <td>0.5536</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>256 rows Ã— 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    [PCBP1]    5 nM   20 nM   80 nM  320 nM  1300 nM\n",
       "0      CCCC  1.1704  1.2545  2.2991  3.6549   3.5584\n",
       "1      GCCC  1.5532  1.5971  2.4116  3.4126   3.5377\n",
       "2      CCCA  1.0617  1.1591  1.8562  2.8203   2.9079\n",
       "3      CCCG  1.5667  1.5468  2.0563  2.6961   2.7707\n",
       "4      ACCC  1.0986  1.1723  1.7423  2.5136   2.6790\n",
       "..      ...     ...     ...     ...     ...      ...\n",
       "251    GAAA  0.9294  0.9057  0.8140  0.6666   0.5976\n",
       "252    TAGA  0.8989  0.8808  0.8041  0.6606   0.5888\n",
       "253    TGAA  0.9222  0.9076  0.8075  0.6521   0.5876\n",
       "254    GAAT  0.9098  0.8883  0.7917  0.6393   0.5715\n",
       "255    AGAA  0.9096  0.8724  0.7830  0.6258   0.5536\n",
       "\n",
       "[256 rows x 6 columns]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "four_mers = pd.read_csv(\"PCBP1/PCBP1_4.tsv\", sep='\\t')\n",
    "five_mers = pd.read_csv(\"PCBP1/PCBP1_5.tsv\", sep='\\t',)\n",
    "six_mers = pd.read_csv(\"PCBP1/PCBP1_6.tsv\", sep='\\t',)\n",
    "four_mers"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "af6ffd4e",
   "metadata": {},
   "source": [
    "Load knockdown differential rna-seq data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "d0e331a6",
   "metadata": {},
   "outputs": [],
   "source": [
    "# get other diff expr for subsetting to shared genes later\n",
    "diff_exp_HNRNPC = pd.read_csv(\"HNRNPC/HNRNPC.tsv\", sep='\\t',)\n",
    "diff_exp_SRSF9 = pd.read_csv(\"SRSF9/SRSF9.tsv\", sep='\\t',)\n",
    "diff_exp_PCBP1 = pd.read_csv(\"PCBP1/PCBP1.tsv\", sep='\\t',)\n",
    "diff_exp_TIA1 = pd.read_csv(\"TIA1/TIA1.tsv\", sep='\\t',)\n",
    "diff_exp_TRA2A = pd.read_csv(\"TRA2A/TRA2A.tsv\", sep='\\t',)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1612ca75",
   "metadata": {},
   "source": [
    "## Generate positive pairs"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b0787ccc",
   "metadata": {},
   "source": [
    "EClip positive binding sites"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "592b1f36",
   "metadata": {},
   "outputs": [],
   "source": [
    "in_vivo_pos_seqs = eclip[1].values"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "18057862",
   "metadata": {},
   "outputs": [],
   "source": [
    "# filter for length and unique values\n",
    "in_vivo_pos_seqs = [s for s in in_vivo_pos_seqs if len(s) >= 10]\n",
    "in_vivo_pos_seqs = list(set(in_vivo_pos_seqs))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "adf61072",
   "metadata": {},
   "source": [
    "RBNS significant binding affinity"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "fd120254",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_signficant_sequences(df):\n",
    "    num_cols = df.columns[1:]\n",
    "    filtered = df[df[num_cols].gt(2).any(axis=1)]\n",
    "    pairs = [\n",
    "        (row[df.columns[0]], tuple(row[num_cols]))\n",
    "        for _, row in filtered.iterrows()\n",
    "    ]\n",
    "    return pairs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "ed65dc0f",
   "metadata": {},
   "outputs": [],
   "source": [
    "pos_four_mers = get_signficant_sequences(four_mers)\n",
    "pos_five_mers = get_signficant_sequences(five_mers)\n",
    "pos_six_mers = get_signficant_sequences(six_mers)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "15b45a48",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(13, 70, 340)"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(pos_four_mers), len(pos_five_mers), len(pos_six_mers), "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "6894c643",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "<__array_function__ internals>:180: VisibleDeprecationWarning: Creating an ndarray from ragged nested sequences (which is a list-or-tuple of lists-or-tuples-or ndarrays with different lengths or shapes) is deprecated. If you meant to do this, you must specify 'dtype=object' when creating the ndarray.\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "(423, 2)"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "in_vitro_pos_seqs = np.concatenate((pos_four_mers, pos_five_mers,pos_six_mers))\n",
    "in_vitro_pos_seqs.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "3bc23f33",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "423"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "unique_rbns = {}\n",
    "for s, arr in in_vitro_pos_seqs:\n",
    "    if s not in unique_rbns:\n",
    "        unique_rbns[s] = arr\n",
    "\n",
    "in_vitro_pos_seqs = list(unique_rbns.items())\n",
    "len(in_vitro_pos_seqs)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6ad2eb50",
   "metadata": {},
   "source": [
    "Filter diff expr"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "7a6b7538",
   "metadata": {},
   "outputs": [],
   "source": [
    "# remove when gene expression is always 0 or if logfold change is inf/-inf\n",
    "diff_exp_filtered_HNRNPC = diff_exp_HNRNPC[\n",
    "    ~(\n",
    "        ((diff_exp_HNRNPC[\"value_1\"] == 0) & (diff_exp_HNRNPC[\"value_2\"] == 0)) |\n",
    "        (diff_exp_HNRNPC[\"log2(fold_change)\"].isin([np.inf, -np.inf]))\n",
    "    )\n",
    "]\n",
    "\n",
    "diff_exp_filtered_SRSF9 = diff_exp_SRSF9[\n",
    "    ~(\n",
    "        ((diff_exp_SRSF9[\"value_1\"] == 0) & (diff_exp_SRSF9[\"value_2\"] == 0)) |\n",
    "        (diff_exp_SRSF9[\"log2(fold_change)\"].isin([np.inf, -np.inf]))\n",
    "    )\n",
    "]\n",
    "\n",
    "diff_exp_filtered_PCBP1 = diff_exp_PCBP1[\n",
    "    ~(\n",
    "        ((diff_exp_PCBP1[\"value_1\"] == 0) & (diff_exp_PCBP1[\"value_2\"] == 0)) |\n",
    "        (diff_exp_PCBP1[\"log2(fold_change)\"].isin([np.inf, -np.inf]))\n",
    "    )\n",
    "]\n",
    "\n",
    "diff_exp_filtered_TIA1 = diff_exp_TIA1[\n",
    "    ~(\n",
    "        ((diff_exp_TIA1[\"value_1\"] == 0) & (diff_exp_TIA1[\"value_2\"] == 0)) |\n",
    "        (diff_exp_TIA1[\"log2(fold_change)\"].isin([np.inf, -np.inf]))\n",
    "    )\n",
    "]\n",
    "\n",
    "diff_exp_filtered_TRA2A = diff_exp_TRA2A[\n",
    "    ~(\n",
    "        ((diff_exp_TRA2A[\"value_1\"] == 0) & (diff_exp_TRA2A[\"value_2\"] == 0)) |\n",
    "        (diff_exp_TRA2A[\"log2(fold_change)\"].isin([np.inf, -np.inf]))\n",
    "    )\n",
    "]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "377252f2",
   "metadata": {},
   "outputs": [],
   "source": [
    "dfs = [diff_exp_filtered_HNRNPC, diff_exp_filtered_SRSF9, diff_exp_filtered_PCBP1, diff_exp_filtered_TIA1, diff_exp_filtered_TRA2A]\n",
    "col = \"gene_id\"\n",
    "shared_keys = set.intersection(*(set(df[col]) for df in dfs))\n",
    "\n",
    "# 2. Subset the dataframe you want (example: df1)\n",
    "diff_exp_filtered_PCBP1 = diff_exp_filtered_PCBP1[diff_exp_filtered_PCBP1[col].isin(shared_keys)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "fdeaa9e5",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>test_id</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene</th>\n",
       "      <th>locus</th>\n",
       "      <th>sample_1</th>\n",
       "      <th>sample_2</th>\n",
       "      <th>status</th>\n",
       "      <th>value_1</th>\n",
       "      <th>value_2</th>\n",
       "      <th>log2(fold_change)</th>\n",
       "      <th>test_stat</th>\n",
       "      <th>p_value</th>\n",
       "      <th>q_value</th>\n",
       "      <th>significant</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>XLOC_000008</td>\n",
       "      <td>XLOC_000008</td>\n",
       "      <td>OR4F29</td>\n",
       "      <td>chr1:317719-461954</td>\n",
       "      <td>PCBP1-BGHLV12-HepG2</td>\n",
       "      <td>Control</td>\n",
       "      <td>OK</td>\n",
       "      <td>0.644145</td>\n",
       "      <td>0.256604</td>\n",
       "      <td>-1.327840</td>\n",
       "      <td>-0.570958</td>\n",
       "      <td>0.02810</td>\n",
       "      <td>0.088555</td>\n",
       "      <td>no</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>XLOC_000012</td>\n",
       "      <td>XLOC_000012</td>\n",
       "      <td>MTND1P23</td>\n",
       "      <td>chr1:536815-660283</td>\n",
       "      <td>PCBP1-BGHLV12-HepG2</td>\n",
       "      <td>Control</td>\n",
       "      <td>OK</td>\n",
       "      <td>260.081000</td>\n",
       "      <td>195.625000</td>\n",
       "      <td>-0.410873</td>\n",
       "      <td>-0.205877</td>\n",
       "      <td>0.70775</td>\n",
       "      <td>0.836839</td>\n",
       "      <td>no</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>XLOC_000013</td>\n",
       "      <td>XLOC_000013</td>\n",
       "      <td>MTND2P28</td>\n",
       "      <td>chr1:536815-660283</td>\n",
       "      <td>PCBP1-BGHLV12-HepG2</td>\n",
       "      <td>Control</td>\n",
       "      <td>OK</td>\n",
       "      <td>763.174000</td>\n",
       "      <td>674.709000</td>\n",
       "      <td>-0.177747</td>\n",
       "      <td>-0.351151</td>\n",
       "      <td>0.61965</td>\n",
       "      <td>0.776942</td>\n",
       "      <td>no</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>XLOC_000014</td>\n",
       "      <td>XLOC_000014</td>\n",
       "      <td>hsa-mir-6723</td>\n",
       "      <td>chr1:536815-660283</td>\n",
       "      <td>PCBP1-BGHLV12-HepG2</td>\n",
       "      <td>Control</td>\n",
       "      <td>OK</td>\n",
       "      <td>1402.760000</td>\n",
       "      <td>1272.220000</td>\n",
       "      <td>-0.140916</td>\n",
       "      <td>-0.478016</td>\n",
       "      <td>0.49485</td>\n",
       "      <td>0.682270</td>\n",
       "      <td>no</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>XLOC_000015</td>\n",
       "      <td>XLOC_000015</td>\n",
       "      <td>RP5-857K21.7</td>\n",
       "      <td>chr1:536815-660283</td>\n",
       "      <td>PCBP1-BGHLV12-HepG2</td>\n",
       "      <td>Control</td>\n",
       "      <td>OK</td>\n",
       "      <td>829.868000</td>\n",
       "      <td>657.946000</td>\n",
       "      <td>-0.334911</td>\n",
       "      <td>-0.503453</td>\n",
       "      <td>0.46355</td>\n",
       "      <td>0.656650</td>\n",
       "      <td>no</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        test_id      gene_id          gene               locus  \\\n",
       "7   XLOC_000008  XLOC_000008        OR4F29  chr1:317719-461954   \n",
       "11  XLOC_000012  XLOC_000012      MTND1P23  chr1:536815-660283   \n",
       "12  XLOC_000013  XLOC_000013      MTND2P28  chr1:536815-660283   \n",
       "13  XLOC_000014  XLOC_000014  hsa-mir-6723  chr1:536815-660283   \n",
       "14  XLOC_000015  XLOC_000015  RP5-857K21.7  chr1:536815-660283   \n",
       "\n",
       "               sample_1 sample_2 status      value_1      value_2  \\\n",
       "7   PCBP1-BGHLV12-HepG2  Control     OK     0.644145     0.256604   \n",
       "11  PCBP1-BGHLV12-HepG2  Control     OK   260.081000   195.625000   \n",
       "12  PCBP1-BGHLV12-HepG2  Control     OK   763.174000   674.709000   \n",
       "13  PCBP1-BGHLV12-HepG2  Control     OK  1402.760000  1272.220000   \n",
       "14  PCBP1-BGHLV12-HepG2  Control     OK   829.868000   657.946000   \n",
       "\n",
       "    log2(fold_change)  test_stat  p_value   q_value significant  \n",
       "7           -1.327840  -0.570958  0.02810  0.088555          no  \n",
       "11          -0.410873  -0.205877  0.70775  0.836839          no  \n",
       "12          -0.177747  -0.351151  0.61965  0.776942          no  \n",
       "13          -0.140916  -0.478016  0.49485  0.682270          no  \n",
       "14          -0.334911  -0.503453  0.46355  0.656650          no  "
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "diff_exp_filtered_PCBP1.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "1046db6c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(17985, 14)"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "diff_exp_filtered_PCBP1.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "d8fe4752",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([-1.32784   , -0.410873  , -0.177747  , ..., -0.00184541,\n",
       "       -0.781436  ,  1.14329   ])"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "logfold_changes = diff_exp_filtered_PCBP1['log2(fold_change)'].values\n",
    "logfold_changes"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "14a14525",
   "metadata": {},
   "source": [
    "## Generate negative pairs"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "eaa775d1",
   "metadata": {},
   "source": [
    "Shuffle eCLIP sequences"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "68b0a794",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "32074"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import random\n",
    "def shuffle_by_pairs(seq):\n",
    "    pairs = [seq[i:i+2] for i in range(0, len(seq), 2)]\n",
    "    random.shuffle(pairs)\n",
    "    return \"\".join(pairs)\n",
    "in_vivo_neg_seqs = [shuffle_by_pairs(seq) for seq in eclip[1].values]\n",
    "len(in_vivo_neg_seqs)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "1ec9c58c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "30452"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# filter for length and unique values\n",
    "in_vivo_neg_seqs = [s for s in in_vivo_neg_seqs if len(s) >= 10]\n",
    "in_vivo_neg_seqs = list(set(in_vivo_neg_seqs))\n",
    "in_vivo_neg_seqs = list(set(in_vivo_neg_seqs) - set(in_vivo_pos_seqs))\n",
    "len(in_vivo_neg_seqs)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "02a5e383",
   "metadata": {},
   "source": [
    "Retrieve RBNS sequences that are not enriched"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "669ff6f6",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_insignficant_sequences(df):\n",
    "    num_cols = df.columns[1:]\n",
    "    filtered = df[df[num_cols].lt(2).any(axis=1)]\n",
    "    pairs = [\n",
    "        (row[df.columns[0]], tuple(row[num_cols]))\n",
    "        for _, row in filtered.iterrows()\n",
    "    ]\n",
    "    return pairs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "4a0c3f24",
   "metadata": {},
   "outputs": [],
   "source": [
    "neg_four_mers = get_insignficant_sequences(four_mers)\n",
    "neg_five_mers = get_insignficant_sequences(five_mers)\n",
    "neg_six_mers = get_insignficant_sequences(six_mers)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "7920a3d9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[('CCCC', (1.1704, 1.2545, 2.2991, 3.6549, 3.5584)),\n",
       " ('GCCC', (1.5532, 1.5971, 2.4116, 3.4126, 3.5377)),\n",
       " ('CCCA', (1.0617, 1.1591, 1.8562, 2.8203, 2.9079)),\n",
       " ('CCCG', (1.5667, 1.5468, 2.0563, 2.6961, 2.7707)),\n",
       " ('ACCC', (1.0986, 1.1723, 1.7423, 2.5136, 2.679)),\n",
       " ('CCCT', (1.0137, 1.077, 1.6233, 2.408, 2.5261)),\n",
       " ('TCCC', (0.9922, 1.0517, 1.5783, 2.3736, 2.4432)),\n",
       " ('CGCC', (1.4975, 1.4734, 1.7883, 2.2296, 2.4665)),\n",
       " ('CCGC', (1.4534, 1.4531, 1.6498, 1.9877, 2.1969)),\n",
       " ('AGCC', (1.2636, 1.2923, 1.5775, 1.9674, 2.1408)),\n",
       " ('TGCC', (1.2428, 1.2825, 1.5485, 1.9047, 2.107)),\n",
       " ('CCAA', (0.952, 1.0202, 1.3377, 1.837, 1.9983)),\n",
       " ('CCTC', (0.9722, 1.0262, 1.3282, 1.8297, 2.0515)),\n",
       " ('CCTG', (1.2009, 1.2293, 1.4568, 1.7762, 1.8481)),\n",
       " ('CCTA', (0.92, 0.9846, 1.2776, 1.7716, 1.9853)),\n",
       " ('ATCC', (0.9284, 0.9896, 1.2851, 1.7657, 1.8976)),\n",
       " ('GTCC', (1.267, 1.3153, 1.4899, 1.7634, 1.8479)),\n",
       " ('CCAC', (1.0477, 1.1006, 1.3564, 1.7629, 1.982)),\n",
       " ('TACC', (0.9621, 1.0339, 1.2965, 1.7611, 2.0112)),\n",
       " ('CACC', (1.0908, 1.1299, 1.3741, 1.7468, 1.9864)),\n",
       " ('CCAG', (1.206, 1.2485, 1.4537, 1.744, 1.8003)),\n",
       " ('CCGG', (1.5263, 1.4784, 1.6496, 1.7378, 1.5575)),\n",
       " ('TTCC', (0.9186, 0.9645, 1.2079, 1.6755, 1.7977)),\n",
       " ('CCTT', (0.9103, 0.9589, 1.2044, 1.648, 1.8466)),\n",
       " ('CGTC', (1.284, 1.2842, 1.4198, 1.6251, 1.6686)),\n",
       " ('GCCT', (1.1806, 1.1913, 1.355, 1.6044, 1.8331)),\n",
       " ('CCGT', (1.282, 1.2954, 1.4019, 1.5818, 1.643)),\n",
       " ('AACC', (0.9602, 1.0089, 1.1955, 1.5232, 1.7405)),\n",
       " ('GCCA', (1.2439, 1.269, 1.3532, 1.5061, 1.6858)),\n",
       " ('CCAT', (0.9302, 0.9875, 1.1607, 1.4942, 1.6367)),\n",
       " ('GCCG', (1.5794, 1.5056, 1.5199, 1.4929, 1.5037)),\n",
       " ('TCCT', (0.8817, 0.9269, 1.1101, 1.4692, 1.6318)),\n",
       " ('ACCT', (0.9284, 0.98, 1.1393, 1.4633, 1.7222)),\n",
       " ('GGCC', (1.5401, 1.5094, 1.5172, 1.4609, 1.4203)),\n",
       " ('TCCG', (1.2138, 1.2314, 1.2928, 1.4232, 1.4686)),\n",
       " ('CAGC', (1.2322, 1.2466, 1.3071, 1.414, 1.4887)),\n",
       " ('GTCG', (1.3174, 1.2469, 1.3748, 1.4067, 1.2061)),\n",
       " ('CTGC', (1.2126, 1.224, 1.2792, 1.3874, 1.4701)),\n",
       " ('TCCA', (0.9135, 0.9647, 1.1044, 1.3735, 1.5206)),\n",
       " ('CGGC', (1.5179, 1.4557, 1.4527, 1.3701, 1.2551)),\n",
       " ('CGCG', (1.5728, 1.5094, 1.4702, 1.3677, 1.2959)),\n",
       " ('CTCC', (0.9216, 0.9516, 1.0881, 1.3487, 1.5047)),\n",
       " ('GCGC', (1.5706, 1.4952, 1.4559, 1.3311, 1.302)),\n",
       " ('CCGA', (1.2292, 1.2071, 1.2461, 1.2956, 1.3249)),\n",
       " ('GTCA', (1.0991, 1.1121, 1.1887, 1.2934, 1.3032)),\n",
       " ('GGTC', (1.2786, 1.2699, 1.3311, 1.2921, 1.1046)),\n",
       " ('CTAC', (0.9252, 0.9749, 1.0725, 1.2891, 1.419)),\n",
       " ('ACCG', (1.2599, 1.2606, 1.2624, 1.2876, 1.3569)),\n",
       " ('TCGG', (1.2402, 1.1787, 1.2721, 1.2717, 1.0497)),\n",
       " ('CACG', (1.2564, 1.2644, 1.2629, 1.2635, 1.2741)),\n",
       " ('ACCA', (0.9439, 0.9965, 1.0707, 1.2566, 1.4698)),\n",
       " ('GGGC', (1.5091, 1.4083, 1.3973, 1.2414, 1.0649)),\n",
       " ('CAAC', (0.9546, 0.9843, 1.0683, 1.2377, 1.3537)),\n",
       " ('GCGG', (1.5253, 1.4314, 1.4076, 1.2353, 1.0461)),\n",
       " ('TGTC', (1.1093, 1.1341, 1.1738, 1.235, 1.2376)),\n",
       " ('CGGT', (1.3081, 1.2902, 1.3055, 1.226, 1.0379)),\n",
       " ('CGGG', (1.5028, 1.396, 1.3869, 1.2198, 1.029)),\n",
       " ('CATC', (0.9265, 0.9569, 1.0415, 1.2192, 1.3121)),\n",
       " ('AGTC', (1.0807, 1.0882, 1.1422, 1.2138, 1.2081)),\n",
       " ('GACC', (1.2035, 1.2092, 1.2164, 1.2083, 1.2339)),\n",
       " ('CTTC', (0.8995, 0.9308, 1.0076, 1.2028, 1.3047)),\n",
       " ('CTGG', (1.2323, 1.2148, 1.2524, 1.1992, 1.0856)),\n",
       " ('CTGT', (1.1066, 1.1135, 1.1513, 1.1974, 1.1799)),\n",
       " ('ACGC', (1.2744, 1.2635, 1.2388, 1.1937, 1.2091)),\n",
       " ('GGCG', (1.5435, 1.4383, 1.3945, 1.185, 1.0083)),\n",
       " ('CTCG', (1.1911, 1.1686, 1.1741, 1.161, 1.1401)),\n",
       " ('CGCA', (1.2587, 1.2418, 1.2027, 1.1577, 1.1761)),\n",
       " ('GCAC', (1.2622, 1.2208, 1.2122, 1.1517, 1.1779)),\n",
       " ('TCGC', (1.2168, 1.1702, 1.1781, 1.146, 1.1234)),\n",
       " ('CAAG', (1.0793, 1.0868, 1.1056, 1.1336, 1.1197)),\n",
       " ('CGGA', (1.2297, 1.1781, 1.2065, 1.1331, 0.9676)),\n",
       " ('GGAC', (1.2375, 1.171, 1.2054, 1.1262, 0.9695)),\n",
       " ('CGAC', (1.2218, 1.211, 1.1825, 1.123, 1.093)),\n",
       " ('CGTG', (1.3235, 1.3051, 1.2429, 1.1208, 1.0422)),\n",
       " ('CAGT', (1.0954, 1.1093, 1.1098, 1.1159, 1.1023)),\n",
       " ('GTGC', (1.3341, 1.3134, 1.2451, 1.1132, 1.0683)),\n",
       " ('CGCT', (1.2166, 1.1899, 1.1599, 1.1017, 1.1192)),\n",
       " ('CAGG', (1.2592, 1.2491, 1.2065, 1.1003, 0.9984)),\n",
       " ('AAGC', (1.094, 1.0914, 1.1007, 1.0968, 1.096)),\n",
       " ('CAAA', (0.8844, 0.8992, 0.9682, 1.0857, 1.159)),\n",
       " ('CTTT', (0.879, 0.9024, 0.9462, 1.0824, 1.1725)),\n",
       " ('GCGT', (1.3578, 1.334, 1.2412, 1.0822, 1.0116)),\n",
       " ('CTCA', (0.8935, 0.9248, 0.9717, 1.0822, 1.1905)),\n",
       " ('ATCG', (1.0658, 1.0411, 1.0852, 1.0811, 0.9801)),\n",
       " ('GATC', (1.0398, 1.0187, 1.076, 1.0802, 0.999)),\n",
       " ('TTTC', (0.8785, 0.9019, 0.9414, 1.0784, 1.1513)),\n",
       " ('CAAT', (0.8751, 0.8994, 0.9587, 1.0765, 1.1325)),\n",
       " ('CGTT', (1.1503, 1.1466, 1.116, 1.074, 1.0614)),\n",
       " ('CACA', (0.9561, 0.9796, 0.9991, 1.0663, 1.1644)),\n",
       " ('CTAT', (0.8546, 0.8735, 0.9336, 1.0643, 1.1548)),\n",
       " ('ATGC', (1.0765, 1.0874, 1.0847, 1.0606, 1.0805)),\n",
       " ('AGCG', (1.2975, 1.2751, 1.2077, 1.0584, 0.9567)),\n",
       " ('GCTG', (1.2749, 1.2371, 1.1833, 1.0525, 0.9522)),\n",
       " ('TCGT', (1.1038, 1.083, 1.0913, 1.0524, 0.9969)),\n",
       " ('GGCA', (1.2768, 1.2087, 1.185, 1.051, 0.9511)),\n",
       " ('TGGC', (1.2774, 1.2198, 1.19, 1.0469, 0.9739)),\n",
       " ('TGCG', (1.3212, 1.2913, 1.1978, 1.0424, 0.9593)),\n",
       " ('TATC', (0.8617, 0.8886, 0.9296, 1.0406, 1.1215)),\n",
       " ('AATC', (0.8579, 0.8724, 0.932, 1.0311, 1.067)),\n",
       " ('GACG', (1.3025, 1.2513, 1.194, 1.0237, 0.8822)),\n",
       " ('CTCT', (0.8776, 0.8934, 0.9272, 1.0235, 1.1214)),\n",
       " ('CTAG', (1.0248, 1.0288, 1.0298, 1.0184, 1.0132)),\n",
       " ('ACGT', (1.1345, 1.1403, 1.0942, 1.0178, 0.993)),\n",
       " ('AGGC', (1.2779, 1.2415, 1.1772, 1.0145, 0.9172)),\n",
       " ('ACGG', (1.2761, 1.2534, 1.1765, 1.0143, 0.8821)),\n",
       " ('ACAC', (0.9693, 0.994, 0.9907, 1.0097, 1.075)),\n",
       " ('CATG', (1.0588, 1.0797, 1.0402, 1.007, 1.0114)),\n",
       " ('TCAC', (0.9333, 0.9586, 0.961, 1.0059, 1.1005)),\n",
       " ('CACT', (0.9266, 0.9459, 0.9568, 1.0019, 1.0865)),\n",
       " ('GTTC', (1.1211, 1.1131, 1.0593, 0.9967, 0.9705)),\n",
       " ('GCAG', (1.2961, 1.2492, 1.1623, 0.9938, 0.8975)),\n",
       " ('CTTG', (1.0366, 1.0315, 1.0186, 0.9915, 0.9966)),\n",
       " ('GTCT', (1.0821, 1.0927, 1.0464, 0.9871, 0.9718)),\n",
       " ('GAGC', (1.2458, 1.2095, 1.1403, 0.9836, 0.8832)),\n",
       " ('GCTC', (1.1608, 1.1307, 1.0759, 0.9823, 0.9832)),\n",
       " ('CGAG', (1.2845, 1.1878, 1.1379, 0.9756, 0.8553)),\n",
       " ('TCAG', (1.0464, 1.0507, 1.0196, 0.9711, 0.9261)),\n",
       " ('TAGC', (1.0544, 1.0597, 1.0197, 0.9693, 0.9867)),\n",
       " ('CGTA', (1.1033, 1.1046, 1.0511, 0.9677, 0.9476)),\n",
       " ('TCTC', (0.8797, 0.9006, 0.9002, 0.9591, 1.0546)),\n",
       " ('TTGC', (1.0698, 1.0648, 1.0155, 0.9591, 0.9697)),\n",
       " ('CTAA', (0.8486, 0.8665, 0.8953, 0.9587, 1.0262)),\n",
       " ('CATT', (0.8734, 0.9004, 0.8963, 0.9557, 1.0164)),\n",
       " ('GGCT', (1.2405, 1.1856, 1.1193, 0.9513, 0.8305)),\n",
       " ('GGGG', (1.4036, 1.3169, 1.2204, 0.9502, 0.7802)),\n",
       " ('CGAT', (1.0647, 1.0639, 1.0194, 0.946, 0.8793)),\n",
       " ('AACG', (1.102, 1.0963, 1.0413, 0.9432, 0.8855)),\n",
       " ('ATAC', (0.8691, 0.8929, 0.9024, 0.9422, 0.993)),\n",
       " ('GCGA', (1.2785, 1.2298, 1.1311, 0.94, 0.822)),\n",
       " ('GACA', (1.0733, 1.0598, 1.0264, 0.9382, 0.8787)),\n",
       " ('GTAC', (1.102, 1.113, 1.0393, 0.9369, 0.9107)),\n",
       " ('TCAA', (0.8665, 0.8804, 0.8918, 0.9354, 1.0006)),\n",
       " ('TTCG', (1.0712, 1.0543, 1.0094, 0.9298, 0.8659)),\n",
       " ('TTAC', (0.8776, 0.9006, 0.8939, 0.9287, 1.0014)),\n",
       " ('TCTG', (1.0414, 1.0417, 0.9933, 0.9266, 0.8945)),\n",
       " ('ACTG', (1.0643, 1.0593, 1.0124, 0.9261, 0.8695)),\n",
       " ('CTTA', (0.8481, 0.8637, 0.8733, 0.9253, 0.9981)),\n",
       " ('CATA', (0.8616, 0.8838, 0.8817, 0.9236, 0.9868)),\n",
       " ('GTTG', (1.1268, 1.0992, 1.0489, 0.9236, 0.8052)),\n",
       " ('AAAC', (0.8884, 0.896, 0.9006, 0.9211, 0.9691)),\n",
       " ('TACG', (1.0906, 1.1014, 1.023, 0.9141, 0.8642)),\n",
       " ('TCGA', (1.0399, 1.0179, 0.9937, 0.9102, 0.8179)),\n",
       " ('GTGG', (1.2452, 1.2099, 1.1167, 0.9085, 0.7684)),\n",
       " ('GGTG', (1.2704, 1.2453, 1.1346, 0.9082, 0.7582)),\n",
       " ('TCAT', (0.8601, 0.8801, 0.8692, 0.9005, 0.9613)),\n",
       " ('TGAC', (1.0562, 1.0557, 1.0043, 0.8977, 0.8586)),\n",
       " ('CTGA', (1.0237, 1.0183, 0.9793, 0.895, 0.854)),\n",
       " ('TGGT', (1.0888, 1.0937, 1.0386, 0.8948, 0.7609)),\n",
       " ('CAGA', (1.0463, 1.0334, 0.9843, 0.8922, 0.8348)),\n",
       " ('AGAC', (1.0632, 1.0518, 1.0038, 0.8883, 0.821)),\n",
       " ('ATTC', (0.8462, 0.8582, 0.8513, 0.8853, 0.9323)),\n",
       " ('GCAA', (1.0857, 1.0645, 0.9949, 0.8852, 0.8556)),\n",
       " ('ACAG', (1.0748, 1.0688, 0.9918, 0.8775, 0.8319)),\n",
       " ('CGAA', (1.0534, 1.0412, 0.9771, 0.8754, 0.8245)),\n",
       " ('TCTT', (0.8569, 0.8717, 0.8523, 0.8741, 0.9377)),\n",
       " ('TGGG', (1.2151, 1.1719, 1.0772, 0.871, 0.7248)),\n",
       " ('GCAT', (1.0722, 1.0711, 0.9808, 0.8706, 0.8486)),\n",
       " ('TTGG', (1.0576, 1.028, 0.9897, 0.865, 0.7435)),\n",
       " ('GGGT', (1.2533, 1.2119, 1.0966, 0.8562, 0.7039)),\n",
       " ('TAAC', (0.8753, 0.8911, 0.8639, 0.8534, 0.9048)),\n",
       " ('TGCA', (1.0813, 1.0742, 0.9814, 0.8521, 0.8341)),\n",
       " ('AGGG', (1.2264, 1.1825, 1.0753, 0.8484, 0.7032)),\n",
       " ('TTTT', (0.8951, 0.8955, 0.8523, 0.8464, 0.8752)),\n",
       " ('AGCA', (1.0764, 1.0704, 0.9736, 0.845, 0.8183)),\n",
       " ('TCTA', (0.8439, 0.8578, 0.8364, 0.8433, 0.9061)),\n",
       " ('TGCT', (1.061, 1.0582, 0.9655, 0.8428, 0.8226)),\n",
       " ('TGTG', (1.1073, 1.104, 0.9956, 0.8419, 0.7623)),\n",
       " ('GTGT', (1.153, 1.147, 1.0222, 0.8416, 0.7604)),\n",
       " ('GGAG', (1.1971, 1.1438, 1.0587, 0.8407, 0.6895)),\n",
       " ('AGTG', (1.0905, 1.0762, 0.9933, 0.8379, 0.7461)),\n",
       " ('GCTA', (1.0457, 1.0308, 0.9545, 0.8339, 0.8047)),\n",
       " ('GCTT', (1.056, 1.0394, 0.9508, 0.8327, 0.8084)),\n",
       " ('GGTT', (1.11, 1.085, 1.0006, 0.8258, 0.6861)),\n",
       " ('AAGT', (0.9687, 0.9512, 0.9054, 0.8257, 0.7866)),\n",
       " ('ATCA', (0.8608, 0.8739, 0.8428, 0.8251, 0.8635)),\n",
       " ('GACT', (1.0162, 1.0028, 0.945, 0.8249, 0.7355)),\n",
       " ('GAGG', (1.1989, 1.1371, 1.0453, 0.82, 0.6685)),\n",
       " ('ACGA', (1.0732, 1.0666, 0.9712, 0.8191, 0.7389)),\n",
       " ('ACTC', (0.8669, 0.8736, 0.8451, 0.8175, 0.8554)),\n",
       " ('TTCT', (0.8608, 0.8751, 0.8254, 0.8139, 0.8813)),\n",
       " ('ACAA', (0.884, 0.888, 0.8504, 0.8133, 0.8345)),\n",
       " ('GAAC', (1.0373, 1.0228, 0.9495, 0.813, 0.7549)),\n",
       " ('AGCT', (1.0425, 1.0304, 0.94, 0.8102, 0.7721)),\n",
       " ('TGTT', (1.0176, 1.0144, 0.9237, 0.8102, 0.7626)),\n",
       " ('ACAT', (0.8758, 0.8884, 0.8429, 0.8087, 0.8382)),\n",
       " ('GGGA', (1.1932, 1.1268, 1.0342, 0.8087, 0.6615)),\n",
       " ('TTCA', (0.8704, 0.8857, 0.8316, 0.8065, 0.8578)),\n",
       " ('AAAG', (0.9547, 0.9302, 0.887, 0.7993, 0.7458)),\n",
       " ('AATG', (0.9403, 0.9417, 0.8847, 0.797, 0.7534)),\n",
       " ('ATCT', (0.8507, 0.8605, 0.8184, 0.7964, 0.8404)),\n",
       " ('ATGT', (0.9757, 0.9843, 0.8968, 0.7962, 0.7713)),\n",
       " ('TTGT', (0.9898, 0.996, 0.9066, 0.7957, 0.7449)),\n",
       " ('GTTA', (0.9828, 0.9699, 0.8994, 0.7918, 0.7527)),\n",
       " ('TTTG', (0.964, 0.9621, 0.8819, 0.7918, 0.7543)),\n",
       " ('GTTT', (1.0276, 1.0121, 0.9185, 0.7898, 0.7383)),\n",
       " ('AAGG', (1.0708, 1.0471, 0.9592, 0.7891, 0.6773)),\n",
       " ('AGGT', (1.0841, 1.0673, 0.9659, 0.7839, 0.6706)),\n",
       " ('TGGA', (1.0305, 1.0183, 0.9396, 0.7837, 0.6719)),\n",
       " ('GATG', (1.0643, 1.0484, 0.9539, 0.7837, 0.6838)),\n",
       " ('GGAT', (1.0287, 0.9968, 0.9338, 0.7826, 0.6526)),\n",
       " ('TAGT', (0.949, 0.9448, 0.8808, 0.7821, 0.7398)),\n",
       " ('GTAG', (1.0661, 1.045, 0.9515, 0.7787, 0.6845)),\n",
       " ('TACA', (0.876, 0.8924, 0.8294, 0.7708, 0.79)),\n",
       " ('AAAT', (0.8174, 0.8083, 0.7901, 0.7671, 0.772)),\n",
       " ('AGTT', (0.9847, 0.9598, 0.8855, 0.767, 0.7148)),\n",
       " ('TATG', (0.936, 0.9408, 0.8634, 0.7639, 0.7326)),\n",
       " ('ATGG', (1.055, 1.0517, 0.9376, 0.7624, 0.6666)),\n",
       " ('ATTG', (0.9372, 0.9426, 0.8665, 0.7622, 0.6985)),\n",
       " ('ACTA', (0.8428, 0.8575, 0.8114, 0.7614, 0.7738)),\n",
       " ('ACTT', (0.8602, 0.8704, 0.8096, 0.7597, 0.7785)),\n",
       " ('GAAG', (1.0655, 1.0303, 0.9346, 0.7596, 0.6506)),\n",
       " ('ATTT', (0.8435, 0.8431, 0.7922, 0.7587, 0.767)),\n",
       " ('TGAT', (0.9199, 0.9207, 0.8621, 0.7569, 0.685)),\n",
       " ('GTGA', (1.0686, 1.0505, 0.9467, 0.7569, 0.6493)),\n",
       " ('AACA', (0.8827, 0.8862, 0.8262, 0.7564, 0.7697)),\n",
       " ('GGTA', (1.0726, 1.0612, 0.9499, 0.7537, 0.6379)),\n",
       " ('TAGG', (1.0407, 1.0229, 0.9302, 0.7519, 0.6375)),\n",
       " ('AGAG', (1.0503, 1.0056, 0.9222, 0.7463, 0.6393)),\n",
       " ('TTAT', (0.8249, 0.8193, 0.7786, 0.7443, 0.7672)),\n",
       " ('TAAG', (0.928, 0.9182, 0.8567, 0.744, 0.6929)),\n",
       " ('AAAA', (0.8271, 0.8045, 0.7817, 0.7422, 0.7532)),\n",
       " ('TTTA', (0.8472, 0.8402, 0.7919, 0.742, 0.7564)),\n",
       " ('AGGA', (1.0307, 1.0039, 0.9211, 0.741, 0.6157)),\n",
       " ('TACT', (0.8496, 0.8682, 0.8026, 0.7391, 0.7563)),\n",
       " ('TTAG', (0.9265, 0.9166, 0.8493, 0.7374, 0.6807)),\n",
       " ('GAGT', (1.0681, 1.0309, 0.9267, 0.737, 0.637)),\n",
       " ('TGAG', (1.0345, 1.0065, 0.9081, 0.7346, 0.6413)),\n",
       " ('TATT', (0.8193, 0.8188, 0.7709, 0.7312, 0.7455)),\n",
       " ('TTGA', (0.9222, 0.9196, 0.8455, 0.7294, 0.6694)),\n",
       " ('GGAA', (1.0346, 1.011, 0.9129, 0.7245, 0.6092)),\n",
       " ('ATAG', (0.9176, 0.9066, 0.8322, 0.7187, 0.6686)),\n",
       " ('AATA', (0.7968, 0.7896, 0.7591, 0.7164, 0.7246)),\n",
       " ('TAAT', (0.8058, 0.8029, 0.7623, 0.716, 0.7233)),\n",
       " ('AACT', (0.8466, 0.8469, 0.7867, 0.7116, 0.71)),\n",
       " ('TAAA', (0.8215, 0.8057, 0.7661, 0.7089, 0.7101)),\n",
       " ('AGAT', (0.9163, 0.897, 0.8351, 0.7083, 0.6255)),\n",
       " ('GTAT', (0.9578, 0.9508, 0.8497, 0.7082, 0.6531)),\n",
       " ('AATT', (0.818, 0.8098, 0.7611, 0.708, 0.7057)),\n",
       " ('TGTA', (0.962, 0.9603, 0.8519, 0.7053, 0.6564)),\n",
       " ('ATAT', (0.8036, 0.7972, 0.7521, 0.7024, 0.7159)),\n",
       " ('GTAA', (0.9571, 0.9425, 0.848, 0.7002, 0.6447)),\n",
       " ('TTAA', (0.8272, 0.8164, 0.7637, 0.6994, 0.698)),\n",
       " ('AGTA', (0.9376, 0.9268, 0.837, 0.6978, 0.6468)),\n",
       " ('TATA', (0.8006, 0.7862, 0.7484, 0.6966, 0.7139)),\n",
       " ('GAGA', (1.038, 0.9659, 0.8865, 0.6955, 0.5815)),\n",
       " ('GATT', (0.919, 0.9089, 0.8268, 0.6938, 0.6083)),\n",
       " ('ATTA', (0.7991, 0.7976, 0.7472, 0.6901, 0.6909)),\n",
       " ('AAGA', (0.9253, 0.9038, 0.8249, 0.6811, 0.6058)),\n",
       " ('GATA', (0.9112, 0.9037, 0.8153, 0.6771, 0.6049)),\n",
       " ('ATAA', (0.8001, 0.7912, 0.7434, 0.6764, 0.6772)),\n",
       " ('ATGA', (0.9117, 0.9082, 0.8167, 0.6756, 0.6142)),\n",
       " ('GAAA', (0.9294, 0.9057, 0.814, 0.6666, 0.5976)),\n",
       " ('TAGA', (0.8989, 0.8808, 0.8041, 0.6606, 0.5888)),\n",
       " ('TGAA', (0.9222, 0.9076, 0.8075, 0.6521, 0.5876)),\n",
       " ('GAAT', (0.9098, 0.8883, 0.7917, 0.6393, 0.5715)),\n",
       " ('AGAA', (0.9096, 0.8724, 0.783, 0.6258, 0.5536))]"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "neg_four_mers"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "47bc768a",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "<__array_function__ internals>:180: VisibleDeprecationWarning: Creating an ndarray from ragged nested sequences (which is a list-or-tuple of lists-or-tuples-or ndarrays with different lengths or shapes) is deprecated. If you meant to do this, you must specify 'dtype=object' when creating the ndarray.\n",
      "/tmp/ipykernel_614325/4178471540.py:3: VisibleDeprecationWarning: Creating an ndarray from ragged nested sequences (which is a list-or-tuple of lists-or-tuples-or ndarrays with different lengths or shapes) is deprecated. If you meant to do this, you must specify 'dtype=object' when creating the ndarray.\n",
      "  in_vitro_neg_seqs = np.array([(s, v) for s, v in in_vitro_neg_seqs_all if s not in pos_rbns_seqs])\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "(4953, 2)"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "in_vitro_neg_seqs_all = np.concatenate((neg_four_mers,neg_five_mers,neg_six_mers))\n",
    "pos_rbns_seqs = {s for s, _ in in_vitro_pos_seqs}\n",
    "in_vitro_neg_seqs = np.array([(s, v) for s, v in in_vitro_neg_seqs_all if s not in pos_rbns_seqs])\n",
    "in_vitro_neg_seqs.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "88397cff",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "4953"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "unique_rbns = {}\n",
    "for s, arr in in_vitro_neg_seqs:\n",
    "    if s not in unique_rbns:\n",
    "        unique_rbns[s] = arr\n",
    "\n",
    "in_vitro_neg_seqs = list(unique_rbns.items())\n",
    "len(in_vitro_neg_seqs)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "94988d48",
   "metadata": {},
   "source": [
    "Construct placeholder expression change vector"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "950281ea",
   "metadata": {},
   "outputs": [],
   "source": [
    "placeholder_logfold_changes = np.zeros(shape=logfold_changes.shape)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b14cf59a",
   "metadata": {},
   "source": [
    "## Split into train/val/test"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "e9d8ef74",
   "metadata": {},
   "outputs": [],
   "source": [
    "# number of total samples per class\n",
    "combinations_size = 500000"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a99c5d06",
   "metadata": {},
   "source": [
    "Positive samples"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "c5fdbf5e",
   "metadata": {},
   "outputs": [],
   "source": [
    "from itertools import product\n",
    "\n",
    "random.shuffle(in_vivo_pos_seqs)\n",
    "random.shuffle(in_vitro_pos_seqs)\n",
    "\n",
    "n_eclip = len(in_vivo_pos_seqs)\n",
    "n_rbns  = len(in_vitro_pos_seqs)\n",
    "\n",
    "# Split indices\n",
    "train_eclip = in_vivo_pos_seqs[:int(0.7*n_eclip)]\n",
    "val_eclip   = in_vivo_pos_seqs[int(0.7*n_eclip):int(0.85*n_eclip)]\n",
    "test_eclip  = in_vivo_pos_seqs[int(0.85*n_eclip):]\n",
    "\n",
    "train_rbns = in_vitro_pos_seqs[:int(0.7*n_rbns)]\n",
    "val_rbns   = in_vitro_pos_seqs[int(0.7*n_rbns):int(0.85*n_rbns)]\n",
    "test_rbns  = in_vitro_pos_seqs[int(0.85*n_rbns):]\n",
    "\n",
    "pos_train_samples = [\n",
    "    (random.choice(train_eclip), random.choice(train_rbns))\n",
    "    for _ in range(combinations_size)\n",
    "]\n",
    "pos_val_samples = [\n",
    "    (random.choice(val_eclip), random.choice(val_rbns))\n",
    "    for _ in range(combinations_size)\n",
    "]\n",
    "pos_test_samples = [\n",
    "    (random.choice(test_eclip), random.choice(test_rbns))\n",
    "    for _ in range(combinations_size)\n",
    "]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "91baf0a1",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "eCLIP sequences are disjoint across splits\n"
     ]
    }
   ],
   "source": [
    "pos_train_eclips = set(eclip for eclip, _ in pos_train_samples)\n",
    "pos_val_eclips   = set(eclip for eclip, _ in pos_val_samples)\n",
    "pos_test_eclips  = set(eclip for eclip, _ in pos_test_samples)\n",
    "\n",
    "assert pos_train_eclips.isdisjoint(pos_val_eclips), \"Overlap between train and val eCLIP sequences!\"\n",
    "assert pos_train_eclips.isdisjoint(pos_test_eclips), \"Overlap between train and test eCLIP sequences!\"\n",
    "assert pos_val_eclips.isdisjoint(pos_test_eclips), \"Overlap between val and test eCLIP sequences!\"\n",
    "print(\"eCLIP sequences are disjoint across splits\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "0ffaaf81",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "RBNS sequences are disjoint across splits\n"
     ]
    }
   ],
   "source": [
    "pos_train_rbns = set(rbns for _, (rbns, _) in pos_train_samples)\n",
    "pos_val_rbns   = set(rbns for _, (rbns, _) in pos_val_samples)\n",
    "pos_test_rbns  = set(rbns for _, (rbns, _) in pos_test_samples)\n",
    "\n",
    "assert pos_train_rbns.isdisjoint(val_rbns), \"Overlap between train and val RBNS sequences!\"\n",
    "assert pos_val_rbns.isdisjoint(test_rbns), \"Overlap between train and test RBNS sequences!\"\n",
    "assert pos_test_rbns.isdisjoint(test_rbns), \"Overlap between val and test RBNS sequences!\"\n",
    "print(\"RBNS sequences are disjoint across splits\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "330d4bc3",
   "metadata": {},
   "outputs": [],
   "source": [
    "pos_train_label = np.ones(len(pos_train_samples))\n",
    "pos_val_label = np.ones(len(pos_val_samples))\n",
    "pos_test_label = np.ones(len(pos_test_samples))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c3da2c3a",
   "metadata": {},
   "source": [
    "Negative samples"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "d7c965b7",
   "metadata": {},
   "outputs": [],
   "source": [
    "random.shuffle(in_vivo_neg_seqs)\n",
    "random.shuffle(in_vitro_neg_seqs)\n",
    "\n",
    "n_eclip = len(in_vivo_neg_seqs)\n",
    "n_rbns  = len(in_vitro_neg_seqs)\n",
    "\n",
    "# Split indices\n",
    "train_eclip = in_vivo_neg_seqs[:int(0.7*n_eclip)]\n",
    "val_eclip   = in_vivo_neg_seqs[int(0.7*n_eclip):int(0.85*n_eclip)]\n",
    "test_eclip  = in_vivo_neg_seqs[int(0.85*n_eclip):]\n",
    "\n",
    "train_rbns = in_vitro_neg_seqs[:int(0.7*n_rbns)]\n",
    "val_rbns   = in_vitro_neg_seqs[int(0.7*n_rbns):int(0.85*n_rbns)]\n",
    "test_rbns  = in_vitro_neg_seqs[int(0.85*n_rbns):]\n",
    "\n",
    "neg_train_samples = [\n",
    "    (random.choice(train_eclip), random.choice(train_rbns))\n",
    "    for _ in range(combinations_size)\n",
    "]\n",
    "neg_val_samples = [\n",
    "    (random.choice(val_eclip), random.choice(val_rbns))\n",
    "    for _ in range(combinations_size)\n",
    "]\n",
    "neg_test_samples = [\n",
    "    (random.choice(test_eclip), random.choice(test_rbns))\n",
    "    for _ in range(combinations_size)\n",
    "]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "id": "19d1f6e3",
   "metadata": {},
   "outputs": [],
   "source": [
    "neg_train_label = np.zeros(len(neg_train_samples))\n",
    "neg_val_label = np.zeros(len(neg_val_samples))\n",
    "neg_test_label = np.zeros(len(neg_test_samples))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "id": "89d4bbb4",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "eCLIP sequences are disjoint across splits\n"
     ]
    }
   ],
   "source": [
    "train_eclips = set(eclip for eclip, _ in neg_train_samples)\n",
    "val_eclips   = set(eclip for eclip, _ in neg_val_samples)\n",
    "test_eclips  = set(eclip for eclip, _ in neg_test_samples)\n",
    "\n",
    "assert train_eclips.isdisjoint(val_eclips), \"Overlap between train and val eCLIP sequences!\"\n",
    "assert train_eclips.isdisjoint(test_eclips), \"Overlap between train and test eCLIP sequences!\"\n",
    "assert val_eclips.isdisjoint(test_eclips), \"Overlap between val and test eCLIP sequences!\"\n",
    "print(\"eCLIP sequences are disjoint across splits\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "15d118d8",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "RBNS sequences are disjoint across splits\n"
     ]
    }
   ],
   "source": [
    "train_rbns = set(rbns for _, (rbns, _) in neg_train_samples)\n",
    "val_rbns   = set(rbns for _, (rbns, _) in neg_val_samples)\n",
    "test_rbns  = set(rbns for _, (rbns, _) in neg_test_samples)\n",
    "\n",
    "assert train_rbns.isdisjoint(val_rbns), \"Overlap between train and val RBNS sequences!\"\n",
    "assert train_rbns.isdisjoint(test_rbns), \"Overlap between train and test RBNS sequences!\"\n",
    "assert val_rbns.isdisjoint(test_rbns), \"Overlap between val and test RBNS sequences!\"\n",
    "print(\"RBNS sequences are disjoint across splits\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "id": "593743cd",
   "metadata": {},
   "outputs": [],
   "source": [
    "# check across pos and neg samples\n",
    "assert train_eclips.isdisjoint(pos_train_eclips), \"Overlap between train and val eCLIP sequences!\"\n",
    "assert val_eclips.isdisjoint(pos_val_eclips), \"Overlap between train and test eCLIP sequences!\"\n",
    "assert test_eclips.isdisjoint(pos_test_eclips), \"Overlap between val and test eCLIP sequences!\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "id": "e8ff8e67",
   "metadata": {},
   "outputs": [],
   "source": [
    "assert train_rbns.isdisjoint(pos_train_rbns), \"Overlap between train and val RBNS sequences!\"\n",
    "assert val_rbns.isdisjoint(pos_val_rbns), \"Overlap between train and test RBNS sequences!\"\n",
    "assert test_rbns.isdisjoint(pos_test_rbns), \"Overlap between val and test RBNS sequences!\""
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9cc54a9e",
   "metadata": {},
   "source": [
    "Combine for full splits"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "id": "726d4838",
   "metadata": {},
   "outputs": [],
   "source": [
    "train_data = np.concatenate((pos_train_samples, neg_train_samples))\n",
    "val_data = np.concatenate((pos_val_samples, neg_val_samples))\n",
    "test_data = np.concatenate((pos_test_samples, neg_test_samples))\n",
    "\n",
    "train_labels = np.concatenate((pos_train_label, neg_train_label))\n",
    "val_labels = np.concatenate((pos_val_label, neg_val_label))\n",
    "test_labels = np.concatenate((pos_test_label, neg_test_label))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "id": "8ce65d9c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# get positive/negative label values\n",
    "positive_value = (\"PCBP1\", logfold_changes)\n",
    "negative_value = (\"Not PCBP1\", placeholder_logfold_changes)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "877f0473",
   "metadata": {},
   "source": [
    "## Save as dictionaries"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "16dcdac6",
   "metadata": {},
   "source": [
    "Save as pickle files since there are nested data structures"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "id": "eff9d82c",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pickle"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "id": "cfda55f9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'\\nSaved data format for each split:\\n- Input\\n    - List containing:\\n        - eCLIP sequence: text\\n        - Tuple of RBNS sequence and binding affinity: (text, vector)\\n- Output\\n    - List containing:\\n        - RBP label: text\\n        - Logfold gene expression change: vector\\n        \\nLabels save separately to save storage:\\n- List containing:\\n    - RBP label: text\\n    - Logfold gene expression change: vector\\n'"
      ]
     },
     "execution_count": 46,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "'''\n",
    "Saved data format for each split:\n",
    "- Input\n",
    "    - List containing:\n",
    "        - eCLIP sequence: text\n",
    "        - Tuple of RBNS sequence and binding affinity: (text, vector)\n",
    "        \n",
    "Labels save separately to save storage:\n",
    "- List containing:\n",
    "    - RBP label: text\n",
    "    - Logfold gene expression change: vector\n",
    "'''"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "id": "6c84bd69",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Save train\n",
    "with open(\"PCBP1/train_split.pkl\", \"wb\") as f:\n",
    "    pickle.dump((train_data, train_labels), f)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "id": "afb7be9c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Save val\n",
    "with open(\"PCBP1/val_split.pkl\", \"wb\") as f:\n",
    "    pickle.dump((val_data, val_labels), f)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "id": "6266df79",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Save test\n",
    "with open(\"PCBP1/test_split.pkl\", \"wb\") as f:\n",
    "    pickle.dump((test_data, test_labels), f)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "id": "95900830",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Save positive label\n",
    "with open(\"PCBP1/PCBP1_positive_label.pkl\", \"wb\") as f:\n",
    "    pickle.dump(positive_value, f)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "id": "fe972c58",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Save negative label\n",
    "with open(\"PCBP1/PCBP1_negative_label.pkl\", \"wb\") as f:\n",
    "    pickle.dump(negative_value, f)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "521fb44e",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "perturb",
   "language": "python",
   "name": "perturb"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
