{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "3ee91ff5",
   "metadata": {},
   "outputs": [],
   "source": [
    "%load_ext autoreload\n",
    "%autoreload 2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "15865688",
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ff5d5753",
   "metadata": {},
   "source": [
    "## Read data"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "235e3bf2",
   "metadata": {},
   "source": [
    "Load eCLIP-seq sequences"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "abe42208",
   "metadata": {},
   "outputs": [],
   "source": [
    "eclip = pd.read_csv(\"SRSF9/SRSF9_eclip.txt\", sep='\\t', header=None)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "32e0bc25",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr7:7222551-7222679</td>\n",
       "      <td>CAAACCCCGGGCGCGGCGGAACAGGTGTCTCGGGGTCAAAGCCGGG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr7:138145159-138145275</td>\n",
       "      <td>TCCCTCCCTCGCTGGCGCTGCCGCGAGTCCACCGAGCGGCCTCTGA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr7:1025772-1025833</td>\n",
       "      <td>TCACTACTGGCGATCGCGGACGCCCCAGGAAGGCGAGTGGCACGAG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr7:150767028-150767100</td>\n",
       "      <td>GCTAGGGGCTGTGCTCTTGGATCTGGACCAGCAGACCCTGCCCGGA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr7:65551730-65551807</td>\n",
       "      <td>CCACGCCGTGGCACTGACCCGAGACTCTGAGCGGCTGCTGGAGGTG...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                          0                                                  1\n",
       "0      chr7:7222551-7222679  CAAACCCCGGGCGCGGCGGAACAGGTGTCTCGGGGTCAAAGCCGGG...\n",
       "1  chr7:138145159-138145275  TCCCTCCCTCGCTGGCGCTGCCGCGAGTCCACCGAGCGGCCTCTGA...\n",
       "2      chr7:1025772-1025833  TCACTACTGGCGATCGCGGACGCCCCAGGAAGGCGAGTGGCACGAG...\n",
       "3  chr7:150767028-150767100  GCTAGGGGCTGTGCTCTTGGATCTGGACCAGCAGACCCTGCCCGGA...\n",
       "4    chr7:65551730-65551807  CCACGCCGTGGCACTGACCCGAGACTCTGAGCGGCTGCTGGAGGTG..."
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "eclip.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8e72135a",
   "metadata": {},
   "source": [
    "Load RBNS binding affinities"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "7bc981f7",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>[SRSF9]</th>\n",
       "      <th>5 nM</th>\n",
       "      <th>20 nM</th>\n",
       "      <th>80 nM</th>\n",
       "      <th>320 nM</th>\n",
       "      <th>1300 nM</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GGGG</td>\n",
       "      <td>1.0975</td>\n",
       "      <td>1.1679</td>\n",
       "      <td>1.2271</td>\n",
       "      <td>1.2953</td>\n",
       "      <td>1.7139</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GGAG</td>\n",
       "      <td>1.0840</td>\n",
       "      <td>1.2955</td>\n",
       "      <td>1.4331</td>\n",
       "      <td>1.6534</td>\n",
       "      <td>1.6673</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>GGGA</td>\n",
       "      <td>1.0948</td>\n",
       "      <td>1.2341</td>\n",
       "      <td>1.3277</td>\n",
       "      <td>1.4853</td>\n",
       "      <td>1.5885</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AGGG</td>\n",
       "      <td>1.0835</td>\n",
       "      <td>1.2028</td>\n",
       "      <td>1.2505</td>\n",
       "      <td>1.3731</td>\n",
       "      <td>1.5563</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>GAGG</td>\n",
       "      <td>1.0562</td>\n",
       "      <td>1.1583</td>\n",
       "      <td>1.2415</td>\n",
       "      <td>1.3768</td>\n",
       "      <td>1.5015</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>251</th>\n",
       "      <td>TTAA</td>\n",
       "      <td>0.9898</td>\n",
       "      <td>0.9762</td>\n",
       "      <td>0.9261</td>\n",
       "      <td>0.9024</td>\n",
       "      <td>0.6550</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>252</th>\n",
       "      <td>TATT</td>\n",
       "      <td>0.9665</td>\n",
       "      <td>0.9633</td>\n",
       "      <td>0.9140</td>\n",
       "      <td>0.8942</td>\n",
       "      <td>0.6303</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>253</th>\n",
       "      <td>TTTT</td>\n",
       "      <td>0.9425</td>\n",
       "      <td>0.8991</td>\n",
       "      <td>0.8394</td>\n",
       "      <td>0.7928</td>\n",
       "      <td>0.6199</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>254</th>\n",
       "      <td>TTTA</td>\n",
       "      <td>0.9502</td>\n",
       "      <td>0.9413</td>\n",
       "      <td>0.8936</td>\n",
       "      <td>0.8685</td>\n",
       "      <td>0.6146</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>255</th>\n",
       "      <td>TTAT</td>\n",
       "      <td>0.9498</td>\n",
       "      <td>0.9534</td>\n",
       "      <td>0.9017</td>\n",
       "      <td>0.8789</td>\n",
       "      <td>0.6093</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>256 rows Ã— 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    [SRSF9]    5 nM   20 nM   80 nM  320 nM  1300 nM\n",
       "0      GGGG  1.0975  1.1679  1.2271  1.2953   1.7139\n",
       "1      GGAG  1.0840  1.2955  1.4331  1.6534   1.6673\n",
       "2      GGGA  1.0948  1.2341  1.3277  1.4853   1.5885\n",
       "3      AGGG  1.0835  1.2028  1.2505  1.3731   1.5563\n",
       "4      GAGG  1.0562  1.1583  1.2415  1.3768   1.5015\n",
       "..      ...     ...     ...     ...     ...      ...\n",
       "251    TTAA  0.9898  0.9762  0.9261  0.9024   0.6550\n",
       "252    TATT  0.9665  0.9633  0.9140  0.8942   0.6303\n",
       "253    TTTT  0.9425  0.8991  0.8394  0.7928   0.6199\n",
       "254    TTTA  0.9502  0.9413  0.8936  0.8685   0.6146\n",
       "255    TTAT  0.9498  0.9534  0.9017  0.8789   0.6093\n",
       "\n",
       "[256 rows x 6 columns]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "four_mers = pd.read_csv(\"SRSF9/SRSF9_4.tsv\", sep='\\t')\n",
    "five_mers = pd.read_csv(\"SRSF9/SRSF9_5.tsv\", sep='\\t',)\n",
    "six_mers = pd.read_csv(\"SRSF9/SRSF9_6.tsv\", sep='\\t',)\n",
    "four_mers"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "af6ffd4e",
   "metadata": {},
   "source": [
    "Load knockdown differential rna-seq data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "d0e331a6",
   "metadata": {},
   "outputs": [],
   "source": [
    "# get other diff expr for subsetting to shared genes later\n",
    "diff_exp_HNRNPC = pd.read_csv(\"HNRNPC/HNRNPC.tsv\", sep='\\t',)\n",
    "diff_exp_SRSF9 = pd.read_csv(\"SRSF9/SRSF9.tsv\", sep='\\t',)\n",
    "diff_exp_PCBP1 = pd.read_csv(\"PCBP1/PCBP1.tsv\", sep='\\t',)\n",
    "diff_exp_TIA1 = pd.read_csv(\"TIA1/TIA1.tsv\", sep='\\t',)\n",
    "diff_exp_TRA2A = pd.read_csv(\"TRA2A/TRA2A.tsv\", sep='\\t',)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1612ca75",
   "metadata": {},
   "source": [
    "## Generate positive pairs"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b0787ccc",
   "metadata": {},
   "source": [
    "EClip positive binding sites"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "592b1f36",
   "metadata": {},
   "outputs": [],
   "source": [
    "in_vivo_pos_seqs = eclip[1].values"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "18057862",
   "metadata": {},
   "outputs": [],
   "source": [
    "# filter for length and unique values\n",
    "in_vivo_pos_seqs = [s for s in in_vivo_pos_seqs if len(s) >= 10]\n",
    "in_vivo_pos_seqs = list(set(in_vivo_pos_seqs))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "adf61072",
   "metadata": {},
   "source": [
    "RBNS significant binding affinity"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "fd120254",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_signficant_sequences(df):\n",
    "    num_cols = df.columns[1:]\n",
    "    filtered = df[df[num_cols].gt(2).any(axis=1)]\n",
    "    pairs = [\n",
    "        (row[df.columns[0]], tuple(row[num_cols]))\n",
    "        for _, row in filtered.iterrows()\n",
    "    ]\n",
    "    return pairs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "ed65dc0f",
   "metadata": {},
   "outputs": [],
   "source": [
    "pos_four_mers = get_signficant_sequences(four_mers)\n",
    "pos_five_mers = get_signficant_sequences(five_mers)\n",
    "pos_six_mers = get_signficant_sequences(six_mers)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "1b63d4fe",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(0, 1, 120)"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(pos_four_mers), len(pos_five_mers), len(pos_six_mers), "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "6894c643",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "<__array_function__ internals>:180: VisibleDeprecationWarning: Creating an ndarray from ragged nested sequences (which is a list-or-tuple of lists-or-tuples-or ndarrays with different lengths or shapes) is deprecated. If you meant to do this, you must specify 'dtype=object' when creating the ndarray.\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "(121, 2)"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "in_vitro_pos_seqs = np.concatenate((pos_five_mers,pos_six_mers))\n",
    "in_vitro_pos_seqs.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "3bc23f33",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "121"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "unique_rbns = {}\n",
    "for s, arr in in_vitro_pos_seqs:\n",
    "    if s not in unique_rbns:\n",
    "        unique_rbns[s] = arr\n",
    "\n",
    "in_vitro_pos_seqs = list(unique_rbns.items())\n",
    "len(in_vitro_pos_seqs)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6ad2eb50",
   "metadata": {},
   "source": [
    "Filter diff expr"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "7a6b7538",
   "metadata": {},
   "outputs": [],
   "source": [
    "# remove when gene expression is always 0 or if logfold change is inf/-inf\n",
    "diff_exp_filtered_HNRNPC = diff_exp_HNRNPC[\n",
    "    ~(\n",
    "        ((diff_exp_HNRNPC[\"value_1\"] == 0) & (diff_exp_HNRNPC[\"value_2\"] == 0)) |\n",
    "        (diff_exp_HNRNPC[\"log2(fold_change)\"].isin([np.inf, -np.inf]))\n",
    "    )\n",
    "]\n",
    "\n",
    "diff_exp_filtered_SRSF9 = diff_exp_SRSF9[\n",
    "    ~(\n",
    "        ((diff_exp_SRSF9[\"value_1\"] == 0) & (diff_exp_SRSF9[\"value_2\"] == 0)) |\n",
    "        (diff_exp_SRSF9[\"log2(fold_change)\"].isin([np.inf, -np.inf]))\n",
    "    )\n",
    "]\n",
    "\n",
    "diff_exp_filtered_PCBP1 = diff_exp_PCBP1[\n",
    "    ~(\n",
    "        ((diff_exp_PCBP1[\"value_1\"] == 0) & (diff_exp_PCBP1[\"value_2\"] == 0)) |\n",
    "        (diff_exp_PCBP1[\"log2(fold_change)\"].isin([np.inf, -np.inf]))\n",
    "    )\n",
    "]\n",
    "\n",
    "diff_exp_filtered_TIA1 = diff_exp_TIA1[\n",
    "    ~(\n",
    "        ((diff_exp_TIA1[\"value_1\"] == 0) & (diff_exp_TIA1[\"value_2\"] == 0)) |\n",
    "        (diff_exp_TIA1[\"log2(fold_change)\"].isin([np.inf, -np.inf]))\n",
    "    )\n",
    "]\n",
    "\n",
    "diff_exp_filtered_TRA2A = diff_exp_TRA2A[\n",
    "    ~(\n",
    "        ((diff_exp_TRA2A[\"value_1\"] == 0) & (diff_exp_TRA2A[\"value_2\"] == 0)) |\n",
    "        (diff_exp_TRA2A[\"log2(fold_change)\"].isin([np.inf, -np.inf]))\n",
    "    )\n",
    "]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "377252f2",
   "metadata": {},
   "outputs": [],
   "source": [
    "dfs = [diff_exp_filtered_HNRNPC, diff_exp_filtered_SRSF9, diff_exp_filtered_PCBP1, diff_exp_filtered_TIA1, diff_exp_filtered_TRA2A]\n",
    "col = \"gene_id\"\n",
    "shared_keys = set.intersection(*(set(df[col]) for df in dfs))\n",
    "\n",
    "# 2. Subset the dataframe you want (example: df1)\n",
    "diff_exp_filtered_SRSF9 = diff_exp_filtered_SRSF9[diff_exp_filtered_SRSF9[col].isin(shared_keys)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "fdeaa9e5",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>test_id</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene</th>\n",
       "      <th>locus</th>\n",
       "      <th>sample_1</th>\n",
       "      <th>sample_2</th>\n",
       "      <th>status</th>\n",
       "      <th>value_1</th>\n",
       "      <th>value_2</th>\n",
       "      <th>log2(fold_change)</th>\n",
       "      <th>test_stat</th>\n",
       "      <th>p_value</th>\n",
       "      <th>q_value</th>\n",
       "      <th>significant</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>XLOC_000008</td>\n",
       "      <td>XLOC_000008</td>\n",
       "      <td>OR4F29</td>\n",
       "      <td>chr1:317719-461954</td>\n",
       "      <td>SRSF9-BGHLV12-HepG2</td>\n",
       "      <td>Control</td>\n",
       "      <td>OK</td>\n",
       "      <td>0.626629</td>\n",
       "      <td>0.248264</td>\n",
       "      <td>-1.335730</td>\n",
       "      <td>-0.528539</td>\n",
       "      <td>0.03450</td>\n",
       "      <td>0.088801</td>\n",
       "      <td>no</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>XLOC_000012</td>\n",
       "      <td>XLOC_000012</td>\n",
       "      <td>MTND1P23</td>\n",
       "      <td>chr1:536815-660283</td>\n",
       "      <td>SRSF9-BGHLV12-HepG2</td>\n",
       "      <td>Control</td>\n",
       "      <td>OK</td>\n",
       "      <td>170.655000</td>\n",
       "      <td>189.246000</td>\n",
       "      <td>0.149173</td>\n",
       "      <td>0.074780</td>\n",
       "      <td>0.89060</td>\n",
       "      <td>0.932908</td>\n",
       "      <td>no</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>XLOC_000013</td>\n",
       "      <td>XLOC_000013</td>\n",
       "      <td>MTND2P28</td>\n",
       "      <td>chr1:536815-660283</td>\n",
       "      <td>SRSF9-BGHLV12-HepG2</td>\n",
       "      <td>Control</td>\n",
       "      <td>OK</td>\n",
       "      <td>513.752000</td>\n",
       "      <td>652.723000</td>\n",
       "      <td>0.345401</td>\n",
       "      <td>0.693741</td>\n",
       "      <td>0.32070</td>\n",
       "      <td>0.478673</td>\n",
       "      <td>no</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>XLOC_000014</td>\n",
       "      <td>XLOC_000014</td>\n",
       "      <td>hsa-mir-6723</td>\n",
       "      <td>chr1:536815-660283</td>\n",
       "      <td>SRSF9-BGHLV12-HepG2</td>\n",
       "      <td>Control</td>\n",
       "      <td>OK</td>\n",
       "      <td>1090.390000</td>\n",
       "      <td>1230.730000</td>\n",
       "      <td>0.174662</td>\n",
       "      <td>0.626937</td>\n",
       "      <td>0.36875</td>\n",
       "      <td>0.528346</td>\n",
       "      <td>no</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>XLOC_000015</td>\n",
       "      <td>XLOC_000015</td>\n",
       "      <td>RP5-857K21.7</td>\n",
       "      <td>chr1:536815-660283</td>\n",
       "      <td>SRSF9-BGHLV12-HepG2</td>\n",
       "      <td>Control</td>\n",
       "      <td>OK</td>\n",
       "      <td>538.127000</td>\n",
       "      <td>636.520000</td>\n",
       "      <td>0.242260</td>\n",
       "      <td>0.359909</td>\n",
       "      <td>0.60395</td>\n",
       "      <td>0.735486</td>\n",
       "      <td>no</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        test_id      gene_id          gene               locus  \\\n",
       "7   XLOC_000008  XLOC_000008        OR4F29  chr1:317719-461954   \n",
       "11  XLOC_000012  XLOC_000012      MTND1P23  chr1:536815-660283   \n",
       "12  XLOC_000013  XLOC_000013      MTND2P28  chr1:536815-660283   \n",
       "13  XLOC_000014  XLOC_000014  hsa-mir-6723  chr1:536815-660283   \n",
       "14  XLOC_000015  XLOC_000015  RP5-857K21.7  chr1:536815-660283   \n",
       "\n",
       "               sample_1 sample_2 status      value_1      value_2  \\\n",
       "7   SRSF9-BGHLV12-HepG2  Control     OK     0.626629     0.248264   \n",
       "11  SRSF9-BGHLV12-HepG2  Control     OK   170.655000   189.246000   \n",
       "12  SRSF9-BGHLV12-HepG2  Control     OK   513.752000   652.723000   \n",
       "13  SRSF9-BGHLV12-HepG2  Control     OK  1090.390000  1230.730000   \n",
       "14  SRSF9-BGHLV12-HepG2  Control     OK   538.127000   636.520000   \n",
       "\n",
       "    log2(fold_change)  test_stat  p_value   q_value significant  \n",
       "7           -1.335730  -0.528539  0.03450  0.088801          no  \n",
       "11           0.149173   0.074780  0.89060  0.932908          no  \n",
       "12           0.345401   0.693741  0.32070  0.478673          no  \n",
       "13           0.174662   0.626937  0.36875  0.528346          no  \n",
       "14           0.242260   0.359909  0.60395  0.735486          no  "
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "diff_exp_filtered_SRSF9.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "1046db6c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(17985, 14)"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "diff_exp_filtered_SRSF9.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "d8fe4752",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([-1.33573 ,  0.149173,  0.345401, ..., -1.01329 , -0.139774,\n",
       "       -1.51089 ])"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "logfold_changes = diff_exp_filtered_SRSF9['log2(fold_change)'].values\n",
    "logfold_changes"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "14a14525",
   "metadata": {},
   "source": [
    "## Generate negative pairs"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "eaa775d1",
   "metadata": {},
   "source": [
    "Shuffle eCLIP sequences"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "68b0a794",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "128774"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import random\n",
    "def shuffle_by_pairs(seq):\n",
    "    pairs = [seq[i:i+2] for i in range(0, len(seq), 2)]\n",
    "    random.shuffle(pairs)\n",
    "    return \"\".join(pairs)\n",
    "in_vivo_neg_seqs = [shuffle_by_pairs(seq) for seq in eclip[1].values]\n",
    "len(in_vivo_neg_seqs)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "1ec9c58c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "123719"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# filter for length and unique values\n",
    "in_vivo_neg_seqs = [s for s in in_vivo_neg_seqs if len(s) >= 10]\n",
    "in_vivo_neg_seqs = list(set(in_vivo_neg_seqs))\n",
    "len(in_vivo_neg_seqs)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "02a5e383",
   "metadata": {},
   "source": [
    "Retrieve RBNS sequences that are not enriched"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "669ff6f6",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_insignficant_sequences(df):\n",
    "    num_cols = df.columns[1:]\n",
    "    filtered = df[df[num_cols].lt(2).any(axis=1)]\n",
    "    pairs = [\n",
    "        (row[df.columns[0]], tuple(row[num_cols]))\n",
    "        for _, row in filtered.iterrows()\n",
    "    ]\n",
    "    return pairs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "4a0c3f24",
   "metadata": {},
   "outputs": [],
   "source": [
    "neg_four_mers = get_insignficant_sequences(four_mers)\n",
    "neg_five_mers = get_insignficant_sequences(five_mers)\n",
    "neg_six_mers = get_insignficant_sequences(six_mers)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "7920a3d9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[('GGGG', (1.0975, 1.1679, 1.2271, 1.2953, 1.7139)),\n",
       " ('GGAG', (1.084, 1.2955, 1.4331, 1.6534, 1.6673)),\n",
       " ('GGGA', (1.0948, 1.2341, 1.3277, 1.4853, 1.5885)),\n",
       " ('AGGG', (1.0835, 1.2028, 1.2505, 1.3731, 1.5563)),\n",
       " ('GAGG', (1.0562, 1.1583, 1.2415, 1.3768, 1.5015)),\n",
       " ('AGGA', (1.0795, 1.2664, 1.3685, 1.5765, 1.4914)),\n",
       " ('TGGG', (1.0713, 1.1198, 1.175, 1.2301, 1.4846)),\n",
       " ('GGGC', (1.0371, 1.0663, 1.1299, 1.123, 1.4796)),\n",
       " ('GGAA', (1.088, 1.2584, 1.3455, 1.554, 1.4778)),\n",
       " ('CAGG', (1.0332, 1.1257, 1.1791, 1.2627, 1.4751)),\n",
       " ('GGCC', (1.0099, 1.0264, 1.1019, 1.0934, 1.4428)),\n",
       " ('GGTG', (1.0333, 1.0965, 1.1215, 1.1694, 1.4261)),\n",
       " ('GGGT', (1.0623, 1.1022, 1.1354, 1.1614, 1.4242)),\n",
       " ('CGGG', (1.0429, 1.0561, 1.1317, 1.1167, 1.4201)),\n",
       " ('GTGG', (1.0395, 1.0887, 1.1268, 1.1691, 1.4162)),\n",
       " ('GCGC', (0.9708, 1.0122, 1.0704, 1.0824, 1.402)),\n",
       " ('GCGG', (1.0052, 1.0331, 1.0877, 1.0612, 1.3789)),\n",
       " ('GCAG', (1.0121, 1.0756, 1.1213, 1.1783, 1.3773)),\n",
       " ('GGCA', (1.0079, 1.0643, 1.1165, 1.1847, 1.3691)),\n",
       " ('GGCG', (0.9933, 1.0185, 1.0641, 1.0559, 1.3636)),\n",
       " ('AGGC', (1.019, 1.0636, 1.1012, 1.1613, 1.3471)),\n",
       " ('CTGG', (1.0404, 1.058, 1.1127, 1.1194, 1.3443)),\n",
       " ('ATGG', (1.0393, 1.1261, 1.1702, 1.2756, 1.3205)),\n",
       " ('TGGA', (1.0504, 1.148, 1.2231, 1.3291, 1.3174)),\n",
       " ('TGGC', (0.9984, 1.0183, 1.0623, 1.0733, 1.3169)),\n",
       " ('GTGC', (0.9911, 1.0417, 1.0625, 1.101, 1.313)),\n",
       " ('AAGG', (1.0567, 1.1469, 1.1788, 1.2872, 1.3048)),\n",
       " ('GCCA', (0.9924, 1.0476, 1.082, 1.1197, 1.3026)),\n",
       " ('GAGC', (1.0092, 1.0661, 1.124, 1.1806, 1.2968)),\n",
       " ('GCCC', (0.9985, 0.9885, 1.0584, 1.0496, 1.2932)),\n",
       " ('GAGA', (1.0645, 1.1138, 1.1837, 1.3117, 1.2813)),\n",
       " ('GCAC', (0.975, 1.021, 1.0615, 1.1123, 1.2777)),\n",
       " ('GCTG', (1.0094, 1.0279, 1.068, 1.0674, 1.2766)),\n",
       " ('CCGG', (1.0443, 1.0051, 1.0733, 1.0012, 1.2714)),\n",
       " ('CAGC', (0.9805, 1.0231, 1.0489, 1.0782, 1.2675)),\n",
       " ('AGAG', (1.072, 1.1162, 1.1493, 1.2443, 1.2612)),\n",
       " ('GAAG', (1.05, 1.1026, 1.1475, 1.2484, 1.2462)),\n",
       " ('CCAG', (1.019, 1.0245, 1.0507, 1.0454, 1.236)),\n",
       " ('GCCG', (0.9821, 0.9605, 1.0187, 0.9522, 1.2342)),\n",
       " ('GGAT', (1.0533, 1.1056, 1.157, 1.2107, 1.2297)),\n",
       " ('CCGC', (0.9813, 0.9696, 1.0196, 0.9692, 1.2251)),\n",
       " ('AGCG', (0.9936, 1.0245, 1.0474, 1.0647, 1.2207)),\n",
       " ('GGAC', (1.02, 1.0682, 1.1236, 1.127, 1.2191)),\n",
       " ('TGCG', (0.9862, 1.0016, 1.0222, 1.0214, 1.2188)),\n",
       " ('CGCG', (0.965, 0.9565, 0.9931, 0.9463, 1.2186)),\n",
       " ('ACGG', (1.0135, 1.0304, 1.0667, 1.0666, 1.2153)),\n",
       " ('CGGC', (0.9753, 0.9577, 1.0112, 0.9521, 1.215)),\n",
       " ('GATG', (1.013, 1.0647, 1.1027, 1.19, 1.2147)),\n",
       " ('AGCA', (0.9965, 1.0827, 1.0985, 1.2019, 1.2115)),\n",
       " ('GGCT', (1.0097, 1.0258, 1.0598, 1.0508, 1.2102)),\n",
       " ('AGCC', (0.9967, 1.0203, 1.0617, 1.0785, 1.2088)),\n",
       " ('TAGG', (1.0521, 1.1071, 1.1389, 1.2007, 1.2055)),\n",
       " ('GGTA', (1.0383, 1.077, 1.0929, 1.1429, 1.2013)),\n",
       " ('CGCA', (0.9645, 0.9971, 1.0201, 1.0324, 1.1977)),\n",
       " ('TGCC', (0.9773, 0.9839, 1.0151, 1.0236, 1.1962)),\n",
       " ('CTGC', (0.9931, 0.9927, 1.0162, 1.0016, 1.1959)),\n",
       " ('GCGA', (0.9979, 1.0073, 1.0358, 1.0412, 1.1855)),\n",
       " ('CACC', (0.975, 0.9686, 1.0268, 1.0591, 1.1811)),\n",
       " ('CCAC', (0.997, 1.0193, 1.054, 1.0468, 1.1794)),\n",
       " ('CATG', (0.9926, 1.0332, 1.032, 1.0833, 1.1741)),\n",
       " ('CGGA', (1.0318, 1.0497, 1.0978, 1.0745, 1.1716)),\n",
       " ('CACG', (0.9734, 0.9845, 0.9989, 0.9939, 1.1713)),\n",
       " ('AGTG', (1.031, 1.0594, 1.0614, 1.1188, 1.1634)),\n",
       " ('GCAA', (1.0001, 1.0531, 1.0582, 1.119, 1.1554)),\n",
       " ('TGCA', (0.9833, 1.0377, 1.0381, 1.0988, 1.1554)),\n",
       " ('ACAG', (1.0115, 1.0679, 1.0801, 1.1305, 1.1553)),\n",
       " ('CCCC', (1.0547, 1.0075, 1.0545, 0.9879, 1.1547)),\n",
       " ('ATGC', (0.9912, 1.0463, 1.0506, 1.1233, 1.1508)),\n",
       " ('TGAG', (1.02, 1.0362, 1.0548, 1.1101, 1.1499)),\n",
       " ('GTAG', (1.0388, 1.085, 1.1092, 1.1408, 1.1497)),\n",
       " ('CGTG', (0.9793, 0.9705, 0.9796, 0.9597, 1.1486)),\n",
       " ('CCTG', (1.0044, 0.9694, 1.0042, 0.9746, 1.148)),\n",
       " ('ACGC', (0.9642, 0.9904, 1.0083, 1.0236, 1.1474)),\n",
       " ('CCCG', (1.0216, 0.9391, 1.0296, 0.9881, 1.1467)),\n",
       " ('AGGT', (1.0143, 1.0332, 1.0416, 1.0723, 1.1444)),\n",
       " ('GCCT', (0.9599, 0.9459, 0.99, 0.9678, 1.1443)),\n",
       " ('CGCC', (0.9507, 0.9085, 0.9574, 0.8926, 1.1437)),\n",
       " ('CAGA', (1.0359, 1.0363, 1.0531, 1.0571, 1.1367)),\n",
       " ('TGTG', (1.0069, 1.0262, 1.0256, 1.0507, 1.1364)),\n",
       " ('CGAG', (0.9908, 0.9702, 1.0376, 1.0612, 1.1339)),\n",
       " ('AGAA', (1.093, 1.14, 1.1685, 1.2574, 1.1337)),\n",
       " ('GTGA', (1.0267, 1.0339, 1.035, 1.0685, 1.1251)),\n",
       " ('CACA', (0.9778, 1.0287, 1.0291, 1.0813, 1.118)),\n",
       " ('GAAC', (1.0321, 1.0721, 1.1015, 1.127, 1.1157)),\n",
       " ('ACTG', (1.0247, 1.0408, 1.0609, 1.0633, 1.115)),\n",
       " ('CAAG', (1.0067, 1.021, 1.0234, 1.0433, 1.1143)),\n",
       " ('AAGC', (1.0051, 1.0316, 1.0443, 1.0902, 1.1125)),\n",
       " ('GCAT', (0.9761, 1.029, 1.0301, 1.0918, 1.1121)),\n",
       " ('GCGT', (0.9667, 0.9717, 0.9754, 0.9571, 1.1121)),\n",
       " ('TGGT', (1.0033, 0.9992, 0.9934, 0.9839, 1.1088)),\n",
       " ('CGGT', (0.9977, 0.9698, 0.9879, 0.9233, 1.1076)),\n",
       " ('GACG', (0.9838, 0.9806, 1.0134, 1.0014, 1.1067)),\n",
       " ('GACA', (1.0002, 1.0196, 1.0495, 1.0957, 1.0971)),\n",
       " ('GACC', (0.9975, 0.9647, 0.9984, 0.9506, 1.0922)),\n",
       " ('GAGT', (1.0199, 1.0489, 1.0812, 1.1391, 1.092)),\n",
       " ('ACCG', (1.004, 0.9655, 0.9924, 0.9403, 1.0912)),\n",
       " ('ACCA', (1.0037, 1.0303, 1.0279, 1.052, 1.0904)),\n",
       " ('ACAC', (0.985, 1.0285, 1.0531, 1.1192, 1.0904)),\n",
       " ('GCTA', (1.0026, 1.0305, 1.0355, 1.0731, 1.0868)),\n",
       " ('GAAA', (1.093, 1.1121, 1.113, 1.1827, 1.0844)),\n",
       " ('ACCC', (1.0252, 0.9821, 1.0342, 1.0261, 1.0843)),\n",
       " ('TTGG', (1.0029, 0.9858, 1.0093, 0.987, 1.0815)),\n",
       " ('AGAC', (1.0288, 1.0175, 1.0354, 1.034, 1.0732)),\n",
       " ('CGCT', (0.9554, 0.9462, 0.9581, 0.9228, 1.073)),\n",
       " ('TAGC', (0.9997, 1.0344, 1.0317, 1.0724, 1.0651)),\n",
       " ('GTGT', (0.995, 1.0125, 0.9984, 1.0109, 1.0641)),\n",
       " ('GGTC', (0.9892, 0.9403, 0.9469, 0.8561, 1.0615)),\n",
       " ('TGCT', (0.9788, 0.987, 0.9803, 0.9837, 1.0594)),\n",
       " ('CTAG', (1.0289, 1.0174, 1.0223, 1.0205, 1.0566)),\n",
       " ('GGTT', (1.0003, 0.9801, 0.9849, 0.9599, 1.0542)),\n",
       " ('AAGA', (1.0721, 1.0696, 1.0667, 1.1065, 1.0465)),\n",
       " ('AGCT', (0.9962, 0.9942, 0.9996, 1.0148, 1.0401)),\n",
       " ('GCTC', (0.9709, 0.9405, 0.9451, 0.8875, 1.0385)),\n",
       " ('CCCA', (1.0075, 0.9821, 0.9942, 0.955, 1.0378)),\n",
       " ('AAAG', (1.0712, 1.0749, 1.0613, 1.0932, 1.0341)),\n",
       " ('GTAC', (1.0046, 1.0062, 1.0042, 1.0035, 1.0297)),\n",
       " ('AACC', (1.0247, 1.0096, 1.0103, 1.0038, 1.0166)),\n",
       " ('CAAC', (0.9709, 0.9805, 0.9667, 0.9663, 1.014)),\n",
       " ('CTGT', (1.0022, 0.9899, 0.9978, 0.9655, 1.0132)),\n",
       " ('AACA', (1.0094, 1.0753, 1.0591, 1.1237, 1.0105)),\n",
       " ('CCGA', (1.0011, 0.9309, 0.9848, 0.9451, 1.0104)),\n",
       " ('TTGC', (0.9595, 0.9442, 0.945, 0.927, 1.0103)),\n",
       " ('AATG', (1.0162, 1.0519, 1.0455, 1.1052, 1.0099)),\n",
       " ('CTGA', (1.0147, 0.9792, 0.9843, 0.9509, 1.0097)),\n",
       " ('ATGA', (1.0241, 1.0386, 1.0365, 1.0986, 1.0085)),\n",
       " ('GTCC', (0.9727, 0.928, 0.9369, 0.8602, 1.0071)),\n",
       " ('TAGA', (1.064, 1.0848, 1.0988, 1.1278, 1.0071)),\n",
       " ('CCTC', (0.9771, 0.9276, 0.9315, 0.8684, 0.9991)),\n",
       " ('CCAA', (1.0091, 0.997, 0.9863, 0.969, 0.9991)),\n",
       " ('CAGT', (0.977, 0.985, 0.9703, 0.9599, 0.999)),\n",
       " ('CACT', (0.9776, 0.9816, 0.9792, 0.9807, 0.997)),\n",
       " ('CCCT', (0.9871, 0.9423, 0.9602, 0.9036, 0.9966)),\n",
       " ('TGAC', (0.9822, 0.9508, 0.9613, 0.9518, 0.9952)),\n",
       " ('GTTG', (0.9714, 0.9388, 0.9391, 0.8877, 0.9866)),\n",
       " ('CCAT', (0.9741, 0.979, 0.9681, 0.9682, 0.9862)),\n",
       " ('CGAC', (0.9715, 0.9127, 0.9378, 0.873, 0.986)),\n",
       " ('GAAT', (1.0392, 1.0605, 1.0734, 1.1274, 0.9859)),\n",
       " ('ATAG', (1.0499, 1.087, 1.0833, 1.1313, 0.984)),\n",
       " ('ACGA', (0.997, 0.9812, 0.9841, 0.9768, 0.9835)),\n",
       " ('GACT', (1.008, 0.9889, 1.01, 0.9789, 0.9802)),\n",
       " ('TACC', (1.0053, 0.9914, 0.9879, 0.9846, 0.979)),\n",
       " ('AACG', (0.994, 0.9882, 0.9787, 0.9651, 0.9779)),\n",
       " ('TCAG', (0.9778, 0.9574, 0.9374, 0.8943, 0.975)),\n",
       " ('GCTT', (0.9578, 0.952, 0.9491, 0.9368, 0.9749)),\n",
       " ('CAAA', (1.0323, 1.0382, 1.0078, 1.02, 0.9747)),\n",
       " ('ACAA', (1.0089, 1.047, 1.026, 1.0676, 0.9744)),\n",
       " ('TCCA', (0.9868, 0.9593, 0.9517, 0.9174, 0.9742)),\n",
       " ('CTTG', (0.9642, 0.9142, 0.9308, 0.8915, 0.9731)),\n",
       " ('CGAA', (1.0152, 0.9774, 0.9788, 0.9493, 0.9676)),\n",
       " ('TCGG', (0.9875, 0.9354, 0.9548, 0.8505, 0.9648)),\n",
       " ('CCGT', (0.9721, 0.9076, 0.9212, 0.8387, 0.964)),\n",
       " ('AGAT', (1.0325, 1.0083, 1.0092, 1.0167, 0.9623)),\n",
       " ('CTAC', (0.989, 0.9752, 0.9672, 0.9572, 0.9613)),\n",
       " ('ACCT', (0.9807, 0.9475, 0.9516, 0.9365, 0.9591)),\n",
       " ('GATA', (1.0316, 1.0441, 1.0418, 1.0798, 0.9576)),\n",
       " ('TCCG', (0.9845, 0.9199, 0.9325, 0.8485, 0.9571)),\n",
       " ('TGAA', (1.042, 1.0279, 1.0179, 1.0353, 0.954)),\n",
       " ('TCTG', (1.0052, 0.9451, 0.9339, 0.8633, 0.9524)),\n",
       " ('CTCA', (0.967, 0.9511, 0.9277, 0.8934, 0.9521)),\n",
       " ('CATC', (0.952, 0.925, 0.9142, 0.8907, 0.9521)),\n",
       " ('AAAA', (1.1411, 1.1404, 1.0922, 1.1327, 0.9502)),\n",
       " ('AAAC', (1.0434, 1.0446, 1.0148, 1.0283, 0.9483)),\n",
       " ('GATC', (0.9996, 0.927, 0.926, 0.8417, 0.9443)),\n",
       " ('TCAC', (0.9575, 0.9393, 0.9191, 0.8856, 0.9405)),\n",
       " ('GTCA', (0.9609, 0.9447, 0.9234, 0.8768, 0.9397)),\n",
       " ('TACG', (0.9869, 0.9766, 0.9641, 0.9496, 0.9358)),\n",
       " ('CTCG', (0.9748, 0.894, 0.9013, 0.8018, 0.9326)),\n",
       " ('ATCC', (0.9958, 0.9459, 0.9405, 0.8961, 0.9214)),\n",
       " ('TCGC', (0.9303, 0.8724, 0.8726, 0.7699, 0.9177)),\n",
       " ('TATG', (1.0091, 1.0296, 1.0089, 1.0411, 0.9122)),\n",
       " ('AGTA', (1.0196, 1.051, 1.0367, 1.0812, 0.9106)),\n",
       " ('TGTA', (1.0041, 1.0236, 1.0132, 1.0233, 0.9088)),\n",
       " ('CGTA', (0.9795, 0.9515, 0.9406, 0.8983, 0.9084)),\n",
       " ('CGAT', (0.976, 0.924, 0.9201, 0.8686, 0.9056)),\n",
       " ('CCTA', (0.9872, 0.9572, 0.943, 0.9104, 0.9037)),\n",
       " ('ATGT', (0.9771, 0.9993, 0.9721, 0.9946, 0.8961)),\n",
       " ('ACGT', (0.9567, 0.9408, 0.9167, 0.8791, 0.8951)),\n",
       " ('GTAA', (1.0223, 1.0275, 1.0085, 1.0247, 0.8936)),\n",
       " ('TAAG', (1.0243, 1.0257, 1.0075, 1.0223, 0.8935)),\n",
       " ('ACAT', (0.9661, 1.0072, 0.9843, 1.0304, 0.8901)),\n",
       " ('TGTC', (0.9572, 0.916, 0.8952, 0.8319, 0.8897)),\n",
       " ('TGAT', (0.9927, 0.9579, 0.9366, 0.919, 0.8893)),\n",
       " ('ATAC', (1.0129, 1.0371, 1.0209, 1.0574, 0.8846)),\n",
       " ('TACA', (0.9797, 1.0241, 1.0058, 1.0544, 0.8826)),\n",
       " ('ATTG', (0.9915, 0.9559, 0.9358, 0.9061, 0.8821)),\n",
       " ('GTCG', (0.9482, 0.89, 0.8939, 0.7709, 0.8757)),\n",
       " ('TCCC', (0.9859, 0.9168, 0.9176, 0.8249, 0.872)),\n",
       " ('CTCC', (0.9769, 0.907, 0.9021, 0.8144, 0.8692)),\n",
       " ('GTTC', (0.9522, 0.8939, 0.8798, 0.8047, 0.8691)),\n",
       " ('CATA', (0.9794, 1.0204, 0.9979, 1.0359, 0.8683)),\n",
       " ('GTCT', (0.9694, 0.9085, 0.8975, 0.8142, 0.8669)),\n",
       " ('AGTC', (0.9712, 0.927, 0.9177, 0.8479, 0.8661)),\n",
       " ('ACTA', (1.0119, 1.0131, 1.0013, 1.0125, 0.866)),\n",
       " ('AAGT', (1.0034, 0.9922, 0.9683, 0.9688, 0.8655)),\n",
       " ('CTCT', (0.9785, 0.9178, 0.8928, 0.8188, 0.8638)),\n",
       " ('CAAT', (0.9745, 0.9871, 0.9541, 0.9604, 0.8631)),\n",
       " ('ACTC', (0.9914, 0.9361, 0.9186, 0.8575, 0.8597)),\n",
       " ('TTCC', (0.9686, 0.9047, 0.8869, 0.8153, 0.8578)),\n",
       " ('GTAT', (0.9925, 0.9971, 0.9774, 0.9781, 0.8578)),\n",
       " ('ATCA', (0.9714, 0.9542, 0.9137, 0.8963, 0.8516)),\n",
       " ('TTGA', (0.9836, 0.932, 0.9197, 0.8926, 0.8503)),\n",
       " ('TTAG', (0.9843, 0.9812, 0.9662, 0.9598, 0.8483)),\n",
       " ('CGTC', (0.9319, 0.8591, 0.8514, 0.7297, 0.8471)),\n",
       " ('CCTT', (0.9255, 0.8788, 0.8771, 0.8263, 0.837)),\n",
       " ('AACT', (1.0063, 0.9826, 0.9567, 0.9292, 0.8369)),\n",
       " ('CTTC', (0.9393, 0.8823, 0.8537, 0.7715, 0.836)),\n",
       " ('TCGA', (0.9757, 0.8904, 0.879, 0.7907, 0.8349)),\n",
       " ('GATT', (0.9935, 0.937, 0.9257, 0.8757, 0.833)),\n",
       " ('ATCG', (0.9705, 0.8866, 0.8689, 0.763, 0.8317)),\n",
       " ('TTTG', (0.9615, 0.9163, 0.8984, 0.8641, 0.8301)),\n",
       " ('GTTA', (0.9649, 0.9645, 0.9367, 0.9229, 0.8294)),\n",
       " ('TCAA', (0.9702, 0.9419, 0.8961, 0.8531, 0.8245)),\n",
       " ('CTAA', (1.0121, 0.9818, 0.9544, 0.9331, 0.8228)),\n",
       " ('CATT', (0.9374, 0.9416, 0.9084, 0.9073, 0.8228)),\n",
       " ('TAAC', (0.996, 0.9902, 0.9603, 0.9578, 0.8212)),\n",
       " ('TAGT', (0.9962, 0.9752, 0.9486, 0.9247, 0.8179)),\n",
       " ('TACT', (0.992, 0.9664, 0.944, 0.9217, 0.8141)),\n",
       " ('AAAT', (1.0635, 1.064, 1.0211, 1.0445, 0.8135)),\n",
       " ('AATC', (1.0042, 0.9533, 0.9167, 0.8702, 0.8092)),\n",
       " ('TCCT', (0.9528, 0.8804, 0.8573, 0.7715, 0.8061)),\n",
       " ('AGTT', (0.9612, 0.9375, 0.9166, 0.8913, 0.8026)),\n",
       " ('TCTC', (0.9587, 0.8871, 0.8537, 0.7636, 0.8018)),\n",
       " ('CGTT', (0.9265, 0.8673, 0.8423, 0.7567, 0.8018)),\n",
       " ('TTCG', (0.9493, 0.8677, 0.8487, 0.7482, 0.8008)),\n",
       " ('TGTT', (0.9455, 0.9226, 0.8887, 0.8522, 0.7999)),\n",
       " ('CTAT', (0.9844, 0.9751, 0.9486, 0.9354, 0.7959)),\n",
       " ('TTCA', (0.9358, 0.9128, 0.8649, 0.8246, 0.7934)),\n",
       " ('AATA', (1.0589, 1.1001, 1.0686, 1.1243, 0.7855)),\n",
       " ('ATTC', (0.9705, 0.9222, 0.8892, 0.8399, 0.7843)),\n",
       " ('ATCT', (0.9826, 0.9232, 0.889, 0.8335, 0.782)),\n",
       " ('ACTT', (0.9582, 0.9308, 0.9036, 0.8674, 0.779)),\n",
       " ('TTAC', (0.9538, 0.95, 0.9241, 0.9149, 0.7752)),\n",
       " ('TTGT', (0.9491, 0.8923, 0.8627, 0.7953, 0.7749)),\n",
       " ('TCAT', (0.9329, 0.9097, 0.866, 0.8263, 0.7735)),\n",
       " ('GTTT', (0.944, 0.9072, 0.8797, 0.8382, 0.7731)),\n",
       " ('TCTA', (0.993, 0.9328, 0.9022, 0.8416, 0.7671)),\n",
       " ('ATAA', (1.0511, 1.0769, 1.0384, 1.0704, 0.7659)),\n",
       " ('TAAA', (1.0631, 1.0553, 1.0063, 1.0062, 0.7619)),\n",
       " ('TCGT', (0.9371, 0.8467, 0.8195, 0.6897, 0.7493)),\n",
       " ('CTTA', (0.9482, 0.9403, 0.9085, 0.8853, 0.7488)),\n",
       " ('TATC', (0.9705, 0.9322, 0.8934, 0.8543, 0.7456)),\n",
       " ('CTTT', (0.9222, 0.8793, 0.8527, 0.7985, 0.7313)),\n",
       " ('TTTC', (0.9375, 0.8765, 0.8311, 0.7622, 0.7309)),\n",
       " ('TTCT', (0.9491, 0.878, 0.8407, 0.7573, 0.7263)),\n",
       " ('AATT', (1.0044, 0.9843, 0.9387, 0.925, 0.7098)),\n",
       " ('TCTT', (0.9304, 0.8661, 0.8264, 0.7422, 0.7049)),\n",
       " ('ATAT', (1.0114, 1.0437, 1.0092, 1.0377, 0.6985)),\n",
       " ('TATA', (1.024, 1.0619, 1.0308, 1.0495, 0.6883)),\n",
       " ('ATTA', (0.9865, 0.9979, 0.9601, 0.964, 0.6724)),\n",
       " ('TAAT', (1.0092, 0.9972, 0.9534, 0.9453, 0.6717)),\n",
       " ('ATTT', (0.9613, 0.942, 0.8886, 0.8683, 0.6609)),\n",
       " ('TTAA', (0.9898, 0.9762, 0.9261, 0.9024, 0.655)),\n",
       " ('TATT', (0.9665, 0.9633, 0.914, 0.8942, 0.6303)),\n",
       " ('TTTT', (0.9425, 0.8991, 0.8394, 0.7928, 0.6199)),\n",
       " ('TTTA', (0.9502, 0.9413, 0.8936, 0.8685, 0.6146)),\n",
       " ('TTAT', (0.9498, 0.9534, 0.9017, 0.8789, 0.6093))]"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "neg_four_mers"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "47bc768a",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "<__array_function__ internals>:180: VisibleDeprecationWarning: Creating an ndarray from ragged nested sequences (which is a list-or-tuple of lists-or-tuples-or ndarrays with different lengths or shapes) is deprecated. If you meant to do this, you must specify 'dtype=object' when creating the ndarray.\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "(17664, 2)"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "in_vitro_neg_seqs = np.concatenate((neg_four_mers,neg_five_mers,neg_six_mers))\n",
    "in_vitro_neg_seqs.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "88397cff",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "17664"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "unique_rbns = {}\n",
    "for s, arr in in_vitro_neg_seqs:\n",
    "    if s not in unique_rbns:\n",
    "        unique_rbns[s] = arr\n",
    "\n",
    "in_vitro_neg_seqs = list(unique_rbns.items())\n",
    "len(in_vitro_neg_seqs)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "94988d48",
   "metadata": {},
   "source": [
    "Construct placeholder expression change vector"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "950281ea",
   "metadata": {},
   "outputs": [],
   "source": [
    "placeholder_logfold_changes = np.zeros(shape=logfold_changes.shape)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b14cf59a",
   "metadata": {},
   "source": [
    "## Split into train/val/test"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "e9d8ef74",
   "metadata": {},
   "outputs": [],
   "source": [
    "# number of total samples per class\n",
    "combinations_size = 100000"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a99c5d06",
   "metadata": {},
   "source": [
    "Positive samples"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "c5fdbf5e",
   "metadata": {},
   "outputs": [],
   "source": [
    "from itertools import product\n",
    "\n",
    "random.shuffle(in_vivo_pos_seqs)\n",
    "random.shuffle(in_vitro_pos_seqs)\n",
    "\n",
    "n_eclip = len(in_vivo_pos_seqs)\n",
    "n_rbns  = len(in_vitro_pos_seqs)\n",
    "\n",
    "# Split indices\n",
    "train_eclip = in_vivo_pos_seqs[:int(0.7*n_eclip)]\n",
    "val_eclip   = in_vivo_pos_seqs[int(0.7*n_eclip):int(0.85*n_eclip)]\n",
    "test_eclip  = in_vivo_pos_seqs[int(0.85*n_eclip):]\n",
    "\n",
    "train_rbns = in_vitro_pos_seqs[:int(0.7*n_rbns)]\n",
    "val_rbns   = in_vitro_pos_seqs[int(0.7*n_rbns):int(0.85*n_rbns)]\n",
    "test_rbns  = in_vitro_pos_seqs[int(0.85*n_rbns):]\n",
    "\n",
    "pos_train_samples = [\n",
    "    (random.choice(train_eclip), random.choice(train_rbns))\n",
    "    for _ in range(combinations_size)\n",
    "]\n",
    "pos_val_samples = [\n",
    "    (random.choice(val_eclip), random.choice(val_rbns))\n",
    "    for _ in range(combinations_size)\n",
    "]\n",
    "pos_test_samples = [\n",
    "    (random.choice(test_eclip), random.choice(test_rbns))\n",
    "    for _ in range(combinations_size)\n",
    "]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "91baf0a1",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "eCLIP sequences are disjoint across splits\n"
     ]
    }
   ],
   "source": [
    "train_eclips = set(eclip for eclip, _ in pos_train_samples)\n",
    "val_eclips   = set(eclip for eclip, _ in pos_val_samples)\n",
    "test_eclips  = set(eclip for eclip, _ in pos_test_samples)\n",
    "\n",
    "assert train_eclips.isdisjoint(val_eclips), \"Overlap between train and val eCLIP sequences!\"\n",
    "assert train_eclips.isdisjoint(test_eclips), \"Overlap between train and test eCLIP sequences!\"\n",
    "assert val_eclips.isdisjoint(test_eclips), \"Overlap between val and test eCLIP sequences!\"\n",
    "print(\"eCLIP sequences are disjoint across splits\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "0ffaaf81",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "RBNS sequences are disjoint across splits\n"
     ]
    }
   ],
   "source": [
    "train_rbns = set(rbns for _, (rbns, _) in pos_train_samples)\n",
    "val_rbns   = set(rbns for _, (rbns, _) in pos_val_samples)\n",
    "test_rbns  = set(rbns for _, (rbns, _) in pos_test_samples)\n",
    "\n",
    "assert train_rbns.isdisjoint(val_rbns), \"Overlap between train and val RBNS sequences!\"\n",
    "assert train_rbns.isdisjoint(test_rbns), \"Overlap between train and test RBNS sequences!\"\n",
    "assert val_rbns.isdisjoint(test_rbns), \"Overlap between val and test RBNS sequences!\"\n",
    "print(\"RBNS sequences are disjoint across splits\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "330d4bc3",
   "metadata": {},
   "outputs": [],
   "source": [
    "pos_train_label = np.ones(len(pos_train_samples))\n",
    "pos_val_label = np.ones(len(pos_val_samples))\n",
    "pos_test_label = np.ones(len(pos_test_samples))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c3da2c3a",
   "metadata": {},
   "source": [
    "Negative samples"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "d7c965b7",
   "metadata": {},
   "outputs": [],
   "source": [
    "random.shuffle(in_vivo_neg_seqs)\n",
    "random.shuffle(in_vitro_neg_seqs)\n",
    "\n",
    "n_eclip = len(in_vivo_neg_seqs)\n",
    "n_rbns  = len(in_vitro_neg_seqs)\n",
    "\n",
    "# Split indices\n",
    "train_eclip = in_vivo_neg_seqs[:int(0.7*n_eclip)]\n",
    "val_eclip   = in_vivo_neg_seqs[int(0.7*n_eclip):int(0.85*n_eclip)]\n",
    "test_eclip  = in_vivo_neg_seqs[int(0.85*n_eclip):]\n",
    "\n",
    "train_rbns = in_vitro_neg_seqs[:int(0.7*n_rbns)]\n",
    "val_rbns   = in_vitro_neg_seqs[int(0.7*n_rbns):int(0.85*n_rbns)]\n",
    "test_rbns  = in_vitro_neg_seqs[int(0.85*n_rbns):]\n",
    "\n",
    "neg_train_samples = [\n",
    "    (random.choice(train_eclip), random.choice(train_rbns))\n",
    "    for _ in range(combinations_size)\n",
    "]\n",
    "neg_val_samples = [\n",
    "    (random.choice(val_eclip), random.choice(val_rbns))\n",
    "    for _ in range(combinations_size)\n",
    "]\n",
    "neg_test_samples = [\n",
    "    (random.choice(test_eclip), random.choice(test_rbns))\n",
    "    for _ in range(combinations_size)\n",
    "]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "19d1f6e3",
   "metadata": {},
   "outputs": [],
   "source": [
    "neg_train_label = np.zeros(len(neg_train_samples))\n",
    "neg_val_label = np.zeros(len(neg_val_samples))\n",
    "neg_test_label = np.zeros(len(neg_test_samples))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9cc54a9e",
   "metadata": {},
   "source": [
    "Combine for full splits"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "726d4838",
   "metadata": {},
   "outputs": [],
   "source": [
    "train_data = np.concatenate((pos_train_samples, neg_train_samples))\n",
    "val_data = np.concatenate((pos_val_samples, neg_val_samples))\n",
    "test_data = np.concatenate((pos_test_samples, neg_test_samples))\n",
    "\n",
    "train_labels = np.concatenate((pos_train_label, neg_train_label))\n",
    "val_labels = np.concatenate((pos_val_label, neg_val_label))\n",
    "test_labels = np.concatenate((pos_test_label, neg_test_label))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "8ce65d9c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# get positive/negative label values\n",
    "positive_value = (\"SRSF9\", logfold_changes)\n",
    "negative_value = (\"Not SRSF9\", placeholder_logfold_changes)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "877f0473",
   "metadata": {},
   "source": [
    "## Save as dictionaries"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "16dcdac6",
   "metadata": {},
   "source": [
    "Save as pickle files since there are nested data structures"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "eff9d82c",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pickle"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "cfda55f9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'\\nSaved data format for each split:\\n- Input\\n    - List containing:\\n        - eCLIP sequence: text\\n        - Tuple of RBNS sequence and binding affinity: (text, vector)\\n- Output\\n    - List containing:\\n        - RBP label: text\\n        - Logfold gene expression change: vector\\n        \\nLabels save separately to save storage:\\n- List containing:\\n    - RBP label: text\\n    - Logfold gene expression change: vector\\n'"
      ]
     },
     "execution_count": 37,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "'''\n",
    "Saved data format for each split:\n",
    "- Input\n",
    "    - List containing:\n",
    "        - eCLIP sequence: text\n",
    "        - Tuple of RBNS sequence and binding affinity: (text, vector)\n",
    "- Output\n",
    "    - List containing:\n",
    "        - RBP label: text\n",
    "        - Logfold gene expression change: vector\n",
    "        \n",
    "Labels save separately to save storage:\n",
    "- List containing:\n",
    "    - RBP label: text\n",
    "    - Logfold gene expression change: vector\n",
    "'''"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "id": "6c84bd69",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Save train\n",
    "with open(\"SRSF9/train_split.pkl\", \"wb\") as f:\n",
    "    pickle.dump((train_data, train_labels), f)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "id": "afb7be9c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Save val\n",
    "with open(\"SRSF9/val_split.pkl\", \"wb\") as f:\n",
    "    pickle.dump((val_data, val_labels), f)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "6266df79",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Save test\n",
    "with open(\"SRSF9/test_split.pkl\", \"wb\") as f:\n",
    "    pickle.dump((test_data, test_labels), f)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "id": "95900830",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Save positive label\n",
    "with open(\"SRSF9/SRSF9_positive_label.pkl\", \"wb\") as f:\n",
    "    pickle.dump(positive_value, f)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "id": "fe972c58",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Save negative label\n",
    "with open(\"SRSF9/SRSF9_negative_label.pkl\", \"wb\") as f:\n",
    "    pickle.dump(negative_value, f)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "521fb44e",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "perturb",
   "language": "python",
   "name": "perturb"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
