{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "3ee91ff5",
   "metadata": {},
   "outputs": [],
   "source": [
    "%load_ext autoreload\n",
    "%autoreload 2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "15865688",
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ff5d5753",
   "metadata": {},
   "source": [
    "## Read data"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "235e3bf2",
   "metadata": {},
   "source": [
    "Load eCLIP-seq sequences"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "abe42208",
   "metadata": {},
   "outputs": [],
   "source": [
    "eclip = pd.read_csv(\"TIA1/TIA1_eclip.txt\", sep='\\t', header=None)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "32e0bc25",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr7:45932874-45932893</td>\n",
       "      <td>CAGTAAAAACTTGTACTAT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr7:45932893-45932940</td>\n",
       "      <td>GTTAATAACTTGTCCTATGTCAATTTGTATATCATGAAACACTTCTC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr7:94296528-94296614</td>\n",
       "      <td>TAATTCTTTGCTACTGCCATTGACCCTGCTGCAGGATTTGTGTCAT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr7:45932862-45932874</td>\n",
       "      <td>ATTTTTTCTACA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr7:127231645-127231726</td>\n",
       "      <td>CTGTTTTGGGTGTACTCTAGGGGCCAGGTTGGGAGGGGGAAGGTGA...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                          0                                                  1\n",
       "0    chr7:45932874-45932893                                CAGTAAAAACTTGTACTAT\n",
       "1    chr7:45932893-45932940    GTTAATAACTTGTCCTATGTCAATTTGTATATCATGAAACACTTCTC\n",
       "2    chr7:94296528-94296614  TAATTCTTTGCTACTGCCATTGACCCTGCTGCAGGATTTGTGTCAT...\n",
       "3    chr7:45932862-45932874                                       ATTTTTTCTACA\n",
       "4  chr7:127231645-127231726  CTGTTTTGGGTGTACTCTAGGGGCCAGGTTGGGAGGGGGAAGGTGA..."
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "eclip.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8e72135a",
   "metadata": {},
   "source": [
    "Load RBNS binding affinities"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "7bc981f7",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>[TIA1]</th>\n",
       "      <th>5 nM</th>\n",
       "      <th>20 nM</th>\n",
       "      <th>80 nM</th>\n",
       "      <th>320 nM</th>\n",
       "      <th>1300 nM</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>TTTT</td>\n",
       "      <td>1.1589</td>\n",
       "      <td>1.2097</td>\n",
       "      <td>1.0931</td>\n",
       "      <td>1.0908</td>\n",
       "      <td>1.5628</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>TTTA</td>\n",
       "      <td>1.1272</td>\n",
       "      <td>1.0755</td>\n",
       "      <td>1.0910</td>\n",
       "      <td>1.0877</td>\n",
       "      <td>1.2518</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ATTT</td>\n",
       "      <td>1.1269</td>\n",
       "      <td>1.1175</td>\n",
       "      <td>1.0903</td>\n",
       "      <td>1.0880</td>\n",
       "      <td>1.2115</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>TTAT</td>\n",
       "      <td>1.1057</td>\n",
       "      <td>1.0367</td>\n",
       "      <td>1.0797</td>\n",
       "      <td>1.0719</td>\n",
       "      <td>1.2024</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>TATT</td>\n",
       "      <td>1.1153</td>\n",
       "      <td>1.0524</td>\n",
       "      <td>1.0917</td>\n",
       "      <td>1.0839</td>\n",
       "      <td>1.2018</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>251</th>\n",
       "      <td>TCGT</td>\n",
       "      <td>0.9579</td>\n",
       "      <td>0.9496</td>\n",
       "      <td>0.9480</td>\n",
       "      <td>0.9331</td>\n",
       "      <td>0.8969</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>252</th>\n",
       "      <td>TCGG</td>\n",
       "      <td>0.9434</td>\n",
       "      <td>0.9667</td>\n",
       "      <td>0.9405</td>\n",
       "      <td>0.9331</td>\n",
       "      <td>0.8898</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>253</th>\n",
       "      <td>CGAG</td>\n",
       "      <td>0.9099</td>\n",
       "      <td>0.9385</td>\n",
       "      <td>0.9129</td>\n",
       "      <td>0.8956</td>\n",
       "      <td>0.8841</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>254</th>\n",
       "      <td>CGTC</td>\n",
       "      <td>0.9235</td>\n",
       "      <td>0.9012</td>\n",
       "      <td>0.9271</td>\n",
       "      <td>0.9114</td>\n",
       "      <td>0.8757</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>255</th>\n",
       "      <td>GTCG</td>\n",
       "      <td>0.9278</td>\n",
       "      <td>0.9291</td>\n",
       "      <td>0.9285</td>\n",
       "      <td>0.9246</td>\n",
       "      <td>0.8716</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>256 rows Ã— 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    [TIA1]    5 nM   20 nM   80 nM  320 nM  1300 nM\n",
       "0     TTTT  1.1589  1.2097  1.0931  1.0908   1.5628\n",
       "1     TTTA  1.1272  1.0755  1.0910  1.0877   1.2518\n",
       "2     ATTT  1.1269  1.1175  1.0903  1.0880   1.2115\n",
       "3     TTAT  1.1057  1.0367  1.0797  1.0719   1.2024\n",
       "4     TATT  1.1153  1.0524  1.0917  1.0839   1.2018\n",
       "..     ...     ...     ...     ...     ...      ...\n",
       "251   TCGT  0.9579  0.9496  0.9480  0.9331   0.8969\n",
       "252   TCGG  0.9434  0.9667  0.9405  0.9331   0.8898\n",
       "253   CGAG  0.9099  0.9385  0.9129  0.8956   0.8841\n",
       "254   CGTC  0.9235  0.9012  0.9271  0.9114   0.8757\n",
       "255   GTCG  0.9278  0.9291  0.9285  0.9246   0.8716\n",
       "\n",
       "[256 rows x 6 columns]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "four_mers = pd.read_csv(\"TIA1/TIA1_4.tsv\", sep='\\t')\n",
    "five_mers = pd.read_csv(\"TIA1/TIA1_5.tsv\", sep='\\t',)\n",
    "six_mers = pd.read_csv(\"TIA1/TIA1_6.tsv\", sep='\\t',)\n",
    "four_mers"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "af6ffd4e",
   "metadata": {},
   "source": [
    "Load knockdown differential rna-seq data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "d0e331a6",
   "metadata": {},
   "outputs": [],
   "source": [
    "# get other diff expr for subsetting to shared genes later\n",
    "diff_exp_HNRNPC = pd.read_csv(\"HNRNPC/HNRNPC.tsv\", sep='\\t',)\n",
    "diff_exp_SRSF9 = pd.read_csv(\"SRSF9/SRSF9.tsv\", sep='\\t',)\n",
    "diff_exp_PCBP1 = pd.read_csv(\"PCBP1/PCBP1.tsv\", sep='\\t',)\n",
    "diff_exp_TIA1 = pd.read_csv(\"TIA1/TIA1.tsv\", sep='\\t',)\n",
    "diff_exp_TRA2A = pd.read_csv(\"TRA2A/TRA2A.tsv\", sep='\\t',)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1612ca75",
   "metadata": {},
   "source": [
    "## Generate positive pairs"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b0787ccc",
   "metadata": {},
   "source": [
    "EClip positive binding sites"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "592b1f36",
   "metadata": {},
   "outputs": [],
   "source": [
    "in_vivo_pos_seqs = eclip[1].values"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "18057862",
   "metadata": {},
   "outputs": [],
   "source": [
    "# filter for length and unique values\n",
    "in_vivo_pos_seqs = [s for s in in_vivo_pos_seqs if len(s) >= 10]\n",
    "in_vivo_pos_seqs = list(set(in_vivo_pos_seqs))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "adf61072",
   "metadata": {},
   "source": [
    "RBNS significant binding affinity"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "fd120254",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_signficant_sequences(df):\n",
    "    num_cols = df.columns[1:]\n",
    "    filtered = df[df[num_cols].gt(2).any(axis=1)]\n",
    "    pairs = [\n",
    "        (row[df.columns[0]], tuple(row[num_cols]))\n",
    "        for _, row in filtered.iterrows()\n",
    "    ]\n",
    "    return pairs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "ed65dc0f",
   "metadata": {},
   "outputs": [],
   "source": [
    "pos_four_mers = get_signficant_sequences(four_mers)\n",
    "pos_five_mers = get_signficant_sequences(five_mers)\n",
    "pos_six_mers = get_signficant_sequences(six_mers)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "1b63d4fe",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(0, 1, 7)"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(pos_four_mers), len(pos_five_mers), len(pos_six_mers), "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "6894c643",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "<__array_function__ internals>:180: VisibleDeprecationWarning: Creating an ndarray from ragged nested sequences (which is a list-or-tuple of lists-or-tuples-or ndarrays with different lengths or shapes) is deprecated. If you meant to do this, you must specify 'dtype=object' when creating the ndarray.\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "(8, 2)"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "in_vitro_pos_seqs = np.concatenate((pos_five_mers,pos_six_mers))\n",
    "in_vitro_pos_seqs.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "3bc23f33",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "8"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "unique_rbns = {}\n",
    "for s, arr in in_vitro_pos_seqs:\n",
    "    if s not in unique_rbns:\n",
    "        unique_rbns[s] = arr\n",
    "\n",
    "in_vitro_pos_seqs = list(unique_rbns.items())\n",
    "len(in_vitro_pos_seqs)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6ad2eb50",
   "metadata": {},
   "source": [
    "Filter diff expr"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "7a6b7538",
   "metadata": {},
   "outputs": [],
   "source": [
    "# remove when gene expression is always 0 or if logfold change is inf/-inf\n",
    "diff_exp_filtered_HNRNPC = diff_exp_HNRNPC[\n",
    "    ~(\n",
    "        ((diff_exp_HNRNPC[\"value_1\"] == 0) & (diff_exp_HNRNPC[\"value_2\"] == 0)) |\n",
    "        (diff_exp_HNRNPC[\"log2(fold_change)\"].isin([np.inf, -np.inf]))\n",
    "    )\n",
    "]\n",
    "\n",
    "diff_exp_filtered_SRSF9 = diff_exp_SRSF9[\n",
    "    ~(\n",
    "        ((diff_exp_SRSF9[\"value_1\"] == 0) & (diff_exp_SRSF9[\"value_2\"] == 0)) |\n",
    "        (diff_exp_SRSF9[\"log2(fold_change)\"].isin([np.inf, -np.inf]))\n",
    "    )\n",
    "]\n",
    "\n",
    "diff_exp_filtered_PCBP1 = diff_exp_PCBP1[\n",
    "    ~(\n",
    "        ((diff_exp_PCBP1[\"value_1\"] == 0) & (diff_exp_PCBP1[\"value_2\"] == 0)) |\n",
    "        (diff_exp_PCBP1[\"log2(fold_change)\"].isin([np.inf, -np.inf]))\n",
    "    )\n",
    "]\n",
    "\n",
    "diff_exp_filtered_TIA1 = diff_exp_TIA1[\n",
    "    ~(\n",
    "        ((diff_exp_TIA1[\"value_1\"] == 0) & (diff_exp_TIA1[\"value_2\"] == 0)) |\n",
    "        (diff_exp_TIA1[\"log2(fold_change)\"].isin([np.inf, -np.inf]))\n",
    "    )\n",
    "]\n",
    "\n",
    "diff_exp_filtered_TRA2A = diff_exp_TRA2A[\n",
    "    ~(\n",
    "        ((diff_exp_TRA2A[\"value_1\"] == 0) & (diff_exp_TRA2A[\"value_2\"] == 0)) |\n",
    "        (diff_exp_TRA2A[\"log2(fold_change)\"].isin([np.inf, -np.inf]))\n",
    "    )\n",
    "]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "377252f2",
   "metadata": {},
   "outputs": [],
   "source": [
    "dfs = [diff_exp_filtered_HNRNPC, diff_exp_filtered_SRSF9, diff_exp_filtered_PCBP1, diff_exp_filtered_TIA1, diff_exp_filtered_TRA2A]\n",
    "col = \"gene_id\"\n",
    "shared_keys = set.intersection(*(set(df[col]) for df in dfs))\n",
    "\n",
    "# 2. Subset the dataframe you want (example: df1)\n",
    "diff_exp_filtered_TIA1 = diff_exp_filtered_TIA1[diff_exp_filtered_TIA1[col].isin(shared_keys)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "fdeaa9e5",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>test_id</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene</th>\n",
       "      <th>locus</th>\n",
       "      <th>sample_1</th>\n",
       "      <th>sample_2</th>\n",
       "      <th>status</th>\n",
       "      <th>value_1</th>\n",
       "      <th>value_2</th>\n",
       "      <th>log2(fold_change)</th>\n",
       "      <th>test_stat</th>\n",
       "      <th>p_value</th>\n",
       "      <th>q_value</th>\n",
       "      <th>significant</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>XLOC_000008</td>\n",
       "      <td>XLOC_000008</td>\n",
       "      <td>OR4F29</td>\n",
       "      <td>chr1:317719-461954</td>\n",
       "      <td>TIA1-BGHLV12-HepG2</td>\n",
       "      <td>Control</td>\n",
       "      <td>OK</td>\n",
       "      <td>0.627869</td>\n",
       "      <td>0.250779</td>\n",
       "      <td>-1.324040</td>\n",
       "      <td>-0.527240</td>\n",
       "      <td>0.02875</td>\n",
       "      <td>0.127831</td>\n",
       "      <td>no</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>XLOC_000012</td>\n",
       "      <td>XLOC_000012</td>\n",
       "      <td>MTND1P23</td>\n",
       "      <td>chr1:536815-660283</td>\n",
       "      <td>TIA1-BGHLV12-HepG2</td>\n",
       "      <td>Control</td>\n",
       "      <td>OK</td>\n",
       "      <td>246.719000</td>\n",
       "      <td>191.170000</td>\n",
       "      <td>-0.368011</td>\n",
       "      <td>-0.180525</td>\n",
       "      <td>0.52255</td>\n",
       "      <td>0.761433</td>\n",
       "      <td>no</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>XLOC_000013</td>\n",
       "      <td>XLOC_000013</td>\n",
       "      <td>MTND2P28</td>\n",
       "      <td>chr1:536815-660283</td>\n",
       "      <td>TIA1-BGHLV12-HepG2</td>\n",
       "      <td>Control</td>\n",
       "      <td>OK</td>\n",
       "      <td>610.343000</td>\n",
       "      <td>659.356000</td>\n",
       "      <td>0.111438</td>\n",
       "      <td>0.204620</td>\n",
       "      <td>0.77205</td>\n",
       "      <td>0.905931</td>\n",
       "      <td>no</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>XLOC_000014</td>\n",
       "      <td>XLOC_000014</td>\n",
       "      <td>hsa-mir-6723</td>\n",
       "      <td>chr1:536815-660283</td>\n",
       "      <td>TIA1-BGHLV12-HepG2</td>\n",
       "      <td>Control</td>\n",
       "      <td>OK</td>\n",
       "      <td>1268.430000</td>\n",
       "      <td>1243.250000</td>\n",
       "      <td>-0.028933</td>\n",
       "      <td>-0.095724</td>\n",
       "      <td>0.89120</td>\n",
       "      <td>0.961140</td>\n",
       "      <td>no</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>XLOC_000015</td>\n",
       "      <td>XLOC_000015</td>\n",
       "      <td>RP5-857K21.7</td>\n",
       "      <td>chr1:536815-660283</td>\n",
       "      <td>TIA1-BGHLV12-HepG2</td>\n",
       "      <td>Control</td>\n",
       "      <td>OK</td>\n",
       "      <td>647.872000</td>\n",
       "      <td>642.983000</td>\n",
       "      <td>-0.010927</td>\n",
       "      <td>-0.015454</td>\n",
       "      <td>0.98180</td>\n",
       "      <td>0.992450</td>\n",
       "      <td>no</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        test_id      gene_id          gene               locus  \\\n",
       "7   XLOC_000008  XLOC_000008        OR4F29  chr1:317719-461954   \n",
       "11  XLOC_000012  XLOC_000012      MTND1P23  chr1:536815-660283   \n",
       "12  XLOC_000013  XLOC_000013      MTND2P28  chr1:536815-660283   \n",
       "13  XLOC_000014  XLOC_000014  hsa-mir-6723  chr1:536815-660283   \n",
       "14  XLOC_000015  XLOC_000015  RP5-857K21.7  chr1:536815-660283   \n",
       "\n",
       "              sample_1 sample_2 status      value_1      value_2  \\\n",
       "7   TIA1-BGHLV12-HepG2  Control     OK     0.627869     0.250779   \n",
       "11  TIA1-BGHLV12-HepG2  Control     OK   246.719000   191.170000   \n",
       "12  TIA1-BGHLV12-HepG2  Control     OK   610.343000   659.356000   \n",
       "13  TIA1-BGHLV12-HepG2  Control     OK  1268.430000  1243.250000   \n",
       "14  TIA1-BGHLV12-HepG2  Control     OK   647.872000   642.983000   \n",
       "\n",
       "    log2(fold_change)  test_stat  p_value   q_value significant  \n",
       "7           -1.324040  -0.527240  0.02875  0.127831          no  \n",
       "11          -0.368011  -0.180525  0.52255  0.761433          no  \n",
       "12           0.111438   0.204620  0.77205  0.905931          no  \n",
       "13          -0.028933  -0.095724  0.89120  0.961140          no  \n",
       "14          -0.010927  -0.015454  0.98180  0.992450          no  "
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "diff_exp_filtered_TIA1.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "1046db6c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(17985, 14)"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "diff_exp_filtered_TIA1.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "d8fe4752",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([-1.32404 , -0.368011,  0.111438, ..., -0.601615, -0.659382,\n",
       "       -1.69771 ])"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "logfold_changes = diff_exp_filtered_TIA1['log2(fold_change)'].values\n",
    "logfold_changes"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "14a14525",
   "metadata": {},
   "source": [
    "## Generate negative pairs"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "eaa775d1",
   "metadata": {},
   "source": [
    "Shuffle eCLIP sequences"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "68b0a794",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "239224"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import random\n",
    "def shuffle_by_pairs(seq):\n",
    "    pairs = [seq[i:i+2] for i in range(0, len(seq), 2)]\n",
    "    random.shuffle(pairs)\n",
    "    return \"\".join(pairs)\n",
    "in_vivo_neg_seqs = [shuffle_by_pairs(seq) for seq in eclip[1].values]\n",
    "len(in_vivo_neg_seqs)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "1ec9c58c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "233638"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# filter for length and unique values\n",
    "in_vivo_neg_seqs = [s for s in in_vivo_neg_seqs if len(s) >= 10]\n",
    "in_vivo_neg_seqs = list(set(in_vivo_neg_seqs))\n",
    "len(in_vivo_neg_seqs)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "02a5e383",
   "metadata": {},
   "source": [
    "Retrieve RBNS sequences that are not enriched"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "669ff6f6",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_insignficant_sequences(df):\n",
    "    num_cols = df.columns[1:]\n",
    "    filtered = df[df[num_cols].lt(2).any(axis=1)]\n",
    "    pairs = [\n",
    "        (row[df.columns[0]], tuple(row[num_cols]))\n",
    "        for _, row in filtered.iterrows()\n",
    "    ]\n",
    "    return pairs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "4a0c3f24",
   "metadata": {},
   "outputs": [],
   "source": [
    "neg_four_mers = get_insignficant_sequences(four_mers)\n",
    "neg_five_mers = get_insignficant_sequences(five_mers)\n",
    "neg_six_mers = get_insignficant_sequences(six_mers)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "7920a3d9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[('TTTT', (1.1589, 1.2097, 1.0931, 1.0908, 1.5628)),\n",
       " ('TTTA', (1.1272, 1.0755, 1.091, 1.0877, 1.2518)),\n",
       " ('ATTT', (1.1269, 1.1175, 1.0903, 1.088, 1.2115)),\n",
       " ('TTAT', (1.1057, 1.0367, 1.0797, 1.0719, 1.2024)),\n",
       " ('TATT', (1.1153, 1.0524, 1.0917, 1.0839, 1.2018)),\n",
       " ('CTTT', (1.0623, 1.076, 1.0189, 1.0049, 1.154)),\n",
       " ('TCTT', (1.0491, 1.0683, 1.0185, 1.0039, 1.1503)),\n",
       " ('AAAA', (1.1711, 1.2103, 1.1931, 1.2454, 1.1352)),\n",
       " ('TTAA', (1.1236, 1.0608, 1.11, 1.1121, 1.1277)),\n",
       " ('TATA', (1.1141, 0.9948, 1.1193, 1.1193, 1.1071)),\n",
       " ('TTCT', (1.0577, 1.0894, 1.0281, 1.0135, 1.1071)),\n",
       " ('TAAA', (1.1406, 1.0761, 1.15, 1.1714, 1.1029)),\n",
       " ('ATTA', (1.1016, 1.0233, 1.0959, 1.0949, 1.1009)),\n",
       " ('TAAT', (1.1089, 1.0295, 1.106, 1.1092, 1.0952)),\n",
       " ('TTTC', (1.0598, 1.0933, 1.0245, 1.0076, 1.092)),\n",
       " ('AATT', (1.1171, 1.093, 1.1048, 1.1057, 1.0919)),\n",
       " ('CTTA', (1.0403, 0.9887, 1.0234, 1.0169, 1.0853)),\n",
       " ('AATA', (1.1129, 1.0157, 1.1333, 1.1501, 1.0822)),\n",
       " ('ATAA', (1.1107, 1.0169, 1.1303, 1.145, 1.08)),\n",
       " ('CTCT', (1.0034, 1.0078, 0.9908, 0.9732, 1.0782)),\n",
       " ('AAAT', (1.1326, 1.1114, 1.142, 1.1668, 1.0773)),\n",
       " ('ATAT', (1.0969, 0.9884, 1.1021, 1.0994, 1.0758)),\n",
       " ('TTAC', (1.0337, 0.9905, 1.0187, 1.0106, 1.0715)),\n",
       " ('CTAT', (1.0395, 0.9778, 1.0333, 1.0235, 1.0631)),\n",
       " ('TACT', (1.0403, 1.0034, 1.0299, 1.0266, 1.0627)),\n",
       " ('ACTT', (1.0362, 1.0309, 1.0199, 1.0145, 1.0582)),\n",
       " ('GTTT', (1.0338, 1.0827, 1.0066, 0.9983, 1.0539)),\n",
       " ('TCTA', (1.0517, 1.0222, 1.0455, 1.0349, 1.0491)),\n",
       " ('ATCT', (1.0443, 1.0365, 1.0365, 1.0221, 1.0488)),\n",
       " ('CATT', (1.0213, 0.9998, 1.0072, 0.9894, 1.0488)),\n",
       " ('TTTG', (1.0427, 1.1211, 1.0107, 1.0064, 1.047)),\n",
       " ('CTAA', (1.0564, 1.0014, 1.0583, 1.0622, 1.0446)),\n",
       " ('ACTA', (1.0445, 0.9768, 1.0516, 1.0565, 1.0414)),\n",
       " ('TTAG', (1.0387, 1.0264, 1.0222, 1.02, 1.0413)),\n",
       " ('TAAC', (1.0385, 0.9795, 1.0425, 1.0404, 1.034)),\n",
       " ('GGGG', (0.9749, 1.0743, 0.9853, 1.0264, 1.0324)),\n",
       " ('AACT', (1.0375, 1.0162, 1.0419, 1.044, 1.0312)),\n",
       " ('CCTT', (1.0022, 0.9865, 0.9661, 0.9375, 1.0307)),\n",
       " ('CCCC', (1.0064, 0.9947, 0.9914, 0.98, 1.0307)),\n",
       " ('AAAG', (1.0651, 1.0997, 1.0808, 1.1033, 1.0305)),\n",
       " ('GCTT', (0.9855, 1.0241, 0.9674, 0.9675, 1.0266)),\n",
       " ('AAAC', (1.0604, 1.0499, 1.0796, 1.0962, 1.0261)),\n",
       " ('CAAA', (1.0582, 1.0523, 1.0783, 1.0932, 1.0252)),\n",
       " ('GAAA', (1.072, 1.1462, 1.085, 1.1081, 1.0246)),\n",
       " ('TGTT', (1.0129, 1.0566, 0.9912, 0.9796, 1.0223)),\n",
       " ('TAAG', (1.0367, 0.996, 1.0421, 1.0495, 1.0214)),\n",
       " ('TGCT', (0.9885, 1.0228, 0.9723, 0.9743, 1.0213)),\n",
       " ('TACC', (1.0087, 0.966, 0.9942, 0.9823, 1.0211)),\n",
       " ('TATC', (1.0232, 0.9732, 1.0159, 0.9979, 1.0206)),\n",
       " ('TATG', (1.0191, 0.9766, 1.0149, 1.0157, 1.02)),\n",
       " ('CCAT', (0.9852, 0.9593, 0.9826, 0.9682, 1.0189)),\n",
       " ('ATTC', (1.0352, 1.0269, 1.0265, 1.0025, 1.0189)),\n",
       " ('GCTA', (0.9902, 0.9791, 0.9871, 0.9998, 1.0188)),\n",
       " ('TCTC', (0.9909, 1.0063, 0.9838, 0.9579, 1.0181)),\n",
       " ('GTTA', (1.0214, 1.0163, 1.0096, 1.005, 1.018)),\n",
       " ('AGAA', (1.0522, 1.0872, 1.066, 1.0737, 1.0178)),\n",
       " ('TAGA', (1.0397, 1.0225, 1.0447, 1.0463, 1.0174)),\n",
       " ('AAGA', (1.048, 1.0772, 1.0619, 1.0741, 1.0164)),\n",
       " ('ATAG', (1.0373, 0.9758, 1.05, 1.056, 1.016)),\n",
       " ('GTAT', (1.0249, 0.9966, 1.0202, 1.015, 1.0159)),\n",
       " ('CTTC', (0.9877, 1.0039, 0.9692, 0.9493, 1.0159)),\n",
       " ('CAAT', (1.0232, 0.98, 1.0344, 1.027, 1.0156)),\n",
       " ('TAGT', (1.0201, 1.0004, 1.0171, 1.0112, 1.0143)),\n",
       " ('AACA', (1.0324, 0.9844, 1.058, 1.0593, 1.014)),\n",
       " ('GTAA', (1.0345, 1.0083, 1.0414, 1.0487, 1.0139)),\n",
       " ('CCTA', (0.9977, 0.9479, 0.9936, 0.9805, 1.0127)),\n",
       " ('ACAA', (1.0344, 0.9883, 1.0598, 1.0629, 1.0118)),\n",
       " ('ATAC', (1.0267, 0.9596, 1.033, 1.0301, 1.0114)),\n",
       " ('TAGC', (0.9852, 0.9686, 0.9865, 0.9916, 1.0111)),\n",
       " ('TTGT', (1.0141, 1.06, 0.9928, 0.979, 1.011)),\n",
       " ('CTCC', (0.9581, 0.9552, 0.9478, 0.9278, 1.0106)),\n",
       " ('GCCA', (0.9575, 0.9782, 0.9645, 0.9902, 1.0105)),\n",
       " ('ACCA', (0.997, 0.9632, 1.0047, 1.0019, 1.0102)),\n",
       " ('CCAA', (1.0035, 0.9928, 1.0137, 1.0104, 1.0093)),\n",
       " ('TGTA', (1.009, 1.0023, 1.0054, 0.999, 1.0093)),\n",
       " ('TACA', (1.0193, 0.9517, 1.0279, 1.0169, 1.0092)),\n",
       " ('CATA', (1.0155, 0.9318, 1.0313, 1.0266, 1.0092)),\n",
       " ('TTCC', (1.0049, 1.0143, 0.9782, 0.9509, 1.0089)),\n",
       " ('CACT', (0.9842, 0.967, 0.9843, 0.9798, 1.0088)),\n",
       " ('AGGG', (0.9776, 1.0332, 0.9901, 1.0157, 1.0077)),\n",
       " ('TTCA', (1.0211, 1.0295, 1.0129, 0.9888, 1.0076)),\n",
       " ('CTAC', (0.9886, 0.9446, 0.9855, 0.9801, 1.0076)),\n",
       " ('GGGT', (0.9825, 1.0524, 0.9818, 1.0004, 1.0071)),\n",
       " ('AGTA', (1.0176, 0.9855, 1.0302, 1.0308, 1.0062)),\n",
       " ('CCCT', (0.9885, 0.9674, 0.9637, 0.9435, 1.004)),\n",
       " ('TGAA', (1.0331, 1.0778, 1.0333, 1.0401, 1.003)),\n",
       " ('ACCT', (0.9924, 0.9689, 0.9808, 0.9648, 1.0024)),\n",
       " ('AATG', (1.025, 1.0357, 1.0319, 1.0422, 1.0024)),\n",
       " ('AACC', (1.0114, 0.9859, 1.015, 1.0091, 1.0011)),\n",
       " ('ACTC', (0.9786, 0.9662, 0.9775, 0.9606, 1.0011)),\n",
       " ('GGTA', (0.9976, 1.0231, 0.9968, 1.0074, 1.001)),\n",
       " ('TTGC', (0.9763, 1.0306, 0.9631, 0.9594, 1.0004)),\n",
       " ('ACAT', (1.0109, 0.9492, 1.0179, 1.0074, 1.0002)),\n",
       " ('TCCT', (0.9928, 0.9785, 0.97, 0.9432, 0.9994)),\n",
       " ('CTAG', (0.9973, 0.9898, 0.9994, 1.0036, 0.9992)),\n",
       " ('TCCC', (0.9832, 0.969, 0.9625, 0.9393, 0.9989)),\n",
       " ('TGGG', (0.9742, 1.0448, 0.9727, 0.9882, 0.9984)),\n",
       " ('GCCT', (0.9664, 0.981, 0.9485, 0.9519, 0.9982)),\n",
       " ('TAGG', (0.9967, 0.99, 0.999, 1.0065, 0.998)),\n",
       " ('CCTC', (0.965, 0.9694, 0.9539, 0.9425, 0.9972)),\n",
       " ('AGCT', (0.969, 0.9833, 0.9682, 0.9747, 0.9968)),\n",
       " ('TCCA', (0.99, 0.9886, 0.9894, 0.9759, 0.9966)),\n",
       " ('GCAA', (0.9779, 0.9875, 0.9936, 1.0075, 0.996)),\n",
       " ('GAAT', (1.0336, 1.0419, 1.035, 1.0316, 0.9957)),\n",
       " ('AATC', (1.0399, 1.0201, 1.0453, 1.0362, 0.9949)),\n",
       " ('ATGA', (1.0109, 1.0161, 1.0165, 1.0229, 0.9929)),\n",
       " ('ATGT', (1.0029, 1.0012, 0.9995, 0.9976, 0.9928)),\n",
       " ('AGTT', (1.0075, 1.0325, 1.002, 0.9913, 0.9928)),\n",
       " ('AAGT', (1.0238, 1.0292, 1.0327, 1.0358, 0.9919)),\n",
       " ('TTGA', (1.0253, 1.0847, 1.0026, 0.993, 0.9917)),\n",
       " ('ATTG', (1.022, 1.045, 1.012, 1.0069, 0.9913)),\n",
       " ('ATGC', (0.9738, 0.9783, 0.9702, 0.975, 0.9909)),\n",
       " ('GATA', (1.0264, 1.0012, 1.0282, 1.0259, 0.9907)),\n",
       " ('AAGC', (0.982, 0.9986, 0.9952, 1.0069, 0.9903)),\n",
       " ('AAGG', (0.9985, 1.0341, 1.0088, 1.0256, 0.9901)),\n",
       " ('GTCT', (0.9912, 1.0141, 0.986, 0.9772, 0.9885)),\n",
       " ('GGGA', (0.9695, 1.0383, 0.979, 0.9904, 0.9882)),\n",
       " ('CCCA', (0.9719, 0.959, 0.9709, 0.9541, 0.9881)),\n",
       " ('CCAC', (0.9568, 0.9444, 0.9598, 0.9635, 0.9878)),\n",
       " ('TGCA', (0.9655, 0.9821, 0.9672, 0.9698, 0.9873)),\n",
       " ('ATGG', (0.9878, 1.0098, 0.9873, 1.005, 0.9869)),\n",
       " ('CTTG', (0.9853, 1.0381, 0.9606, 0.9438, 0.9867)),\n",
       " ('AGCA', (0.9682, 0.9692, 0.9864, 0.9939, 0.9867)),\n",
       " ('GATT', (1.0308, 1.0643, 1.0204, 1.0101, 0.9866)),\n",
       " ('CTGT', (0.9792, 1.0006, 0.9709, 0.9653, 0.9861)),\n",
       " ('GTAC', (0.9862, 0.9808, 0.9842, 0.9828, 0.9858)),\n",
       " ('TGAT', (1.0198, 1.0514, 1.0094, 1.0029, 0.9857)),\n",
       " ('GCAT', (0.9622, 0.9608, 0.9665, 0.9715, 0.9856)),\n",
       " ('GCTC', (0.9389, 0.9653, 0.9406, 0.9386, 0.9853)),\n",
       " ('TGCC', (0.9561, 0.9778, 0.9421, 0.9477, 0.9852)),\n",
       " ('GGCC', (0.9406, 0.9802, 0.9463, 0.9801, 0.9851)),\n",
       " ('GGAA', (1.0055, 1.0581, 1.0102, 1.0215, 0.9851)),\n",
       " ('TCAA', (1.0296, 1.0169, 1.0384, 1.0318, 0.9846)),\n",
       " ('AGAT', (1.0302, 1.0371, 1.0329, 1.032, 0.9845)),\n",
       " ('CTGG', (0.9723, 1.0232, 0.9701, 0.9803, 0.9843)),\n",
       " ('ACTG', (0.9835, 0.9877, 0.9873, 0.9931, 0.9843)),\n",
       " ('CGCT', (0.9386, 0.9592, 0.9393, 0.9449, 0.984)),\n",
       " ('TGGT', (0.9878, 1.043, 0.9747, 0.9778, 0.9832)),\n",
       " ('ATCC', (0.9941, 0.9797, 0.9884, 0.9717, 0.9832)),\n",
       " ('CTGC', (0.9454, 0.9705, 0.9458, 0.9574, 0.983)),\n",
       " ('TCAT', (1.0056, 0.977, 1.0012, 0.9797, 0.9822)),\n",
       " ('CTCA', (0.9755, 0.97, 0.9822, 0.9724, 0.9821)),\n",
       " ('CACA', (0.9764, 0.9442, 0.9947, 0.9938, 0.9814)),\n",
       " ('CATG', (0.968, 0.9762, 0.9703, 0.9756, 0.981)),\n",
       " ('CAAC', (0.9725, 0.9454, 0.9896, 0.9831, 0.9808)),\n",
       " ('GGCT', (0.9504, 0.9851, 0.9503, 0.9693, 0.9808)),\n",
       " ('GACT', (0.9841, 0.9852, 0.9831, 0.9811, 0.9798)),\n",
       " ('GTAG', (0.969, 0.988, 0.9739, 0.9747, 0.9791)),\n",
       " ('GTGC', (0.9386, 0.9893, 0.9348, 0.9541, 0.9789)),\n",
       " ('GGTG', (0.9532, 1.0298, 0.9454, 0.9733, 0.9787)),\n",
       " ('GCGC', (0.9101, 0.9562, 0.9218, 0.9496, 0.9777)),\n",
       " ('AGGC', (0.9541, 0.9899, 0.9628, 0.9875, 0.9774)),\n",
       " ('TACG', (0.9709, 0.9434, 0.9711, 0.979, 0.9755)),\n",
       " ('TGGC', (0.9515, 1.0014, 0.9491, 0.9779, 0.9753)),\n",
       " ('CCAG', (0.9664, 0.9855, 0.9681, 0.9738, 0.9751)),\n",
       " ('AGCC', (0.9521, 0.9576, 0.953, 0.9568, 0.975)),\n",
       " ('CAAG', (0.9834, 0.9956, 0.996, 0.9997, 0.9747)),\n",
       " ('AGAG', (0.9895, 1.0357, 1.0016, 1.0097, 0.9746)),\n",
       " ('GGTT', (0.9867, 1.0523, 0.9782, 0.9755, 0.9743)),\n",
       " ('AGGT', (0.9791, 1.0203, 0.9794, 0.9867, 0.9739)),\n",
       " ('TCTG', (0.9956, 1.034, 0.9845, 0.9758, 0.9734)),\n",
       " ('GCTG', (0.9439, 0.9961, 0.9462, 0.9693, 0.9719)),\n",
       " ('AGGA', (0.9878, 1.0071, 1.0001, 1.0064, 0.9714)),\n",
       " ('GGGC', (0.9386, 0.9709, 0.9474, 0.9722, 0.9713)),\n",
       " ('GTGG', (0.9522, 1.0167, 0.9479, 0.9735, 0.9702)),\n",
       " ('CAGC', (0.9279, 0.9443, 0.943, 0.9537, 0.9697)),\n",
       " ('GTGT', (0.9553, 1.0046, 0.9497, 0.9496, 0.9692)),\n",
       " ('TGTG', (0.9605, 1.0145, 0.9513, 0.9548, 0.9691)),\n",
       " ('GAAC', (0.9815, 0.9846, 0.9879, 0.9831, 0.9691)),\n",
       " ('ACAC', (0.9756, 0.9306, 0.9846, 0.9674, 0.9686)),\n",
       " ('GCAG', (0.9381, 0.9779, 0.9484, 0.973, 0.9686)),\n",
       " ('ACAG', (0.9756, 0.9581, 0.9973, 1.0045, 0.9674)),\n",
       " ('GGCA', (0.9429, 0.9717, 0.9502, 0.9666, 0.9674)),\n",
       " ('CAGA', (0.9906, 1.0051, 1.0021, 1.0067, 0.9673)),\n",
       " ('TCAC', (0.9763, 0.9597, 0.9773, 0.9672, 0.967)),\n",
       " ('CAGG', (0.9574, 0.996, 0.9679, 0.9886, 0.967)),\n",
       " ('AGAC', (0.9823, 0.9735, 0.9914, 0.9921, 0.9648)),\n",
       " ('GGAT', (0.9947, 1.0054, 0.9938, 0.994, 0.9647)),\n",
       " ('CGTA', (0.978, 0.9458, 0.978, 0.9747, 0.9644)),\n",
       " ('GAAG', (0.9799, 1.0539, 0.9896, 0.9991, 0.9643)),\n",
       " ('CGCA', (0.9292, 0.9382, 0.9416, 0.9541, 0.9639)),\n",
       " ('CTGA', (0.9899, 1.0231, 0.9848, 0.9811, 0.9636)),\n",
       " ('GCAC', (0.9433, 0.9493, 0.9471, 0.9486, 0.9636)),\n",
       " ('TTGG', (0.9922, 1.0517, 0.9757, 0.9676, 0.9635)),\n",
       " ('CCTG', (0.9672, 0.9901, 0.9547, 0.9471, 0.9629)),\n",
       " ('AGTG', (0.9597, 0.999, 0.9657, 0.9732, 0.9612)),\n",
       " ('ACGC', (0.9276, 0.9271, 0.9369, 0.9441, 0.9612)),\n",
       " ('GCCC', (0.9426, 0.9532, 0.9383, 0.9292, 0.9607)),\n",
       " ('TGCG', (0.9314, 0.9843, 0.9351, 0.9546, 0.9603)),\n",
       " ('TGGA', (0.9891, 1.0215, 0.9853, 0.9869, 0.9599)),\n",
       " ('ATCA', (1.0048, 0.9667, 1.0153, 1.0021, 0.9587)),\n",
       " ('GTCC', (0.9591, 0.9587, 0.9545, 0.9492, 0.9587)),\n",
       " ('TGAG', (0.9683, 1.0274, 0.9682, 0.9666, 0.9585)),\n",
       " ('CCGC', (0.9245, 0.9348, 0.9247, 0.9238, 0.9578)),\n",
       " ('CAGT', (0.9568, 0.955, 0.9627, 0.9606, 0.9578)),\n",
       " ('ACCC', (0.9772, 0.9426, 0.9672, 0.9269, 0.9571)),\n",
       " ('AGCG', (0.9376, 0.9693, 0.9547, 0.971, 0.9561)),\n",
       " ('GAGG', (0.948, 1.0168, 0.9542, 0.9682, 0.9559)),\n",
       " ('AACG', (0.9705, 0.9717, 0.9843, 0.9934, 0.9551)),\n",
       " ('GATC', (0.9983, 1.0075, 1.0029, 0.9995, 0.955)),\n",
       " ('GTTC', (0.972, 1.013, 0.964, 0.9475, 0.9537)),\n",
       " ('GATG', (0.9648, 1.009, 0.9577, 0.9677, 0.9534)),\n",
       " ('GACC', (0.9553, 0.9489, 0.9555, 0.9483, 0.9521)),\n",
       " ('CATC', (0.9568, 0.9464, 0.9575, 0.9322, 0.9518)),\n",
       " ('ACGT', (0.9578, 0.9408, 0.9612, 0.9637, 0.9511)),\n",
       " ('CGAA', (0.9823, 1.0085, 0.9906, 0.9907, 0.9507)),\n",
       " ('CGTT', (0.9609, 0.9725, 0.9476, 0.9366, 0.9503)),\n",
       " ('TGTC', (0.9674, 0.9856, 0.9626, 0.9528, 0.9493)),\n",
       " ('GTGA', (0.9626, 1.0157, 0.9616, 0.9701, 0.9493)),\n",
       " ('ACCG', (0.9635, 0.954, 0.9639, 0.9592, 0.9491)),\n",
       " ('CACG', (0.9367, 0.9415, 0.9429, 0.9602, 0.9489)),\n",
       " ('GCGT', (0.929, 0.9602, 0.9339, 0.9421, 0.947)),\n",
       " ('GGAG', (0.9508, 1.0195, 0.959, 0.9668, 0.947)),\n",
       " ('GCGA', (0.9307, 0.9821, 0.9425, 0.9585, 0.9466)),\n",
       " ('GACA', (0.9671, 0.9443, 0.9738, 0.9649, 0.9458)),\n",
       " ('TGAC', (0.967, 0.9755, 0.9554, 0.9535, 0.945)),\n",
       " ('CGAT', (0.9717, 0.987, 0.9711, 0.97, 0.9446)),\n",
       " ('GGAC', (0.954, 0.9375, 0.9602, 0.9623, 0.9441)),\n",
       " ('GAGC', (0.9288, 0.9649, 0.936, 0.9389, 0.9427)),\n",
       " ('GTTG', (0.9593, 1.0249, 0.9517, 0.9468, 0.9424)),\n",
       " ('CACC', (0.9523, 0.9256, 0.9429, 0.9031, 0.9416)),\n",
       " ('GCCG', (0.9287, 0.9514, 0.9322, 0.95, 0.9413)),\n",
       " ('GAGT', (0.9589, 1.0055, 0.9666, 0.9566, 0.941)),\n",
       " ('GAGA', (0.9761, 1.0325, 0.9821, 0.9759, 0.9409)),\n",
       " ('ACGG', (0.9427, 0.9572, 0.9489, 0.9659, 0.9407)),\n",
       " ('ACGA', (0.9602, 0.9634, 0.9698, 0.9782, 0.9398)),\n",
       " ('CGGG', (0.9408, 0.9793, 0.9452, 0.9546, 0.9393)),\n",
       " ('GTCA', (0.9733, 0.9718, 0.9777, 0.9758, 0.939)),\n",
       " ('GCGG', (0.9198, 0.9729, 0.9331, 0.9632, 0.9384)),\n",
       " ('CCGT', (0.954, 0.9389, 0.9433, 0.9296, 0.9382)),\n",
       " ('AGTC', (0.9723, 0.9638, 0.9773, 0.9657, 0.9369)),\n",
       " ('GGTC', (0.9603, 0.9952, 0.96, 0.9615, 0.9359)),\n",
       " ('TTCG', (0.9743, 1.0011, 0.9653, 0.9544, 0.9344)),\n",
       " ('TCCG', (0.9455, 0.9616, 0.9406, 0.9252, 0.9327)),\n",
       " ('CGCG', (0.9028, 0.9421, 0.9137, 0.9305, 0.9318)),\n",
       " ('GGCG', (0.9079, 0.9609, 0.919, 0.9423, 0.9315)),\n",
       " ('CGTG', (0.9262, 0.964, 0.9225, 0.9331, 0.9308)),\n",
       " ('CGGT', (0.9477, 0.9877, 0.9456, 0.9533, 0.9301)),\n",
       " ('CGCC', (0.9154, 0.9161, 0.9136, 0.9034, 0.9288)),\n",
       " ('TCAG', (0.9584, 0.9751, 0.9634, 0.9554, 0.928)),\n",
       " ('CCGG', (0.9502, 0.9734, 0.9488, 0.9475, 0.9255)),\n",
       " ('ATCG', (0.9737, 0.9577, 0.9775, 0.9688, 0.9205)),\n",
       " ('TCGC', (0.9292, 0.9442, 0.9322, 0.9272, 0.9204)),\n",
       " ('CGGA', (0.9593, 0.9678, 0.9624, 0.9683, 0.9202)),\n",
       " ('TCGA', (0.9661, 0.9822, 0.9637, 0.9542, 0.9192)),\n",
       " ('CGGC', (0.8986, 0.9343, 0.9113, 0.9264, 0.919)),\n",
       " ('CGAC', (0.9222, 0.9235, 0.931, 0.9269, 0.9183)),\n",
       " ('CCGA', (0.946, 0.952, 0.9478, 0.9167, 0.9138)),\n",
       " ('CTCG', (0.9332, 0.958, 0.9274, 0.9153, 0.911)),\n",
       " ('CCCG', (0.941, 0.932, 0.9331, 0.8865, 0.9051)),\n",
       " ('GACG', (0.9128, 0.9331, 0.9189, 0.9218, 0.9033)),\n",
       " ('TCGT', (0.9579, 0.9496, 0.948, 0.9331, 0.8969)),\n",
       " ('TCGG', (0.9434, 0.9667, 0.9405, 0.9331, 0.8898)),\n",
       " ('CGAG', (0.9099, 0.9385, 0.9129, 0.8956, 0.8841)),\n",
       " ('CGTC', (0.9235, 0.9012, 0.9271, 0.9114, 0.8757)),\n",
       " ('GTCG', (0.9278, 0.9291, 0.9285, 0.9246, 0.8716))]"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "neg_four_mers"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "47bc768a",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "<__array_function__ internals>:180: VisibleDeprecationWarning: Creating an ndarray from ragged nested sequences (which is a list-or-tuple of lists-or-tuples-or ndarrays with different lengths or shapes) is deprecated. If you meant to do this, you must specify 'dtype=object' when creating the ndarray.\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "(5376, 2)"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "in_vitro_neg_seqs = np.concatenate((neg_four_mers,neg_five_mers,neg_six_mers))\n",
    "in_vitro_neg_seqs.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "88397cff",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "5376"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "unique_rbns = {}\n",
    "for s, arr in in_vitro_neg_seqs:\n",
    "    if s not in unique_rbns:\n",
    "        unique_rbns[s] = arr\n",
    "\n",
    "in_vitro_neg_seqs = list(unique_rbns.items())\n",
    "len(in_vitro_neg_seqs)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "94988d48",
   "metadata": {},
   "source": [
    "Construct placeholder expression change vector"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "950281ea",
   "metadata": {},
   "outputs": [],
   "source": [
    "placeholder_logfold_changes = np.zeros(shape=logfold_changes.shape)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b14cf59a",
   "metadata": {},
   "source": [
    "## Split into train/val/test"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "e9d8ef74",
   "metadata": {},
   "outputs": [],
   "source": [
    "# number of total samples per class\n",
    "combinations_size = 100000"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a99c5d06",
   "metadata": {},
   "source": [
    "Positive samples"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "c5fdbf5e",
   "metadata": {},
   "outputs": [],
   "source": [
    "from itertools import product\n",
    "\n",
    "random.shuffle(in_vivo_pos_seqs)\n",
    "random.shuffle(in_vitro_pos_seqs)\n",
    "\n",
    "n_eclip = len(in_vivo_pos_seqs)\n",
    "n_rbns  = len(in_vitro_pos_seqs)\n",
    "\n",
    "# Split indices\n",
    "train_eclip = in_vivo_pos_seqs[:int(0.7*n_eclip)]\n",
    "val_eclip   = in_vivo_pos_seqs[int(0.7*n_eclip):int(0.85*n_eclip)]\n",
    "test_eclip  = in_vivo_pos_seqs[int(0.85*n_eclip):]\n",
    "\n",
    "train_rbns = in_vitro_pos_seqs[:int(0.7*n_rbns)]\n",
    "val_rbns   = in_vitro_pos_seqs[int(0.7*n_rbns):int(0.85*n_rbns)]\n",
    "test_rbns  = in_vitro_pos_seqs[int(0.85*n_rbns):]\n",
    "\n",
    "pos_train_samples = [\n",
    "    (random.choice(train_eclip), random.choice(train_rbns))\n",
    "    for _ in range(combinations_size)\n",
    "]\n",
    "pos_val_samples = [\n",
    "    (random.choice(val_eclip), random.choice(val_rbns))\n",
    "    for _ in range(combinations_size)\n",
    "]\n",
    "pos_test_samples = [\n",
    "    (random.choice(test_eclip), random.choice(test_rbns))\n",
    "    for _ in range(combinations_size)\n",
    "]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "91baf0a1",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "eCLIP sequences are disjoint across splits\n"
     ]
    }
   ],
   "source": [
    "train_eclips = set(eclip for eclip, _ in pos_train_samples)\n",
    "val_eclips   = set(eclip for eclip, _ in pos_val_samples)\n",
    "test_eclips  = set(eclip for eclip, _ in pos_test_samples)\n",
    "\n",
    "assert train_eclips.isdisjoint(val_eclips), \"Overlap between train and val eCLIP sequences!\"\n",
    "assert train_eclips.isdisjoint(test_eclips), \"Overlap between train and test eCLIP sequences!\"\n",
    "assert val_eclips.isdisjoint(test_eclips), \"Overlap between val and test eCLIP sequences!\"\n",
    "print(\"eCLIP sequences are disjoint across splits\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "0ffaaf81",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "RBNS sequences are disjoint across splits\n"
     ]
    }
   ],
   "source": [
    "train_rbns = set(rbns for _, (rbns, _) in pos_train_samples)\n",
    "val_rbns   = set(rbns for _, (rbns, _) in pos_val_samples)\n",
    "test_rbns  = set(rbns for _, (rbns, _) in pos_test_samples)\n",
    "\n",
    "assert train_rbns.isdisjoint(val_rbns), \"Overlap between train and val RBNS sequences!\"\n",
    "assert train_rbns.isdisjoint(test_rbns), \"Overlap between train and test RBNS sequences!\"\n",
    "assert val_rbns.isdisjoint(test_rbns), \"Overlap between val and test RBNS sequences!\"\n",
    "print(\"RBNS sequences are disjoint across splits\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "330d4bc3",
   "metadata": {},
   "outputs": [],
   "source": [
    "pos_train_label = np.ones(len(pos_train_samples))\n",
    "pos_val_label = np.ones(len(pos_val_samples))\n",
    "pos_test_label = np.ones(len(pos_test_samples))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c3da2c3a",
   "metadata": {},
   "source": [
    "Negative samples"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "d7c965b7",
   "metadata": {},
   "outputs": [],
   "source": [
    "random.shuffle(in_vivo_neg_seqs)\n",
    "random.shuffle(in_vitro_neg_seqs)\n",
    "\n",
    "n_eclip = len(in_vivo_neg_seqs)\n",
    "n_rbns  = len(in_vitro_neg_seqs)\n",
    "\n",
    "# Split indices\n",
    "train_eclip = in_vivo_neg_seqs[:int(0.7*n_eclip)]\n",
    "val_eclip   = in_vivo_neg_seqs[int(0.7*n_eclip):int(0.85*n_eclip)]\n",
    "test_eclip  = in_vivo_neg_seqs[int(0.85*n_eclip):]\n",
    "\n",
    "train_rbns = in_vitro_neg_seqs[:int(0.7*n_rbns)]\n",
    "val_rbns   = in_vitro_neg_seqs[int(0.7*n_rbns):int(0.85*n_rbns)]\n",
    "test_rbns  = in_vitro_neg_seqs[int(0.85*n_rbns):]\n",
    "\n",
    "neg_train_samples = [\n",
    "    (random.choice(train_eclip), random.choice(train_rbns))\n",
    "    for _ in range(combinations_size)\n",
    "]\n",
    "neg_val_samples = [\n",
    "    (random.choice(val_eclip), random.choice(val_rbns))\n",
    "    for _ in range(combinations_size)\n",
    "]\n",
    "neg_test_samples = [\n",
    "    (random.choice(test_eclip), random.choice(test_rbns))\n",
    "    for _ in range(combinations_size)\n",
    "]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "19d1f6e3",
   "metadata": {},
   "outputs": [],
   "source": [
    "neg_train_label = np.zeros(len(neg_train_samples))\n",
    "neg_val_label = np.zeros(len(neg_val_samples))\n",
    "neg_test_label = np.zeros(len(neg_test_samples))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9cc54a9e",
   "metadata": {},
   "source": [
    "Combine for full splits"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "726d4838",
   "metadata": {},
   "outputs": [],
   "source": [
    "train_data = np.concatenate((pos_train_samples, neg_train_samples))\n",
    "val_data = np.concatenate((pos_val_samples, neg_val_samples))\n",
    "test_data = np.concatenate((pos_test_samples, neg_test_samples))\n",
    "\n",
    "train_labels = np.concatenate((pos_train_label, neg_train_label))\n",
    "val_labels = np.concatenate((pos_val_label, neg_val_label))\n",
    "test_labels = np.concatenate((pos_test_label, neg_test_label))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "8ce65d9c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# get positive/negative label values\n",
    "positive_value = (\"TIA1\", logfold_changes)\n",
    "negative_value = (\"Not TIA1\", placeholder_logfold_changes)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "877f0473",
   "metadata": {},
   "source": [
    "## Save as dictionaries"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "16dcdac6",
   "metadata": {},
   "source": [
    "Save as pickle files since there are nested data structures"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "eff9d82c",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pickle"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "cfda55f9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'\\nSaved data format for each split:\\n- Input\\n    - List containing:\\n        - eCLIP sequence: text\\n        - Tuple of RBNS sequence and binding affinity: (text, vector)\\n- Output\\n    - List containing:\\n        - RBP label: text\\n        - Logfold gene expression change: vector\\n        \\nLabels save separately to save storage:\\n- List containing:\\n    - RBP label: text\\n    - Logfold gene expression change: vector\\n'"
      ]
     },
     "execution_count": 37,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "'''\n",
    "Saved data format for each split:\n",
    "- Input\n",
    "    - List containing:\n",
    "        - eCLIP sequence: text\n",
    "        - Tuple of RBNS sequence and binding affinity: (text, vector)\n",
    "- Output\n",
    "    - List containing:\n",
    "        - RBP label: text\n",
    "        - Logfold gene expression change: vector\n",
    "        \n",
    "Labels save separately to save storage:\n",
    "- List containing:\n",
    "    - RBP label: text\n",
    "    - Logfold gene expression change: vector\n",
    "'''"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "id": "6c84bd69",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Save train\n",
    "with open(\"TIA1/train_split.pkl\", \"wb\") as f:\n",
    "    pickle.dump((train_data, train_labels), f)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "id": "afb7be9c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Save val\n",
    "with open(\"TIA1/val_split.pkl\", \"wb\") as f:\n",
    "    pickle.dump((val_data, val_labels), f)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "6266df79",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Save test\n",
    "with open(\"TIA1/test_split.pkl\", \"wb\") as f:\n",
    "    pickle.dump((test_data, test_labels), f)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "id": "95900830",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Save positive label\n",
    "with open(\"TIA1/TIA1_positive_label.pkl\", \"wb\") as f:\n",
    "    pickle.dump(positive_value, f)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "id": "fe972c58",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Save negative label\n",
    "with open(\"TIA1/TIA1_negative_label.pkl\", \"wb\") as f:\n",
    "    pickle.dump(negative_value, f)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "521fb44e",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "perturb",
   "language": "python",
   "name": "perturb"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
