{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "3ee91ff5",
   "metadata": {},
   "outputs": [],
   "source": [
    "%load_ext autoreload\n",
    "%autoreload 2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "15865688",
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ff5d5753",
   "metadata": {},
   "source": [
    "## Read data"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "235e3bf2",
   "metadata": {},
   "source": [
    "Load eCLIP-seq sequences"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "abe42208",
   "metadata": {},
   "outputs": [],
   "source": [
    "eclip = pd.read_csv(\"TRA2A/TRA2A_eclip.txt\", sep='\\t', header=None)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "32e0bc25",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr7:94285693-94285822</td>\n",
       "      <td>ACTTCGGTTGGTGTGTGTCGAAGAAACCTGACTGCGCCCTGAGGAG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr7:94292667-94292751</td>\n",
       "      <td>AAAAAAAGAAGGAGTGGAAGAGGGGGCCAGGATCCAGGCCTCCATC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr7:100780742-100780921</td>\n",
       "      <td>CATCTGGGACAAAACTGGAGATGCATCGGGAAAGAAGAAACTCCGA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr7:11022589-11022767</td>\n",
       "      <td>ACAATGAGGATGATGAAGATGAGGGAAGCGGGAGTGATGAAGACGA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr7:155493474-155493644</td>\n",
       "      <td>GGAAACATTGGAACTTCAAAAGGACATCAAAGAAGAATCAGATGAA...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                          0                                                  1\n",
       "0    chr7:94285693-94285822  ACTTCGGTTGGTGTGTGTCGAAGAAACCTGACTGCGCCCTGAGGAG...\n",
       "1    chr7:94292667-94292751  AAAAAAAGAAGGAGTGGAAGAGGGGGCCAGGATCCAGGCCTCCATC...\n",
       "2  chr7:100780742-100780921  CATCTGGGACAAAACTGGAGATGCATCGGGAAAGAAGAAACTCCGA...\n",
       "3    chr7:11022589-11022767  ACAATGAGGATGATGAAGATGAGGGAAGCGGGAGTGATGAAGACGA...\n",
       "4  chr7:155493474-155493644  GGAAACATTGGAACTTCAAAAGGACATCAAAGAAGAATCAGATGAA..."
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "eclip.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8e72135a",
   "metadata": {},
   "source": [
    "Load RBNS binding affinities"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "7bc981f7",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>[TRA2A]</th>\n",
       "      <th>0 nM</th>\n",
       "      <th>5 nM</th>\n",
       "      <th>20 nM</th>\n",
       "      <th>80 nM</th>\n",
       "      <th>320 nM</th>\n",
       "      <th>1300 nM</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AGAA</td>\n",
       "      <td>1.0179</td>\n",
       "      <td>1.9157</td>\n",
       "      <td>2.4378</td>\n",
       "      <td>2.8430</td>\n",
       "      <td>2.7831</td>\n",
       "      <td>2.5174</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GAAG</td>\n",
       "      <td>0.9857</td>\n",
       "      <td>2.0081</td>\n",
       "      <td>2.6597</td>\n",
       "      <td>2.8270</td>\n",
       "      <td>2.6370</td>\n",
       "      <td>2.2785</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AAGA</td>\n",
       "      <td>1.0130</td>\n",
       "      <td>1.8014</td>\n",
       "      <td>2.2892</td>\n",
       "      <td>2.5155</td>\n",
       "      <td>2.3421</td>\n",
       "      <td>2.0420</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>GAAA</td>\n",
       "      <td>0.9935</td>\n",
       "      <td>1.5545</td>\n",
       "      <td>1.8787</td>\n",
       "      <td>2.1811</td>\n",
       "      <td>2.2963</td>\n",
       "      <td>2.1849</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>TGAA</td>\n",
       "      <td>1.0309</td>\n",
       "      <td>1.5297</td>\n",
       "      <td>1.8455</td>\n",
       "      <td>2.0781</td>\n",
       "      <td>2.1111</td>\n",
       "      <td>1.9834</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>251</th>\n",
       "      <td>CGGG</td>\n",
       "      <td>0.9202</td>\n",
       "      <td>0.8796</td>\n",
       "      <td>0.8297</td>\n",
       "      <td>0.6913</td>\n",
       "      <td>0.6943</td>\n",
       "      <td>0.6643</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>252</th>\n",
       "      <td>CGTT</td>\n",
       "      <td>1.0362</td>\n",
       "      <td>0.8679</td>\n",
       "      <td>0.7962</td>\n",
       "      <td>0.6736</td>\n",
       "      <td>0.6339</td>\n",
       "      <td>0.6104</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>253</th>\n",
       "      <td>GGGT</td>\n",
       "      <td>0.9313</td>\n",
       "      <td>0.8662</td>\n",
       "      <td>0.7994</td>\n",
       "      <td>0.6708</td>\n",
       "      <td>0.6366</td>\n",
       "      <td>0.6112</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>254</th>\n",
       "      <td>GGTC</td>\n",
       "      <td>0.9870</td>\n",
       "      <td>0.8717</td>\n",
       "      <td>0.8178</td>\n",
       "      <td>0.6655</td>\n",
       "      <td>0.6135</td>\n",
       "      <td>0.5582</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>255</th>\n",
       "      <td>TCGT</td>\n",
       "      <td>1.0746</td>\n",
       "      <td>0.8632</td>\n",
       "      <td>0.7969</td>\n",
       "      <td>0.6320</td>\n",
       "      <td>0.5554</td>\n",
       "      <td>0.4824</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>256 rows Ã— 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    [TRA2A]    0 nM    5 nM   20 nM   80 nM  320 nM  1300 nM\n",
       "0      AGAA  1.0179  1.9157  2.4378  2.8430  2.7831   2.5174\n",
       "1      GAAG  0.9857  2.0081  2.6597  2.8270  2.6370   2.2785\n",
       "2      AAGA  1.0130  1.8014  2.2892  2.5155  2.3421   2.0420\n",
       "3      GAAA  0.9935  1.5545  1.8787  2.1811  2.2963   2.1849\n",
       "4      TGAA  1.0309  1.5297  1.8455  2.0781  2.1111   1.9834\n",
       "..      ...     ...     ...     ...     ...     ...      ...\n",
       "251    CGGG  0.9202  0.8796  0.8297  0.6913  0.6943   0.6643\n",
       "252    CGTT  1.0362  0.8679  0.7962  0.6736  0.6339   0.6104\n",
       "253    GGGT  0.9313  0.8662  0.7994  0.6708  0.6366   0.6112\n",
       "254    GGTC  0.9870  0.8717  0.8178  0.6655  0.6135   0.5582\n",
       "255    TCGT  1.0746  0.8632  0.7969  0.6320  0.5554   0.4824\n",
       "\n",
       "[256 rows x 7 columns]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "four_mers = pd.read_csv(\"TRA2A/TRA2A_4.tsv\", sep='\\t')\n",
    "five_mers = pd.read_csv(\"TRA2A/TRA2A_5.tsv\", sep='\\t',)\n",
    "six_mers = pd.read_csv(\"TRA2A/TRA2A_6.tsv\", sep='\\t',)\n",
    "four_mers"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "af6ffd4e",
   "metadata": {},
   "source": [
    "Load knockdown differential rna-seq data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "d0e331a6",
   "metadata": {},
   "outputs": [],
   "source": [
    "# get other diff expr for subsetting to shared genes later\n",
    "diff_exp_HNRNPC = pd.read_csv(\"HNRNPC/HNRNPC.tsv\", sep='\\t',)\n",
    "diff_exp_SRSF9 = pd.read_csv(\"SRSF9/SRSF9.tsv\", sep='\\t',)\n",
    "diff_exp_PCBP1 = pd.read_csv(\"PCBP1/PCBP1.tsv\", sep='\\t',)\n",
    "diff_exp_TIA1 = pd.read_csv(\"TIA1/TIA1.tsv\", sep='\\t',)\n",
    "diff_exp_TRA2A = pd.read_csv(\"TRA2A/TRA2A.tsv\", sep='\\t',)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1612ca75",
   "metadata": {},
   "source": [
    "## Generate positive pairs"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b0787ccc",
   "metadata": {},
   "source": [
    "EClip positive binding sites"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "592b1f36",
   "metadata": {},
   "outputs": [],
   "source": [
    "in_vivo_pos_seqs = eclip[1].values"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "18057862",
   "metadata": {},
   "outputs": [],
   "source": [
    "# filter for length and unique values\n",
    "in_vivo_pos_seqs = [s for s in in_vivo_pos_seqs if len(s) >= 10]\n",
    "in_vivo_pos_seqs = list(set(in_vivo_pos_seqs))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "adf61072",
   "metadata": {},
   "source": [
    "RBNS significant binding affinity"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "fd120254",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_signficant_sequences(df):\n",
    "    num_cols = df.columns[1:]\n",
    "    filtered = df[df[num_cols].gt(2).any(axis=1)]\n",
    "    pairs = [\n",
    "        (row[df.columns[0]], tuple(row[num_cols]))\n",
    "        for _, row in filtered.iterrows()\n",
    "    ]\n",
    "    return pairs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "ed65dc0f",
   "metadata": {},
   "outputs": [],
   "source": [
    "pos_four_mers = get_signficant_sequences(four_mers)\n",
    "pos_five_mers = get_signficant_sequences(five_mers)\n",
    "pos_six_mers = get_signficant_sequences(six_mers)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "1b63d4fe",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(6, 25, 129)"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(pos_four_mers), len(pos_five_mers), len(pos_six_mers), "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "6894c643",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "<__array_function__ internals>:180: VisibleDeprecationWarning: Creating an ndarray from ragged nested sequences (which is a list-or-tuple of lists-or-tuples-or ndarrays with different lengths or shapes) is deprecated. If you meant to do this, you must specify 'dtype=object' when creating the ndarray.\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "(160, 2)"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "in_vitro_pos_seqs = np.concatenate((pos_four_mers, pos_five_mers,pos_six_mers))\n",
    "in_vitro_pos_seqs.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "3bc23f33",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "160"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "unique_rbns = {}\n",
    "for s, arr in in_vitro_pos_seqs:\n",
    "    if s not in unique_rbns:\n",
    "        unique_rbns[s] = arr\n",
    "\n",
    "in_vitro_pos_seqs = list(unique_rbns.items())\n",
    "len(in_vitro_pos_seqs)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6ad2eb50",
   "metadata": {},
   "source": [
    "Filter diff expr"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "7a6b7538",
   "metadata": {},
   "outputs": [],
   "source": [
    "# remove when gene expression is always 0 or if logfold change is inf/-inf\n",
    "diff_exp_filtered_HNRNPC = diff_exp_HNRNPC[\n",
    "    ~(\n",
    "        ((diff_exp_HNRNPC[\"value_1\"] == 0) & (diff_exp_HNRNPC[\"value_2\"] == 0)) |\n",
    "        (diff_exp_HNRNPC[\"log2(fold_change)\"].isin([np.inf, -np.inf]))\n",
    "    )\n",
    "]\n",
    "\n",
    "diff_exp_filtered_SRSF9 = diff_exp_SRSF9[\n",
    "    ~(\n",
    "        ((diff_exp_SRSF9[\"value_1\"] == 0) & (diff_exp_SRSF9[\"value_2\"] == 0)) |\n",
    "        (diff_exp_SRSF9[\"log2(fold_change)\"].isin([np.inf, -np.inf]))\n",
    "    )\n",
    "]\n",
    "\n",
    "diff_exp_filtered_PCBP1 = diff_exp_PCBP1[\n",
    "    ~(\n",
    "        ((diff_exp_PCBP1[\"value_1\"] == 0) & (diff_exp_PCBP1[\"value_2\"] == 0)) |\n",
    "        (diff_exp_PCBP1[\"log2(fold_change)\"].isin([np.inf, -np.inf]))\n",
    "    )\n",
    "]\n",
    "\n",
    "diff_exp_filtered_TIA1 = diff_exp_TIA1[\n",
    "    ~(\n",
    "        ((diff_exp_TIA1[\"value_1\"] == 0) & (diff_exp_TIA1[\"value_2\"] == 0)) |\n",
    "        (diff_exp_TIA1[\"log2(fold_change)\"].isin([np.inf, -np.inf]))\n",
    "    )\n",
    "]\n",
    "\n",
    "diff_exp_filtered_TRA2A = diff_exp_TRA2A[\n",
    "    ~(\n",
    "        ((diff_exp_TRA2A[\"value_1\"] == 0) & (diff_exp_TRA2A[\"value_2\"] == 0)) |\n",
    "        (diff_exp_TRA2A[\"log2(fold_change)\"].isin([np.inf, -np.inf]))\n",
    "    )\n",
    "]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "377252f2",
   "metadata": {},
   "outputs": [],
   "source": [
    "dfs = [diff_exp_filtered_HNRNPC, diff_exp_filtered_SRSF9, diff_exp_filtered_PCBP1, diff_exp_filtered_TIA1, diff_exp_filtered_TRA2A]\n",
    "col = \"gene_id\"\n",
    "shared_keys = set.intersection(*(set(df[col]) for df in dfs))\n",
    "\n",
    "# 2. Subset the dataframe you want (example: df1)\n",
    "diff_exp_filtered_TRA2A = diff_exp_filtered_TRA2A[diff_exp_filtered_TRA2A[col].isin(shared_keys)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "fdeaa9e5",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>test_id</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene</th>\n",
       "      <th>locus</th>\n",
       "      <th>sample_1</th>\n",
       "      <th>sample_2</th>\n",
       "      <th>status</th>\n",
       "      <th>value_1</th>\n",
       "      <th>value_2</th>\n",
       "      <th>log2(fold_change)</th>\n",
       "      <th>test_stat</th>\n",
       "      <th>p_value</th>\n",
       "      <th>q_value</th>\n",
       "      <th>significant</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>XLOC_000008</td>\n",
       "      <td>XLOC_000008</td>\n",
       "      <td>OR4F29</td>\n",
       "      <td>chr1:317719-461954</td>\n",
       "      <td>TRA2A-BGHLV14-HepG2</td>\n",
       "      <td>Control</td>\n",
       "      <td>OK</td>\n",
       "      <td>0.65528</td>\n",
       "      <td>0.501422</td>\n",
       "      <td>-0.386086</td>\n",
       "      <td>-0.177847</td>\n",
       "      <td>0.57070</td>\n",
       "      <td>0.999131</td>\n",
       "      <td>no</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>XLOC_000012</td>\n",
       "      <td>XLOC_000012</td>\n",
       "      <td>MTND1P23</td>\n",
       "      <td>chr1:536815-660283</td>\n",
       "      <td>TRA2A-BGHLV14-HepG2</td>\n",
       "      <td>Control</td>\n",
       "      <td>OK</td>\n",
       "      <td>136.29500</td>\n",
       "      <td>115.024000</td>\n",
       "      <td>-0.244798</td>\n",
       "      <td>-0.100524</td>\n",
       "      <td>0.69615</td>\n",
       "      <td>0.999131</td>\n",
       "      <td>no</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>XLOC_000013</td>\n",
       "      <td>XLOC_000013</td>\n",
       "      <td>MTND2P28</td>\n",
       "      <td>chr1:536815-660283</td>\n",
       "      <td>TRA2A-BGHLV14-HepG2</td>\n",
       "      <td>Control</td>\n",
       "      <td>OK</td>\n",
       "      <td>711.27600</td>\n",
       "      <td>543.353000</td>\n",
       "      <td>-0.388520</td>\n",
       "      <td>-0.765375</td>\n",
       "      <td>0.27055</td>\n",
       "      <td>0.906203</td>\n",
       "      <td>no</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>XLOC_000014</td>\n",
       "      <td>XLOC_000014</td>\n",
       "      <td>hsa-mir-6723</td>\n",
       "      <td>chr1:536815-660283</td>\n",
       "      <td>TRA2A-BGHLV14-HepG2</td>\n",
       "      <td>Control</td>\n",
       "      <td>OK</td>\n",
       "      <td>642.94600</td>\n",
       "      <td>732.844000</td>\n",
       "      <td>0.188808</td>\n",
       "      <td>0.504172</td>\n",
       "      <td>0.47060</td>\n",
       "      <td>0.993477</td>\n",
       "      <td>no</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>XLOC_000015</td>\n",
       "      <td>XLOC_000015</td>\n",
       "      <td>RP5-857K21.7</td>\n",
       "      <td>chr1:536815-660283</td>\n",
       "      <td>TRA2A-BGHLV14-HepG2</td>\n",
       "      <td>Control</td>\n",
       "      <td>OK</td>\n",
       "      <td>348.26200</td>\n",
       "      <td>359.108000</td>\n",
       "      <td>0.044242</td>\n",
       "      <td>0.050949</td>\n",
       "      <td>0.94195</td>\n",
       "      <td>0.999131</td>\n",
       "      <td>no</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        test_id      gene_id          gene               locus  \\\n",
       "7   XLOC_000008  XLOC_000008        OR4F29  chr1:317719-461954   \n",
       "11  XLOC_000012  XLOC_000012      MTND1P23  chr1:536815-660283   \n",
       "12  XLOC_000013  XLOC_000013      MTND2P28  chr1:536815-660283   \n",
       "13  XLOC_000014  XLOC_000014  hsa-mir-6723  chr1:536815-660283   \n",
       "14  XLOC_000015  XLOC_000015  RP5-857K21.7  chr1:536815-660283   \n",
       "\n",
       "               sample_1 sample_2 status    value_1     value_2  \\\n",
       "7   TRA2A-BGHLV14-HepG2  Control     OK    0.65528    0.501422   \n",
       "11  TRA2A-BGHLV14-HepG2  Control     OK  136.29500  115.024000   \n",
       "12  TRA2A-BGHLV14-HepG2  Control     OK  711.27600  543.353000   \n",
       "13  TRA2A-BGHLV14-HepG2  Control     OK  642.94600  732.844000   \n",
       "14  TRA2A-BGHLV14-HepG2  Control     OK  348.26200  359.108000   \n",
       "\n",
       "    log2(fold_change)  test_stat  p_value   q_value significant  \n",
       "7           -0.386086  -0.177847  0.57070  0.999131          no  \n",
       "11          -0.244798  -0.100524  0.69615  0.999131          no  \n",
       "12          -0.388520  -0.765375  0.27055  0.906203          no  \n",
       "13           0.188808   0.504172  0.47060  0.993477          no  \n",
       "14           0.044242   0.050949  0.94195  0.999131          no  "
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "diff_exp_filtered_TRA2A.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "1046db6c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(17985, 14)"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "diff_exp_filtered_TRA2A.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "d8fe4752",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([-0.386086 , -0.244798 , -0.38852  , ..., -0.434982 , -0.0532165,\n",
       "        1.4967   ])"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "logfold_changes = diff_exp_filtered_TRA2A['log2(fold_change)'].values\n",
    "logfold_changes"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "14a14525",
   "metadata": {},
   "source": [
    "## Generate negative pairs"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "eaa775d1",
   "metadata": {},
   "source": [
    "Shuffle eCLIP sequences"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "68b0a794",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "59110"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import random\n",
    "def shuffle_by_pairs(seq):\n",
    "    pairs = [seq[i:i+2] for i in range(0, len(seq), 2)]\n",
    "    random.shuffle(pairs)\n",
    "    return \"\".join(pairs)\n",
    "in_vivo_neg_seqs = [shuffle_by_pairs(seq) for seq in eclip[1].values]\n",
    "len(in_vivo_neg_seqs)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "1ec9c58c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "56372"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# filter for length and unique values\n",
    "in_vivo_neg_seqs = [s for s in in_vivo_neg_seqs if len(s) >= 10]\n",
    "in_vivo_neg_seqs = list(set(in_vivo_neg_seqs))\n",
    "in_vivo_neg_seqs = list(set(in_vivo_neg_seqs) - set(in_vivo_pos_seqs))\n",
    "len(in_vivo_neg_seqs)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "02a5e383",
   "metadata": {},
   "source": [
    "Retrieve RBNS sequences that are not enriched"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "669ff6f6",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_insignficant_sequences(df):\n",
    "    num_cols = df.columns[1:]\n",
    "    filtered = df[df[num_cols].lt(2).any(axis=1)]\n",
    "    pairs = [\n",
    "        (row[df.columns[0]], tuple(row[num_cols]))\n",
    "        for _, row in filtered.iterrows()\n",
    "    ]\n",
    "    return pairs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "4a0c3f24",
   "metadata": {},
   "outputs": [],
   "source": [
    "neg_four_mers = get_insignficant_sequences(four_mers)\n",
    "neg_five_mers = get_insignficant_sequences(five_mers)\n",
    "neg_six_mers = get_insignficant_sequences(six_mers)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "7920a3d9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[('AGAA', (1.0179, 1.9157, 2.4378, 2.843, 2.7831, 2.5174)),\n",
       " ('GAAG', (0.9857, 2.0081, 2.6597, 2.827, 2.637, 2.2785)),\n",
       " ('AAGA', (1.013, 1.8014, 2.2892, 2.5155, 2.3421, 2.042)),\n",
       " ('GAAA', (0.9935, 1.5545, 1.8787, 2.1811, 2.2963, 2.1849)),\n",
       " ('TGAA', (1.0309, 1.5297, 1.8455, 2.0781, 2.1111, 1.9834)),\n",
       " ('CGAA', (1.0152, 1.4349, 1.7138, 1.9392, 2.0369, 1.9997)),\n",
       " ('GAAT', (1.0332, 1.3938, 1.5998, 1.8862, 1.9037, 1.7997)),\n",
       " ('AAAA', (0.9796, 1.2833, 1.4431, 1.562, 1.683, 1.6166)),\n",
       " ('GAAC', (1.0107, 1.2331, 1.3571, 1.5423, 1.6013, 1.6389)),\n",
       " ('AATA', (0.9756, 1.1941, 1.2813, 1.476, 1.4739, 1.4009)),\n",
       " ('AACA', (0.9932, 1.1455, 1.2187, 1.4302, 1.4291, 1.4719)),\n",
       " ('AAGC', (0.9808, 1.1753, 1.2622, 1.3948, 1.4063, 1.4377)),\n",
       " ('ATGA', (1.0263, 1.1857, 1.2618, 1.3941, 1.3691, 1.3013)),\n",
       " ('AAAG', (0.9915, 1.1844, 1.2719, 1.3921, 1.3641, 1.3028)),\n",
       " ('CCGA', (1.0218, 1.1603, 1.2665, 1.3396, 1.4863, 1.5305)),\n",
       " ('AAAC', (0.9899, 1.1157, 1.1794, 1.3124, 1.3736, 1.4098)),\n",
       " ('CAGA', (0.999, 1.1357, 1.2058, 1.3101, 1.312, 1.3247)),\n",
       " ('AAAT', (0.9841, 1.1378, 1.2071, 1.3053, 1.3681, 1.3371)),\n",
       " ('AATG', (1.0039, 1.124, 1.1725, 1.2726, 1.2447, 1.1861)),\n",
       " ('CTGA', (1.0355, 1.1639, 1.2744, 1.26, 1.2741, 1.2386)),\n",
       " ('AGCA', (0.9705, 1.0864, 1.1232, 1.248, 1.25, 1.3153)),\n",
       " ('AGAG', (0.9806, 1.1341, 1.2048, 1.2287, 1.1877, 1.1657)),\n",
       " ('ACGA', (1.0113, 1.1034, 1.1619, 1.2264, 1.2263, 1.2093)),\n",
       " ('TAGA', (1.0175, 1.1103, 1.15, 1.2218, 1.209, 1.1794)),\n",
       " ('ACAA', (0.9952, 1.0514, 1.0706, 1.2012, 1.162, 1.1893)),\n",
       " ('GGAA', (0.9631, 1.1094, 1.179, 1.1922, 1.1814, 1.1776)),\n",
       " ('CAAG', (0.9854, 1.0707, 1.107, 1.1839, 1.1497, 1.1775)),\n",
       " ('ACAC', (0.9969, 1.0276, 1.0415, 1.1612, 1.2083, 1.3314)),\n",
       " ('ATAA', (0.9831, 1.0585, 1.0615, 1.1609, 1.1319, 1.0833)),\n",
       " ('AATT', (1.0023, 1.0556, 1.0833, 1.1606, 1.2053, 1.185)),\n",
       " ('AACC', (0.9949, 1.0254, 1.0495, 1.1411, 1.2363, 1.3189)),\n",
       " ('CTCC', (1.0205, 1.0665, 1.117, 1.1379, 1.2832, 1.4078)),\n",
       " ('AAGT', (1.0054, 1.0713, 1.0878, 1.1379, 1.0989, 1.0659)),\n",
       " ('AACT', (1.0166, 1.0448, 1.0581, 1.1276, 1.1468, 1.1694)),\n",
       " ('ACCC', (0.9856, 1.0172, 1.0289, 1.1267, 1.3223, 1.5115)),\n",
       " ('ACAG', (0.9754, 1.0268, 1.0486, 1.113, 1.0814, 1.0999)),\n",
       " ('CCCG', (0.988, 1.0388, 1.0651, 1.1082, 1.3372, 1.4848)),\n",
       " ('GCAA', (0.9708, 1.0232, 1.0403, 1.1026, 1.0846, 1.1275)),\n",
       " ('ATAC', (0.9889, 1.0124, 1.0114, 1.0989, 1.09, 1.101)),\n",
       " ('ATAG', (0.9891, 1.0353, 1.0388, 1.0968, 1.046, 1.0078)),\n",
       " ('GAGA', (1.0125, 1.0763, 1.0931, 1.0875, 1.0757, 1.0588)),\n",
       " ('ACAT', (0.9993, 0.999, 0.9943, 1.0868, 1.0587, 1.079)),\n",
       " ('GTGA', (1.0059, 1.0761, 1.1162, 1.0812, 1.0368, 0.9693)),\n",
       " ('AGAT', (1.0289, 1.0629, 1.0865, 1.0784, 1.0036, 0.9482)),\n",
       " ('ATTA', (0.9931, 1.0127, 1.0067, 1.0781, 1.0591, 1.0156)),\n",
       " ('TTGA', (1.0512, 1.0502, 1.0585, 1.0781, 1.0779, 1.0622)),\n",
       " ('CCCC', (0.9718, 1.0103, 1.0403, 1.0774, 1.3354, 1.5024)),\n",
       " ('CACC', (1.001, 0.9891, 0.9916, 1.0758, 1.2391, 1.4456)),\n",
       " ('AACG', (0.9929, 1.0294, 1.0499, 1.0732, 1.0666, 1.0522)),\n",
       " ('ATAT', (0.9715, 1.0101, 0.9982, 1.0728, 1.052, 1.0057)),\n",
       " ('TAAG', (0.9984, 1.0373, 1.0349, 1.0672, 1.0036, 0.9483)),\n",
       " ('TCCC', (1.0226, 1.0195, 1.0555, 1.0639, 1.2214, 1.3134)),\n",
       " ('AAGG', (0.953, 1.0474, 1.0812, 1.0578, 0.9997, 0.9634)),\n",
       " ('AGAC', (1.0054, 1.0302, 1.0497, 1.0522, 1.0027, 0.9903)),\n",
       " ('AGCC', (0.9875, 0.9961, 1.0015, 1.0448, 1.1569, 1.2781)),\n",
       " ('CCTG', (1.0143, 1.0397, 1.0837, 1.0397, 1.1147, 1.165)),\n",
       " ('TACA', (0.9843, 0.971, 0.9576, 1.0367, 1.0055, 1.0459)),\n",
       " ('CACA', (0.9887, 0.9666, 0.957, 1.033, 1.0516, 1.1463)),\n",
       " ('ACTC', (1.0375, 1.0162, 1.0101, 1.0312, 1.0717, 1.1276)),\n",
       " ('CAAA', (0.99, 0.9861, 0.9717, 1.0306, 1.0382, 1.0963)),\n",
       " ('GCGA', (0.9761, 1.0305, 1.0493, 1.0273, 1.0344, 1.0292)),\n",
       " ('CCCT', (1.0123, 0.991, 1.0015, 1.0235, 1.1851, 1.3259)),\n",
       " ('AGGA', (0.9626, 1.0305, 1.0498, 1.0216, 0.9546, 0.9065)),\n",
       " ('GCCC', (0.9757, 0.99, 0.9888, 1.0206, 1.217, 1.3766)),\n",
       " ('CCCA', (0.9807, 0.9771, 0.9718, 1.0203, 1.1718, 1.3329)),\n",
       " ('AGTA', (0.9905, 1.0046, 0.9878, 1.0194, 0.9897, 0.9823)),\n",
       " ('ACTA', (0.9984, 0.9907, 0.978, 1.0123, 0.984, 0.9671)),\n",
       " ('GCAG', (0.9529, 0.998, 0.9942, 1.0049, 1.0122, 1.0518)),\n",
       " ('AGCT', (1.0052, 1.0064, 0.9966, 1.0046, 1.0049, 1.0412)),\n",
       " ('GACA', (1.0007, 0.9959, 0.989, 0.9965, 0.9574, 0.9736)),\n",
       " ('CAAC', (0.9972, 0.957, 0.9433, 0.9949, 1.0047, 1.0752)),\n",
       " ('ACCA', (0.9944, 0.958, 0.9499, 0.993, 1.0287, 1.109)),\n",
       " ('AGCG', (0.9606, 1.0023, 1.0129, 0.9847, 0.9755, 0.9743)),\n",
       " ('TATA', (0.9645, 0.9719, 0.9305, 0.9809, 0.9562, 0.9226)),\n",
       " ('TAAA', (0.9897, 0.9917, 0.9584, 0.9805, 0.9783, 0.9576)),\n",
       " ('ATGC', (0.9934, 0.9766, 0.9615, 0.9792, 0.99, 1.0357)),\n",
       " ('GCAC', (0.9828, 0.9696, 0.9649, 0.9784, 1.0427, 1.1514)),\n",
       " ('TAAC', (1.001, 0.9664, 0.9405, 0.9776, 0.9649, 0.9741)),\n",
       " ('AATC', (1.0206, 0.9701, 0.9584, 0.9748, 0.9836, 1.0079)),\n",
       " ('CAAT', (1.0059, 0.9548, 0.9255, 0.9719, 0.9473, 0.9741)),\n",
       " ('ACCT', (1.0238, 0.969, 0.9472, 0.9715, 1.0405, 1.1306)),\n",
       " ('CATA', (0.9776, 0.9451, 0.9164, 0.9702, 0.9466, 0.9654)),\n",
       " ('TATG', (0.996, 0.9865, 0.96, 0.9677, 0.9193, 0.8642)),\n",
       " ('AGTG', (0.9843, 0.9957, 0.9816, 0.9669, 0.9348, 0.9106)),\n",
       " ('TACC', (1.0025, 0.954, 0.9402, 0.9661, 1.0217, 1.0716)),\n",
       " ('ACTG', (1.0183, 0.9941, 0.9984, 0.9661, 0.929, 0.8885)),\n",
       " ('CCTC', (1.0246, 0.9673, 0.9683, 0.9614, 1.0912, 1.1996)),\n",
       " ('TTAA', (1.0, 0.9736, 0.9389, 0.9599, 0.9468, 0.9232)),\n",
       " ('ACTT', (1.027, 0.9729, 0.9454, 0.9565, 0.9538, 0.9674)),\n",
       " ('GATG', (1.0018, 1.0049, 0.9979, 0.9556, 0.9011, 0.8541)),\n",
       " ('GAGC', (0.9775, 0.9921, 0.9818, 0.9541, 0.987, 1.0416)),\n",
       " ('TCGA', (1.0534, 0.9982, 1.002, 0.9517, 0.9209, 0.8822)),\n",
       " ('CATG', (1.0007, 0.9633, 0.9494, 0.9504, 0.9278, 0.9368)),\n",
       " ('TTAC', (1.0064, 0.9542, 0.9324, 0.947, 0.9431, 0.954)),\n",
       " ('GACT', (1.0319, 1.003, 1.0052, 0.9465, 0.86, 0.7783)),\n",
       " ('CACT', (1.0241, 0.9576, 0.9319, 0.9443, 0.9754, 1.0299)),\n",
       " ('GCAT', (0.9885, 0.9542, 0.9317, 0.9441, 0.9181, 0.94)),\n",
       " ('GATA', (1.0115, 0.9835, 0.9533, 0.9437, 0.8946, 0.8755)),\n",
       " ('TGCA', (0.9875, 0.9486, 0.9252, 0.9428, 0.943, 1.0095)),\n",
       " ('CCAA', (0.9958, 0.9461, 0.922, 0.9403, 0.9914, 1.0882)),\n",
       " ('ATTG', (1.0339, 0.9708, 0.9429, 0.9397, 0.8776, 0.8216)),\n",
       " ('ATCA', (1.0156, 0.9382, 0.9061, 0.9384, 0.9053, 0.93)),\n",
       " ('CCAC', (0.9899, 0.9357, 0.914, 0.9312, 1.0411, 1.1718)),\n",
       " ('TCAA', (1.0173, 0.9417, 0.9067, 0.9272, 0.8983, 0.9386)),\n",
       " ('ACGC', (0.9846, 0.9496, 0.9279, 0.9251, 0.9761, 1.0633)),\n",
       " ('CAGC', (0.971, 0.9331, 0.9081, 0.9242, 0.9897, 1.1048)),\n",
       " ('CCTT', (1.0607, 0.9568, 0.9275, 0.9213, 1.0038, 1.1029)),\n",
       " ('TTAG', (1.0097, 0.9667, 0.9363, 0.9203, 0.8815, 0.8449)),\n",
       " ('TAAT', (0.9863, 0.9514, 0.9051, 0.9202, 0.8953, 0.8587)),\n",
       " ('TCCT', (1.0546, 0.9814, 0.9832, 0.9197, 0.9552, 0.9826)),\n",
       " ('TAGC', (0.9863, 0.949, 0.9222, 0.9195, 0.9296, 0.9761)),\n",
       " ('TACT', (1.0115, 0.9594, 0.9241, 0.9195, 0.902, 0.8892)),\n",
       " ('CCAG', (0.9834, 0.9505, 0.9381, 0.9194, 0.9883, 1.0705)),\n",
       " ('CTAC', (0.9995, 0.9402, 0.9143, 0.9147, 0.9359, 0.9867)),\n",
       " ('ACCG', (0.9904, 0.9499, 0.9445, 0.9084, 0.9359, 0.9406)),\n",
       " ('ATTT', (0.9839, 0.9417, 0.916, 0.9043, 0.9542, 0.9584)),\n",
       " ('GCCA', (0.9858, 0.9325, 0.9112, 0.9019, 0.9662, 1.0672)),\n",
       " ('CATT', (1.0176, 0.9273, 0.8865, 0.9015, 0.8914, 0.9066)),\n",
       " ('TGAC', (1.0301, 0.9673, 0.9384, 0.9014, 0.8773, 0.8781)),\n",
       " ('CTCA', (1.0199, 0.932, 0.8978, 0.9005, 0.9359, 1.0189)),\n",
       " ('GTAG', (0.9842, 0.9749, 0.9583, 0.8999, 0.8824, 0.8661)),\n",
       " ('GGAC', (0.9635, 0.9797, 1.0063, 0.8996, 0.8151, 0.6963)),\n",
       " ('GCTA', (0.9872, 0.9542, 0.9315, 0.8981, 0.8768, 0.876)),\n",
       " ('GCCT', (1.0459, 0.9506, 0.9312, 0.8975, 1.0027, 1.1111)),\n",
       " ('ATCC', (1.0285, 0.9412, 0.9151, 0.8971, 0.9393, 0.9897)),\n",
       " ('TGCC', (1.0123, 0.9443, 0.9204, 0.8969, 0.9875, 1.0836)),\n",
       " ('GAGT', (1.0052, 0.9754, 0.933, 0.8965, 0.8612, 0.8469)),\n",
       " ('TTTA', (0.9783, 0.9351, 0.8998, 0.8956, 0.9204, 0.9211)),\n",
       " ('CGCA', (0.9783, 0.9266, 0.9022, 0.8955, 0.9292, 1.0105)),\n",
       " ('GTAA', (0.9938, 0.9557, 0.9165, 0.8939, 0.8526, 0.8252)),\n",
       " ('TGAG', (1.0069, 0.9714, 0.9289, 0.8938, 0.8761, 0.8759)),\n",
       " ('CTTG', (1.0582, 0.971, 0.942, 0.893, 0.9112, 0.9458)),\n",
       " ('TGGA', (0.9867, 0.9734, 0.9606, 0.8926, 0.8302, 0.7713)),\n",
       " ('TACG', (0.9926, 0.9525, 0.9288, 0.8917, 0.855, 0.8261)),\n",
       " ('CTTA', (1.0144, 0.9397, 0.8967, 0.8907, 0.8963, 0.9235)),\n",
       " ('AGTT', (1.0246, 0.9598, 0.9159, 0.8904, 0.8472, 0.8239)),\n",
       " ('CTAA', (1.0022, 0.9419, 0.8989, 0.8878, 0.8852, 0.9007)),\n",
       " ('CTCT', (1.043, 0.958, 0.9249, 0.8869, 0.9334, 0.9654)),\n",
       " ('GCTC', (1.0118, 0.9637, 0.9409, 0.8868, 0.923, 0.974)),\n",
       " ('CACG', (0.9824, 0.938, 0.9163, 0.8866, 0.9139, 0.9622)),\n",
       " ('CCTA', (1.0043, 0.9267, 0.8856, 0.8859, 0.9506, 1.044)),\n",
       " ('CCAT', (1.0138, 0.9202, 0.8862, 0.8855, 0.9227, 0.9876)),\n",
       " ('GACG', (0.9833, 0.9704, 0.9631, 0.8836, 0.8433, 0.7849)),\n",
       " ('TTAT', (0.9793, 0.9297, 0.8821, 0.8812, 0.873, 0.8467)),\n",
       " ('TCAC', (1.0254, 0.9208, 0.8875, 0.8807, 0.8912, 0.9395)),\n",
       " ('GGTG', (0.9482, 0.9866, 1.0057, 0.8784, 0.8339, 0.774)),\n",
       " ('AGGC', (0.9445, 0.9519, 0.944, 0.8764, 0.8782, 0.9039)),\n",
       " ('GCTG', (0.9871, 0.9632, 0.9488, 0.8757, 0.8526, 0.8321)),\n",
       " ('GGCA', (0.9453, 0.9527, 0.9417, 0.8741, 0.8695, 0.8868)),\n",
       " ('TCTC', (1.0471, 0.9508, 0.9239, 0.8728, 0.8955, 0.9146)),\n",
       " ('TATT', (0.9766, 0.923, 0.8758, 0.8726, 0.8674, 0.8392)),\n",
       " ('ATGT', (1.0072, 0.9508, 0.903, 0.8684, 0.8174, 0.7786)),\n",
       " ('ATGG', (0.9716, 0.9547, 0.9219, 0.8682, 0.8107, 0.7686)),\n",
       " ('CATC', (1.0418, 0.9181, 0.8815, 0.8651, 0.8743, 0.9251)),\n",
       " ('GCGC', (0.9608, 0.9364, 0.9075, 0.8627, 0.9279, 0.9984)),\n",
       " ('CTAG', (0.9964, 0.9403, 0.9041, 0.8621, 0.8581, 0.877)),\n",
       " ('CTAT', (0.9997, 0.9237, 0.8687, 0.8601, 0.8542, 0.8562)),\n",
       " ('GCTT', (1.0281, 0.9496, 0.9079, 0.858, 0.8466, 0.8567)),\n",
       " ('CTTC', (1.0537, 0.9327, 0.8969, 0.8565, 0.8838, 0.9303)),\n",
       " ('GACC', (0.9963, 0.9361, 0.9185, 0.856, 0.8725, 0.8921)),\n",
       " ('CGGA', (0.9741, 0.9544, 0.9603, 0.8554, 0.7965, 0.7107)),\n",
       " ('TCAG', (1.0076, 0.9216, 0.893, 0.8552, 0.8197, 0.8258)),\n",
       " ('CGCC', (1.0164, 0.917, 0.8871, 0.8548, 0.9928, 1.1322)),\n",
       " ('TGTA', (1.0045, 0.9412, 0.8957, 0.8499, 0.8219, 0.8087)),\n",
       " ('GTAC', (0.9916, 0.9258, 0.8911, 0.8477, 0.8172, 0.8081)),\n",
       " ('CGAG', (1.0103, 0.9528, 0.9036, 0.8468, 0.8843, 0.9145)),\n",
       " ('TCCG', (1.0171, 0.9375, 0.9081, 0.8466, 0.8715, 0.8731)),\n",
       " ('TTCC', (1.0523, 0.9293, 0.8886, 0.8425, 0.9069, 0.9507)),\n",
       " ('GTTA', (1.0155, 0.9379, 0.889, 0.8371, 0.7717, 0.7203)),\n",
       " ('CTTT', (1.0301, 0.9191, 0.8668, 0.8331, 0.8658, 0.9024)),\n",
       " ('TTTG', (1.0269, 0.9341, 0.8876, 0.8322, 0.8209, 0.787)),\n",
       " ('TTGC', (1.0237, 0.9227, 0.8751, 0.8311, 0.8319, 0.8607)),\n",
       " ('TTCA', (1.0239, 0.8898, 0.8328, 0.8284, 0.815, 0.8663)),\n",
       " ('GGCT', (0.9635, 0.952, 0.9299, 0.8284, 0.7744, 0.7277)),\n",
       " ('GATT', (1.043, 0.9403, 0.893, 0.8283, 0.7598, 0.7034)),\n",
       " ('TGCT', (1.019, 0.9352, 0.8914, 0.8281, 0.8187, 0.8324)),\n",
       " ('TGTG', (0.9976, 0.9426, 0.8973, 0.8278, 0.7828, 0.7437)),\n",
       " ('CTCG', (1.0396, 0.945, 0.9135, 0.8262, 0.8267, 0.8154)),\n",
       " ('CGAC', (1.0053, 0.9266, 0.8936, 0.8261, 0.8246, 0.8366)),\n",
       " ('ATTC', (1.0319, 0.9058, 0.8472, 0.8215, 0.8222, 0.8364)),\n",
       " ('CTGC', (0.9993, 0.9177, 0.8787, 0.8212, 0.865, 0.9283)),\n",
       " ('CCGC', (0.9838, 0.8989, 0.8557, 0.8209, 0.9423, 1.0712)),\n",
       " ('GTGC', (0.9741, 0.9267, 0.8865, 0.8208, 0.8364, 0.8726)),\n",
       " ('TCTG', (1.0349, 0.9403, 0.9068, 0.8184, 0.7842, 0.7328)),\n",
       " ('GGGA', (0.9248, 0.9392, 0.9127, 0.8143, 0.8009, 0.7878)),\n",
       " ('GCCG', (0.9892, 0.9197, 0.8938, 0.8142, 0.8592, 0.8685)),\n",
       " ('GAGG', (0.9341, 0.9403, 0.9105, 0.813, 0.7944, 0.7757)),\n",
       " ('GTTG', (1.0179, 0.9415, 0.9068, 0.8108, 0.7254, 0.6399)),\n",
       " ('GTAT', (0.998, 0.9167, 0.8549, 0.8075, 0.7613, 0.7198)),\n",
       " ('TATC', (1.0114, 0.8977, 0.8398, 0.8066, 0.7906, 0.7869)),\n",
       " ('TGAT', (1.0506, 0.9336, 0.8674, 0.8044, 0.7544, 0.7229)),\n",
       " ('TCCA', (1.0262, 0.9004, 0.8452, 0.8036, 0.8176, 0.8577)),\n",
       " ('CAGT', (1.0017, 0.9028, 0.8462, 0.8034, 0.7826, 0.8065)),\n",
       " ('CGCT', (1.0152, 0.9141, 0.8699, 0.8033, 0.8301, 0.8668)),\n",
       " ('TGCG', (0.9767, 0.9243, 0.8754, 0.8028, 0.7855, 0.7737)),\n",
       " ('TAGG', (0.9582, 0.9207, 0.8789, 0.8025, 0.7421, 0.7016)),\n",
       " ('ACGG', (0.9494, 0.9271, 0.9029, 0.8024, 0.7523, 0.6937)),\n",
       " ('TCAT', (1.0278, 0.8857, 0.8289, 0.8018, 0.7701, 0.7766)),\n",
       " ('TAGT', (1.0079, 0.9235, 0.8602, 0.8016, 0.7526, 0.7208)),\n",
       " ('GTCC', (1.0104, 0.9257, 0.8958, 0.8003, 0.7946, 0.7775)),\n",
       " ('AGGG', (0.9022, 0.9237, 0.8972, 0.8, 0.773, 0.7553)),\n",
       " ('AGTC', (1.0132, 0.9233, 0.8694, 0.7983, 0.7486, 0.7345)),\n",
       " ('GGCC', (0.957, 0.9177, 0.8958, 0.7982, 0.8587, 0.8928)),\n",
       " ('AGGT', (0.9762, 0.927, 0.8806, 0.7964, 0.7419, 0.7131)),\n",
       " ('TGGC', (0.9795, 0.9178, 0.8804, 0.7951, 0.8053, 0.8236)),\n",
       " ('GGCG', (0.9286, 0.9284, 0.9155, 0.7894, 0.7832, 0.7421)),\n",
       " ('GGGC', (0.9031, 0.9294, 0.918, 0.7883, 0.766, 0.7129)),\n",
       " ('ATCT', (1.031, 0.9039, 0.8411, 0.7876, 0.7773, 0.7794)),\n",
       " ('GTCA', (1.0125, 0.9088, 0.8712, 0.7857, 0.7137, 0.6675)),\n",
       " ('CAGG', (0.9405, 0.9001, 0.8653, 0.7855, 0.7764, 0.792)),\n",
       " ('CTGT', (1.0303, 0.9231, 0.8726, 0.7843, 0.7632, 0.7535)),\n",
       " ('GGAT', (0.9941, 0.922, 0.8785, 0.7777, 0.6783, 0.5887)),\n",
       " ('CGAT', (1.0327, 0.9086, 0.8475, 0.7771, 0.7562, 0.7602)),\n",
       " ('GGAG', (0.9327, 0.9202, 0.8837, 0.7742, 0.7473, 0.7249)),\n",
       " ('CGCG', (0.9705, 0.9082, 0.8599, 0.7716, 0.7963, 0.8118)),\n",
       " ('TTGG', (1.0097, 0.9148, 0.8696, 0.7702, 0.7078, 0.6482)),\n",
       " ('TTTT', (0.9693, 0.8835, 0.8429, 0.7671, 0.8586, 0.8897)),\n",
       " ('ACGT', (1.0066, 0.9048, 0.8565, 0.7663, 0.7264, 0.7047)),\n",
       " ('ATCG', (1.0498, 0.9071, 0.874, 0.7633, 0.6895, 0.6204)),\n",
       " ('CGGC', (0.9524, 0.9132, 0.8835, 0.7625, 0.7593, 0.7286)),\n",
       " ('GCGG', (0.926, 0.9165, 0.8877, 0.7581, 0.7339, 0.6623)),\n",
       " ('GGTA', (0.9647, 0.9054, 0.8473, 0.7574, 0.7122, 0.6828)),\n",
       " ('GTGG', (0.9465, 0.9133, 0.8767, 0.7566, 0.7216, 0.6663)),\n",
       " ('GTGT', (0.9947, 0.9137, 0.8511, 0.7566, 0.7043, 0.6592)),\n",
       " ('GTCG', (1.0335, 0.9168, 0.899, 0.7532, 0.6678, 0.5512)),\n",
       " ('TCGG', (1.0045, 0.9112, 0.8959, 0.7485, 0.6611, 0.5479)),\n",
       " ('GTTT', (1.0161, 0.8953, 0.8322, 0.7467, 0.7161, 0.6803)),\n",
       " ('TCTT', (1.0458, 0.8924, 0.8258, 0.7462, 0.7468, 0.7497)),\n",
       " ('CGTG', (0.9957, 0.9057, 0.8504, 0.7449, 0.7262, 0.7048)),\n",
       " ('TGTT', (1.0382, 0.9109, 0.8343, 0.7439, 0.6939, 0.6503)),\n",
       " ('GATC', (1.0475, 0.9071, 0.8598, 0.7433, 0.6776, 0.6232)),\n",
       " ('TCTA', (1.0165, 0.8815, 0.8055, 0.7409, 0.7241, 0.722)),\n",
       " ('CGGT', (0.9801, 0.9105, 0.8803, 0.7371, 0.6874, 0.6286)),\n",
       " ('CCGG', (0.9532, 0.8953, 0.8574, 0.737, 0.7749, 0.7771)),\n",
       " ('CTGG', (0.9825, 0.8999, 0.8438, 0.7341, 0.7289, 0.7311)),\n",
       " ('TTTC', (1.0218, 0.8699, 0.8068, 0.7333, 0.7644, 0.8013)),\n",
       " ('GTCT', (1.0353, 0.9045, 0.8452, 0.7329, 0.6689, 0.6061)),\n",
       " ('CGTA', (1.0021, 0.8793, 0.819, 0.7315, 0.7103, 0.7036)),\n",
       " ('TGGG', (0.9335, 0.8972, 0.8458, 0.7314, 0.691, 0.6502)),\n",
       " ('TCGC', (1.0452, 0.8888, 0.8354, 0.7307, 0.7115, 0.694)),\n",
       " ('TGTC', (1.0308, 0.8991, 0.8365, 0.7269, 0.6722, 0.6336)),\n",
       " ('TTCG', (1.0415, 0.8939, 0.8337, 0.7264, 0.674, 0.6231)),\n",
       " ('GCGT', (0.9821, 0.8927, 0.8377, 0.7259, 0.6896, 0.6651)),\n",
       " ('TTCT', (1.0328, 0.8793, 0.8108, 0.7226, 0.7268, 0.7327)),\n",
       " ('TTGT', (1.0407, 0.8977, 0.8177, 0.7223, 0.6674, 0.6228)),\n",
       " ('CCGT', (1.0223, 0.8719, 0.8036, 0.7137, 0.7442, 0.7744)),\n",
       " ('GTTC', (1.04, 0.8866, 0.8127, 0.7131, 0.6783, 0.6622)),\n",
       " ('TGGT', (1.008, 0.8981, 0.8296, 0.7058, 0.646, 0.5943)),\n",
       " ('GGTT', (0.9982, 0.8859, 0.8128, 0.6999, 0.6331, 0.5877)),\n",
       " ('GGGG', (0.8731, 0.8785, 0.8406, 0.6977, 0.7151, 0.706)),\n",
       " ('CGTC', (1.0544, 0.8777, 0.8401, 0.692, 0.6464, 0.6053)),\n",
       " ('CGGG', (0.9202, 0.8796, 0.8297, 0.6913, 0.6943, 0.6643)),\n",
       " ('CGTT', (1.0362, 0.8679, 0.7962, 0.6736, 0.6339, 0.6104)),\n",
       " ('GGGT', (0.9313, 0.8662, 0.7994, 0.6708, 0.6366, 0.6112)),\n",
       " ('GGTC', (0.987, 0.8717, 0.8178, 0.6655, 0.6135, 0.5582)),\n",
       " ('TCGT', (1.0746, 0.8632, 0.7969, 0.632, 0.5554, 0.4824))]"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "neg_four_mers"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "47bc768a",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "<__array_function__ internals>:180: VisibleDeprecationWarning: Creating an ndarray from ragged nested sequences (which is a list-or-tuple of lists-or-tuples-or ndarrays with different lengths or shapes) is deprecated. If you meant to do this, you must specify 'dtype=object' when creating the ndarray.\n",
      "/tmp/ipykernel_1051197/4178471540.py:3: VisibleDeprecationWarning: Creating an ndarray from ragged nested sequences (which is a list-or-tuple of lists-or-tuples-or ndarrays with different lengths or shapes) is deprecated. If you meant to do this, you must specify 'dtype=object' when creating the ndarray.\n",
      "  in_vitro_neg_seqs = np.array([(s, v) for s, v in in_vitro_neg_seqs_all if s not in pos_rbns_seqs])\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "(5216, 2)"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "in_vitro_neg_seqs_all = np.concatenate((neg_four_mers,neg_five_mers,neg_six_mers))\n",
    "pos_rbns_seqs = {s for s, _ in in_vitro_pos_seqs}\n",
    "in_vitro_neg_seqs = np.array([(s, v) for s, v in in_vitro_neg_seqs_all if s not in pos_rbns_seqs])\n",
    "in_vitro_neg_seqs.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "88397cff",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "5216"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "unique_rbns = {}\n",
    "for s, arr in in_vitro_neg_seqs:\n",
    "    if s not in unique_rbns:\n",
    "        unique_rbns[s] = arr\n",
    "\n",
    "in_vitro_neg_seqs = list(unique_rbns.items())\n",
    "len(in_vitro_neg_seqs)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "94988d48",
   "metadata": {},
   "source": [
    "Construct placeholder expression change vector"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "950281ea",
   "metadata": {},
   "outputs": [],
   "source": [
    "placeholder_logfold_changes = np.zeros(shape=logfold_changes.shape)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b14cf59a",
   "metadata": {},
   "source": [
    "## Split into train/val/test"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "e9d8ef74",
   "metadata": {},
   "outputs": [],
   "source": [
    "# number of total samples per class\n",
    "combinations_size = 500000"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a99c5d06",
   "metadata": {},
   "source": [
    "Positive samples"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "c5fdbf5e",
   "metadata": {},
   "outputs": [],
   "source": [
    "from itertools import product\n",
    "\n",
    "random.shuffle(in_vivo_pos_seqs)\n",
    "random.shuffle(in_vitro_pos_seqs)\n",
    "\n",
    "n_eclip = len(in_vivo_pos_seqs)\n",
    "n_rbns  = len(in_vitro_pos_seqs)\n",
    "\n",
    "# Split indices\n",
    "train_eclip = in_vivo_pos_seqs[:int(0.7*n_eclip)]\n",
    "val_eclip   = in_vivo_pos_seqs[int(0.7*n_eclip):int(0.85*n_eclip)]\n",
    "test_eclip  = in_vivo_pos_seqs[int(0.85*n_eclip):]\n",
    "\n",
    "train_rbns = in_vitro_pos_seqs[:int(0.7*n_rbns)]\n",
    "val_rbns   = in_vitro_pos_seqs[int(0.7*n_rbns):int(0.85*n_rbns)]\n",
    "test_rbns  = in_vitro_pos_seqs[int(0.85*n_rbns):]\n",
    "\n",
    "pos_train_samples = [\n",
    "    (random.choice(train_eclip), random.choice(train_rbns))\n",
    "    for _ in range(combinations_size)\n",
    "]\n",
    "pos_val_samples = [\n",
    "    (random.choice(val_eclip), random.choice(val_rbns))\n",
    "    for _ in range(combinations_size)\n",
    "]\n",
    "pos_test_samples = [\n",
    "    (random.choice(test_eclip), random.choice(test_rbns))\n",
    "    for _ in range(combinations_size)\n",
    "]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "91baf0a1",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "eCLIP sequences are disjoint across splits\n"
     ]
    }
   ],
   "source": [
    "pos_train_eclips = set(eclip for eclip, _ in pos_train_samples)\n",
    "pos_val_eclips   = set(eclip for eclip, _ in pos_val_samples)\n",
    "pos_test_eclips  = set(eclip for eclip, _ in pos_test_samples)\n",
    "\n",
    "assert pos_train_eclips.isdisjoint(pos_val_eclips), \"Overlap between train and val eCLIP sequences!\"\n",
    "assert pos_train_eclips.isdisjoint(pos_test_eclips), \"Overlap between train and test eCLIP sequences!\"\n",
    "assert pos_val_eclips.isdisjoint(pos_test_eclips), \"Overlap between val and test eCLIP sequences!\"\n",
    "print(\"eCLIP sequences are disjoint across splits\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "0ffaaf81",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "RBNS sequences are disjoint across splits\n"
     ]
    }
   ],
   "source": [
    "pos_train_rbns = set(rbns for _, (rbns, _) in pos_train_samples)\n",
    "pos_val_rbns   = set(rbns for _, (rbns, _) in pos_val_samples)\n",
    "pos_test_rbns  = set(rbns for _, (rbns, _) in pos_test_samples)\n",
    "\n",
    "assert pos_train_rbns.isdisjoint(val_rbns), \"Overlap between train and val RBNS sequences!\"\n",
    "assert pos_val_rbns.isdisjoint(test_rbns), \"Overlap between train and test RBNS sequences!\"\n",
    "assert pos_test_rbns.isdisjoint(test_rbns), \"Overlap between val and test RBNS sequences!\"\n",
    "print(\"RBNS sequences are disjoint across splits\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "330d4bc3",
   "metadata": {},
   "outputs": [],
   "source": [
    "pos_train_label = np.ones(len(pos_train_samples))\n",
    "pos_val_label = np.ones(len(pos_val_samples))\n",
    "pos_test_label = np.ones(len(pos_test_samples))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c3da2c3a",
   "metadata": {},
   "source": [
    "Negative samples"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "d7c965b7",
   "metadata": {},
   "outputs": [],
   "source": [
    "random.shuffle(in_vivo_neg_seqs)\n",
    "random.shuffle(in_vitro_neg_seqs)\n",
    "\n",
    "n_eclip = len(in_vivo_neg_seqs)\n",
    "n_rbns  = len(in_vitro_neg_seqs)\n",
    "\n",
    "# Split indices\n",
    "train_eclip = in_vivo_neg_seqs[:int(0.7*n_eclip)]\n",
    "val_eclip   = in_vivo_neg_seqs[int(0.7*n_eclip):int(0.85*n_eclip)]\n",
    "test_eclip  = in_vivo_neg_seqs[int(0.85*n_eclip):]\n",
    "\n",
    "train_rbns = in_vitro_neg_seqs[:int(0.7*n_rbns)]\n",
    "val_rbns   = in_vitro_neg_seqs[int(0.7*n_rbns):int(0.85*n_rbns)]\n",
    "test_rbns  = in_vitro_neg_seqs[int(0.85*n_rbns):]\n",
    "\n",
    "neg_train_samples = [\n",
    "    (random.choice(train_eclip), random.choice(train_rbns))\n",
    "    for _ in range(combinations_size)\n",
    "]\n",
    "neg_val_samples = [\n",
    "    (random.choice(val_eclip), random.choice(val_rbns))\n",
    "    for _ in range(combinations_size)\n",
    "]\n",
    "neg_test_samples = [\n",
    "    (random.choice(test_eclip), random.choice(test_rbns))\n",
    "    for _ in range(combinations_size)\n",
    "]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "19d1f6e3",
   "metadata": {},
   "outputs": [],
   "source": [
    "neg_train_label = np.zeros(len(neg_train_samples))\n",
    "neg_val_label = np.zeros(len(neg_val_samples))\n",
    "neg_test_label = np.zeros(len(neg_test_samples))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "215a7df0",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "eCLIP sequences are disjoint across splits\n"
     ]
    }
   ],
   "source": [
    "train_eclips = set(eclip for eclip, _ in neg_train_samples)\n",
    "val_eclips   = set(eclip for eclip, _ in neg_val_samples)\n",
    "test_eclips  = set(eclip for eclip, _ in neg_test_samples)\n",
    "\n",
    "assert train_eclips.isdisjoint(val_eclips), \"Overlap between train and val eCLIP sequences!\"\n",
    "assert train_eclips.isdisjoint(test_eclips), \"Overlap between train and test eCLIP sequences!\"\n",
    "assert val_eclips.isdisjoint(test_eclips), \"Overlap between val and test eCLIP sequences!\"\n",
    "print(\"eCLIP sequences are disjoint across splits\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "5f353c01",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "RBNS sequences are disjoint across splits\n"
     ]
    }
   ],
   "source": [
    "train_rbns = set(rbns for _, (rbns, _) in neg_train_samples)\n",
    "val_rbns   = set(rbns for _, (rbns, _) in neg_val_samples)\n",
    "test_rbns  = set(rbns for _, (rbns, _) in neg_test_samples)\n",
    "\n",
    "assert train_rbns.isdisjoint(val_rbns), \"Overlap between train and val RBNS sequences!\"\n",
    "assert train_rbns.isdisjoint(test_rbns), \"Overlap between train and test RBNS sequences!\"\n",
    "assert val_rbns.isdisjoint(test_rbns), \"Overlap between val and test RBNS sequences!\"\n",
    "print(\"RBNS sequences are disjoint across splits\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "5d450562",
   "metadata": {},
   "outputs": [],
   "source": [
    "# check across pos and neg samples\n",
    "assert train_eclips.isdisjoint(pos_train_eclips), \"Overlap between train and val eCLIP sequences!\"\n",
    "assert val_eclips.isdisjoint(pos_val_eclips), \"Overlap between train and test eCLIP sequences!\"\n",
    "assert test_eclips.isdisjoint(pos_test_eclips), \"Overlap between val and test eCLIP sequences!\"\n",
    "\n",
    "assert train_rbns.isdisjoint(pos_train_rbns), \"Overlap between train and val RBNS sequences!\"\n",
    "assert val_rbns.isdisjoint(pos_val_rbns), \"Overlap between train and test RBNS sequences!\"\n",
    "assert test_rbns.isdisjoint(pos_test_rbns), \"Overlap between val and test RBNS sequences!\""
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9cc54a9e",
   "metadata": {},
   "source": [
    "Combine for full splits"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "726d4838",
   "metadata": {},
   "outputs": [],
   "source": [
    "train_data = np.concatenate((pos_train_samples, neg_train_samples))\n",
    "val_data = np.concatenate((pos_val_samples, neg_val_samples))\n",
    "test_data = np.concatenate((pos_test_samples, neg_test_samples))\n",
    "\n",
    "train_labels = np.concatenate((pos_train_label, neg_train_label))\n",
    "val_labels = np.concatenate((pos_val_label, neg_val_label))\n",
    "test_labels = np.concatenate((pos_test_label, neg_test_label))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "id": "8ce65d9c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# get positive/negative label values\n",
    "positive_value = (\"TRA2A\", logfold_changes)\n",
    "negative_value = (\"Not TRA2A\", placeholder_logfold_changes)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "877f0473",
   "metadata": {},
   "source": [
    "## Save as dictionaries"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "16dcdac6",
   "metadata": {},
   "source": [
    "Save as pickle files since there are nested data structures"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "id": "eff9d82c",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pickle"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "cfda55f9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'\\nSaved data format for each split:\\n- Input\\n    - List containing:\\n        - eCLIP sequence: text\\n        - Tuple of RBNS sequence and binding affinity: (text, vector)\\n- Output\\n    - List containing:\\n        - RBP label: text\\n        - Logfold gene expression change: vector\\n        \\nLabels save separately to save storage:\\n- List containing:\\n    - RBP label: text\\n    - Logfold gene expression change: vector\\n'"
      ]
     },
     "execution_count": 40,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "'''\n",
    "Saved data format for each split:\n",
    "- Input\n",
    "    - List containing:\n",
    "        - eCLIP sequence: text\n",
    "        - Tuple of RBNS sequence and binding affinity: (text, vector)\n",
    "- Output\n",
    "    - List containing:\n",
    "        - RBP label: text\n",
    "        - Logfold gene expression change: vector\n",
    "        \n",
    "Labels save separately to save storage:\n",
    "- List containing:\n",
    "    - RBP label: text\n",
    "    - Logfold gene expression change: vector\n",
    "'''"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "id": "6c84bd69",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Save train\n",
    "with open(\"TRA2A/train_split.pkl\", \"wb\") as f:\n",
    "    pickle.dump((train_data, train_labels), f)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "id": "afb7be9c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Save val\n",
    "with open(\"TRA2A/val_split.pkl\", \"wb\") as f:\n",
    "    pickle.dump((val_data, val_labels), f)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "id": "6266df79",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Save test\n",
    "with open(\"TRA2A/test_split.pkl\", \"wb\") as f:\n",
    "    pickle.dump((test_data, test_labels), f)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "id": "95900830",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Save positive label\n",
    "with open(\"TRA2A/TRA2A_positive_label.pkl\", \"wb\") as f:\n",
    "    pickle.dump(positive_value, f)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "id": "fe972c58",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Save negative label\n",
    "with open(\"TRA2A/TRA2A_negative_label.pkl\", \"wb\") as f:\n",
    "    pickle.dump(negative_value, f)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "521fb44e",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "perturb",
   "language": "python",
   "name": "perturb"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
